{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bd4496a9-7902-47f0-ab65-397f90a2823c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pyfaidx import Fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2de0dc1b-817d-4318-ac46-0ad5aac9a23f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ID</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383518</td>\n",
       "      <td>21383521</td>\n",
       "      <td>chr18:21383518-21383521:ATG&gt;TTT</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383517</td>\n",
       "      <td>21383521</td>\n",
       "      <td>chr18:21383517-21383521:CATG&gt;C</td>\n",
       "      <td>CATG</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383517</td>\n",
       "      <td>21383520</td>\n",
       "      <td>chr18:21383517-21383520:CAT&gt;C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383517</td>\n",
       "      <td>21383518</td>\n",
       "      <td>chr18:21383517-21383518:C&gt;CC</td>\n",
       "      <td>C</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383517</td>\n",
       "      <td>21383518</td>\n",
       "      <td>chr18:21383517-21383518:C&gt;CCCC</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383514</td>\n",
       "      <td>21383518</td>\n",
       "      <td>chr18:21383514-21383518:GATC&gt;GATCCCCC</td>\n",
       "      <td>GATC</td>\n",
       "      <td>GATCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383518</td>\n",
       "      <td>21383521</td>\n",
       "      <td>chr18:21383518-21383521:ATG&gt;ATGATG</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ATGATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383672</td>\n",
       "      <td>21383675</td>\n",
       "      <td>chr18:21383672-21383675:CAG&gt;TTT</td>\n",
       "      <td>CAG</td>\n",
       "      <td>TTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383672</td>\n",
       "      <td>21383675</td>\n",
       "      <td>chr18:21383672-21383675:CAG&gt;TTTTT</td>\n",
       "      <td>CAG</td>\n",
       "      <td>TTTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383674</td>\n",
       "      <td>21383675</td>\n",
       "      <td>chr18:21383674-21383675:G&gt;GTTTTT</td>\n",
       "      <td>G</td>\n",
       "      <td>GTTTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383672</td>\n",
       "      <td>21383675</td>\n",
       "      <td>chr18:21383672-21383675:CAG&gt;T</td>\n",
       "      <td>CAG</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383671</td>\n",
       "      <td>21383675</td>\n",
       "      <td>chr18:21383671-21383675:GCAG&gt;G</td>\n",
       "      <td>GCAG</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383530</td>\n",
       "      <td>21383531</td>\n",
       "      <td>chr18:21383530:T&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383530</td>\n",
       "      <td>21383536</td>\n",
       "      <td>chr18:21383530-21383536:TATGCT&gt;ATCGGC</td>\n",
       "      <td>TATGCT</td>\n",
       "      <td>ATCGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383540</td>\n",
       "      <td>21383543</td>\n",
       "      <td>chr18:21383540-21383543:AAC&gt;TAA</td>\n",
       "      <td>AAC</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383521</td>\n",
       "      <td>21383522</td>\n",
       "      <td>chr18:21383521-21383522:G&gt;GT</td>\n",
       "      <td>G</td>\n",
       "      <td>GT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383521</td>\n",
       "      <td>21383522</td>\n",
       "      <td>chr18:21383521-21383522:G&gt;GTTT</td>\n",
       "      <td>G</td>\n",
       "      <td>GTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383531</td>\n",
       "      <td>21383535</td>\n",
       "      <td>chr18:21383531-21383535:ATGC&gt;A</td>\n",
       "      <td>ATGC</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383531</td>\n",
       "      <td>21383534</td>\n",
       "      <td>chr18:21383531-21383534:ATG&gt;A</td>\n",
       "      <td>ATG</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383515</td>\n",
       "      <td>21383521</td>\n",
       "      <td>chr18:21383515-21383521:ATCATG&gt;TTTAGGGG</td>\n",
       "      <td>ATCATG</td>\n",
       "      <td>TTTAGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383515</td>\n",
       "      <td>21383521</td>\n",
       "      <td>chr18:21383515-21383521:ATCATG&gt;T</td>\n",
       "      <td>ATCATG</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383672</td>\n",
       "      <td>21383678</td>\n",
       "      <td>chr18:21383672-21383678:CAGGTA&gt;GGG</td>\n",
       "      <td>CAGGTA</td>\n",
       "      <td>GGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383672</td>\n",
       "      <td>21383678</td>\n",
       "      <td>chr18:21383672-21383678:CAGGTA&gt;G</td>\n",
       "      <td>CAGGTA</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr18</td>\n",
       "      <td>21383672</td>\n",
       "      <td>21383678</td>\n",
       "      <td>chr18:21383672-21383678:CAGGTA&gt;GGGTTTT</td>\n",
       "      <td>CAGGTA</td>\n",
       "      <td>GGGTTTT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome     Start       End                                       ID  \\\n",
       "0       chr18  21383518  21383521          chr18:21383518-21383521:ATG>TTT   \n",
       "1       chr18  21383517  21383521           chr18:21383517-21383521:CATG>C   \n",
       "2       chr18  21383517  21383520            chr18:21383517-21383520:CAT>C   \n",
       "3       chr18  21383517  21383518             chr18:21383517-21383518:C>CC   \n",
       "4       chr18  21383517  21383518           chr18:21383517-21383518:C>CCCC   \n",
       "5       chr18  21383514  21383518    chr18:21383514-21383518:GATC>GATCCCCC   \n",
       "6       chr18  21383518  21383521       chr18:21383518-21383521:ATG>ATGATG   \n",
       "7       chr18  21383672  21383675          chr18:21383672-21383675:CAG>TTT   \n",
       "8       chr18  21383672  21383675        chr18:21383672-21383675:CAG>TTTTT   \n",
       "9       chr18  21383674  21383675         chr18:21383674-21383675:G>GTTTTT   \n",
       "10      chr18  21383672  21383675            chr18:21383672-21383675:CAG>T   \n",
       "11      chr18  21383671  21383675           chr18:21383671-21383675:GCAG>G   \n",
       "12      chr18  21383530  21383531                       chr18:21383530:T>G   \n",
       "13      chr18  21383530  21383536    chr18:21383530-21383536:TATGCT>ATCGGC   \n",
       "14      chr18  21383540  21383543          chr18:21383540-21383543:AAC>TAA   \n",
       "15      chr18  21383521  21383522             chr18:21383521-21383522:G>GT   \n",
       "16      chr18  21383521  21383522           chr18:21383521-21383522:G>GTTT   \n",
       "17      chr18  21383531  21383535           chr18:21383531-21383535:ATGC>A   \n",
       "18      chr18  21383531  21383534            chr18:21383531-21383534:ATG>A   \n",
       "19      chr18  21383515  21383521  chr18:21383515-21383521:ATCATG>TTTAGGGG   \n",
       "20      chr18  21383515  21383521         chr18:21383515-21383521:ATCATG>T   \n",
       "21      chr18  21383672  21383678       chr18:21383672-21383678:CAGGTA>GGG   \n",
       "22      chr18  21383672  21383678         chr18:21383672-21383678:CAGGTA>G   \n",
       "23      chr18  21383672  21383678   chr18:21383672-21383678:CAGGTA>GGGTTTT   \n",
       "\n",
       "       Ref       Alt  \n",
       "0      ATG       TTT  \n",
       "1     CATG         C  \n",
       "2      CAT         C  \n",
       "3        C        CC  \n",
       "4        C      CCCC  \n",
       "5     GATC  GATCCCCC  \n",
       "6      ATG    ATGATG  \n",
       "7      CAG       TTT  \n",
       "8      CAG     TTTTT  \n",
       "9        G    GTTTTT  \n",
       "10     CAG         T  \n",
       "11    GCAG         G  \n",
       "12       T         G  \n",
       "13  TATGCT    ATCGGC  \n",
       "14     AAC       TAA  \n",
       "15       G        GT  \n",
       "16       G      GTTT  \n",
       "17    ATGC         A  \n",
       "18     ATG         A  \n",
       "19  ATCATG  TTTAGGGG  \n",
       "20  ATCATG         T  \n",
       "21  CAGGTA       GGG  \n",
       "22  CAGGTA         G  \n",
       "23  CAGGTA   GGGTTTT  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transcript (ENST00000424526.7) in question: 21383518:21383675\t\n",
    "\n",
    "\n",
    "# TODO: Add Stop codon in middle of CDS\n",
    "\n",
    "test_variants_data = [\n",
    "    \n",
    "    # --- variants within start of CDS ---\n",
    "    \n",
    "    # Start codon substitution\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383518, \"End\": 21383521, \"ID\": \"chr18:21383518-21383521:ATG>TTT\", \"Ref\": \"ATG\", \"Alt\": \"TTT\"}, # CDS starts with TTT --> no start codon\n",
    "    \n",
    "    # Start codon deletion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383517, \"End\": 21383521, \"ID\": \"chr18:21383517-21383521:CATG>C\", \"Ref\": \"CATG\", \"Alt\": \"C\"}, # start Codon deletion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383517, \"End\": 21383520, \"ID\": \"chr18:21383517-21383520:CAT>C\", \"Ref\": \"CAT\", \"Alt\": \"C\"}, # CDS starts with C instead of ATG\n",
    "\n",
    "    # Start codon insertion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383517, \"End\": 21383518, \"ID\": \"chr18:21383517-21383518:C>CC\", \"Ref\": \"C\", \"Alt\": \"CC\"}, # CDS starts with CATG instead of ATG\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383517, \"End\": 21383518, \"ID\": \"chr18:21383517-21383518:C>CCCC\", \"Ref\": \"C\", \"Alt\": \"CCCC\"}, # CDS starts with CCCATG instead of ATG\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383514, \"End\": 21383518, \"ID\": \"chr18:21383514-21383518:GATC>GATCCCCC\", \"Ref\": \"GATC\", \"Alt\": \"GATCCCCC\"}, # CDS starts with CCCATG instead of ATG\n",
    "\n",
    "    # Start codon duplication\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383518, \"End\": 21383521, \"ID\": \"chr18:21383518-21383521:ATG>ATGATG\", \"Ref\": \"ATG\", \"Alt\": \"ATGATG\"}, # CDS starts with ATGATG instead of ATG\n",
    "\n",
    "    \n",
    "    # --- variants within end of CDS ---\n",
    "    \n",
    "    # End substitution\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383672, \"End\": 21383675, \"ID\": \"chr18:21383672-21383675:CAG>TTT\", \"Ref\": \"CAG\", \"Alt\": \"TTT\"}, \n",
    "\n",
    "    # End insertion: longer CDS\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383672, \"End\": 21383675, \"ID\": \"chr18:21383672-21383675:CAG>TTTTT\", \"Ref\": \"CAG\", \"Alt\": \"TTTTT\"},\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383674, \"End\": 21383675, \"ID\": \"chr18:21383674-21383675:G>GTTTTT\", \"Ref\": \"G\", \"Alt\": \"GTTTTT\"},\n",
    "\n",
    "    # End deletion: shorter CDS (If last CDS, maybe need to make new cases for stop codon)\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383672, \"End\": 21383675, \"ID\": \"chr18:21383672-21383675:CAG>T\", \"Ref\": \"CAG\", \"Alt\": \"T\"},\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383671, \"End\": 21383675, \"ID\": \"chr18:21383671-21383675:GCAG>G\", \"Ref\": \"GCAG\", \"Alt\": \"G\"},\n",
    "\n",
    "    # --- variants within CDS ---\n",
    "\n",
    "    # SNV\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383530, \"End\": 21383531, \"ID\": \"chr18:21383530-21383531:T>G\", \"Ref\": \"T\", \"Alt\": \"G\"},\n",
    "\n",
    "    # Substitution\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383530, \"End\": 21383536, \"ID\": \"chr18:21383530-21383536:TATGCT>ATCGGC\", \"Ref\": \"TATGCT\", \"Alt\": \"ATCGGC\"},\n",
    "\n",
    "    # Nonsense mutation\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383540, \"End\": 21383543, \"ID\": \"chr18:21383540-21383543:AAC>TAA\", \"Ref\": \"AAC\", \"Alt\": \"TAA\"},\n",
    "    \n",
    "    # Insertions\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383521, \"End\": 21383522, \"ID\": \"chr18:21383521-21383522:G>GT\", \"Ref\": \"G\", \"Alt\": \"GT\"},\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383521, \"End\": 21383522, \"ID\": \"chr18:21383521-21383522:G>GTTT\", \"Ref\": \"G\", \"Alt\": \"GTTT\"},\n",
    "    \n",
    "    # In-frame deletion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383531, \"End\": 21383535, \"ID\": \"chr18:21383531-21383535:ATGC>A\", \"Ref\": \"ATGC\", \"Alt\": \"A\"},\n",
    "    \n",
    "    # Frameshift deletion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383531, \"End\": 21383534, \"ID\": \"chr18:21383531-21383534:ATG>A\", \"Ref\": \"ATG\", \"Alt\": \"A\"},\n",
    "\n",
    "\n",
    "    # --- variants before CDS ---\n",
    "    \n",
    "    # Complex substitution: Deletion and Insertion at the Beginning of CDS\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383515, \"End\": 21383521, \"ID\": \"chr18:21383515-21383521:ATCATG>TTTAGGGG\", \"Ref\": \"ATCATG\", \"Alt\": \"TTTAGGGG\"}, # CDS starts with AGGGG\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383515, \"End\": 21383521, \"ID\": \"chr18:21383515-21383521:ATCATG>T\", \"Ref\": \"ATCATG\", \"Alt\": \"T\"}, # CDS starts after  ATG , i.e. ATG gets deleted\n",
    "\n",
    "    # --- variants after CDS ---\n",
    "    \n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383672, \"End\": 21383678, \"ID\": \"chr18:21383672-21383678:CAGGTA>GGG\", \"Ref\": \"CAGGTA\", \"Alt\": \"GGG\"}, # CDS same length\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383672, \"End\": 21383678, \"ID\": \"chr18:21383672-21383678:CAGGTA>G\", \"Ref\": \"CAGGTA\", \"Alt\": \"G\"}, # shorter CDS\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 21383672, \"End\": 21383678, \"ID\": \"chr18:21383672-21383678:CAGGTA>GGGTTTT\", \"Ref\": \"CAGGTA\", \"Alt\": \"GGGTTTT\"} # longer CDS\n",
    "    \n",
    "    \n",
    "]\n",
    "\n",
    "test_variants_df = pd.DataFrame(test_variants_data)\n",
    "test_variants_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bfa9f899-8adf-4f2d-a06a-8d77d4382b54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">chr18\n",
      "TTAGTGGCTTTCTCCTTTTTTGCCAACTAC\n",
      ">chr18\n",
      "TACCTA\n"
     ]
    }
   ],
   "source": [
    "fasta_path = \"/s/genomes/Gencode/Gencode_human/release_40/GRCh38.p13.genome.fa\"\n",
    "fasta = Fasta(fasta_path)\n",
    "print(f\">chr18\\n{fasta['chr18'][56603026:56603056]}\") \n",
    "print(f\">chr18\\n{fasta['chr18'][56603053:56603059]}\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b16080b0-aa1a-45a2-81ea-96aae8c6ce9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ID</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603029</td>\n",
       "      <td>56603032</td>\n",
       "      <td>chr18:56603029-56603032:GTG&gt;TTT</td>\n",
       "      <td>GTG</td>\n",
       "      <td>TTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603028</td>\n",
       "      <td>56603032</td>\n",
       "      <td>chr18:56603028-56603032:AGTG&gt;A</td>\n",
       "      <td>AGTG</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603028</td>\n",
       "      <td>56603031</td>\n",
       "      <td>chr18:56603028-56603031:AGT&gt;A</td>\n",
       "      <td>AGT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603029</td>\n",
       "      <td>56603032</td>\n",
       "      <td>chr18:56603029-56603032:GTG&gt;GTGATG</td>\n",
       "      <td>GTG</td>\n",
       "      <td>GTGATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603028</td>\n",
       "      <td>56603032</td>\n",
       "      <td>chr18:56603028-56603032:AGTG&gt;AGTTTATG</td>\n",
       "      <td>AGTG</td>\n",
       "      <td>AGTTTATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603053</td>\n",
       "      <td>56603056</td>\n",
       "      <td>chr18:56603053-56603056:TAC&gt;GGG</td>\n",
       "      <td>TAC</td>\n",
       "      <td>GGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603053</td>\n",
       "      <td>56603056</td>\n",
       "      <td>chr18:56603053-56603056:TAC&gt;TTTTT</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TTTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603055</td>\n",
       "      <td>56603056</td>\n",
       "      <td>chr18:56603055-56603056:C&gt;CTTTTT</td>\n",
       "      <td>C</td>\n",
       "      <td>CTTTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603053</td>\n",
       "      <td>56603056</td>\n",
       "      <td>chr18:56603053-56603056:TAC&gt;T</td>\n",
       "      <td>TAC</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603052</td>\n",
       "      <td>56603056</td>\n",
       "      <td>chr18:56603052-56603056:CTAC&gt;C</td>\n",
       "      <td>CTAC</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603040</td>\n",
       "      <td>56603041</td>\n",
       "      <td>chr18:56603040-56603041:C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603040</td>\n",
       "      <td>56603046</td>\n",
       "      <td>chr18:56603040-56603046:CTTTTT&gt;AAAAAA</td>\n",
       "      <td>CTTTTT</td>\n",
       "      <td>AAAAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603040</td>\n",
       "      <td>56603043</td>\n",
       "      <td>chr18:56603040-56603043:CTT&gt;AAA</td>\n",
       "      <td>CTT</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603040</td>\n",
       "      <td>56603041</td>\n",
       "      <td>chr18:56603040-56603041:C&gt;CT</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603040</td>\n",
       "      <td>56603041</td>\n",
       "      <td>chr18:56603040-56603041:C&gt;CTTT</td>\n",
       "      <td>C</td>\n",
       "      <td>CTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603040</td>\n",
       "      <td>56603044</td>\n",
       "      <td>chr18:56603040-56603044:CTTT&gt;A</td>\n",
       "      <td>CTTT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603040</td>\n",
       "      <td>56603043</td>\n",
       "      <td>chr18:56603040-56603043:CTT&gt;A</td>\n",
       "      <td>CTT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603037</td>\n",
       "      <td>56603043</td>\n",
       "      <td>chr18:56603037-56603043:CTCCTT&gt;TTTAGGGG</td>\n",
       "      <td>CTCCTT</td>\n",
       "      <td>TTTAGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603037</td>\n",
       "      <td>56603043</td>\n",
       "      <td>chr18:56603037-56603043:CTCCTT&gt;T</td>\n",
       "      <td>CTCCTT</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603053</td>\n",
       "      <td>56603059</td>\n",
       "      <td>chr18:56603053-56603059:TACCTA&gt;GGG</td>\n",
       "      <td>TACCTA</td>\n",
       "      <td>GGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603053</td>\n",
       "      <td>56603059</td>\n",
       "      <td>chr18:56603053-56603059:TACCTA&gt;G</td>\n",
       "      <td>TACCTA</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr18</td>\n",
       "      <td>56603053</td>\n",
       "      <td>56603059</td>\n",
       "      <td>chr18:56603053-56603059:TACCTA&gt;GGGTTTT</td>\n",
       "      <td>TACCTA</td>\n",
       "      <td>GGGTTTT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome     Start       End                                       ID  \\\n",
       "0       chr18  56603029  56603032          chr18:56603029-56603032:GTG>TTT   \n",
       "1       chr18  56603028  56603032           chr18:56603028-56603032:AGTG>A   \n",
       "2       chr18  56603028  56603031            chr18:56603028-56603031:AGT>A   \n",
       "3       chr18  56603029  56603032       chr18:56603029-56603032:GTG>GTGATG   \n",
       "4       chr18  56603028  56603032    chr18:56603028-56603032:AGTG>AGTTTATG   \n",
       "5       chr18  56603053  56603056          chr18:56603053-56603056:TAC>GGG   \n",
       "6       chr18  56603053  56603056        chr18:56603053-56603056:TAC>TTTTT   \n",
       "7       chr18  56603055  56603056         chr18:56603055-56603056:C>CTTTTT   \n",
       "8       chr18  56603053  56603056            chr18:56603053-56603056:TAC>T   \n",
       "9       chr18  56603052  56603056           chr18:56603052-56603056:CTAC>C   \n",
       "10      chr18  56603040  56603041              chr18:56603040-56603041:C>T   \n",
       "11      chr18  56603040  56603046    chr18:56603040-56603046:CTTTTT>AAAAAA   \n",
       "12      chr18  56603040  56603043          chr18:56603040-56603043:CTT>AAA   \n",
       "13      chr18  56603040  56603041             chr18:56603040-56603041:C>CT   \n",
       "14      chr18  56603040  56603041           chr18:56603040-56603041:C>CTTT   \n",
       "15      chr18  56603040  56603044           chr18:56603040-56603044:CTTT>A   \n",
       "16      chr18  56603040  56603043            chr18:56603040-56603043:CTT>A   \n",
       "17      chr18  56603037  56603043  chr18:56603037-56603043:CTCCTT>TTTAGGGG   \n",
       "18      chr18  56603037  56603043         chr18:56603037-56603043:CTCCTT>T   \n",
       "19      chr18  56603053  56603059       chr18:56603053-56603059:TACCTA>GGG   \n",
       "20      chr18  56603053  56603059         chr18:56603053-56603059:TACCTA>G   \n",
       "21      chr18  56603053  56603059   chr18:56603053-56603059:TACCTA>GGGTTTT   \n",
       "\n",
       "       Ref       Alt  \n",
       "0      GTG       TTT  \n",
       "1     AGTG         A  \n",
       "2      AGT         A  \n",
       "3      GTG    GTGATG  \n",
       "4     AGTG  AGTTTATG  \n",
       "5      TAC       GGG  \n",
       "6      TAC     TTTTT  \n",
       "7        C    CTTTTT  \n",
       "8      TAC         T  \n",
       "9     CTAC         C  \n",
       "10       C         T  \n",
       "11  CTTTTT    AAAAAA  \n",
       "12     CTT       AAA  \n",
       "13       C        CT  \n",
       "14       C      CTTT  \n",
       "15    CTTT         A  \n",
       "16     CTT         A  \n",
       "17  CTCCTT  TTTAGGGG  \n",
       "18  CTCCTT         T  \n",
       "19  TACCTA       GGG  \n",
       "20  TACCTA         G  \n",
       "21  TACCTA   GGGTTTT  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create variants file for minus strand example: ENST00000217515.11, 56603029:56603056\n",
    "\n",
    "minus_strand_variants_data = [\n",
    "    \n",
    "    # Beginning of CDS (which is the end of the CDS on the minus strand)\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603029, \"End\": 56603032, \"ID\": \"chr18:56603029-56603032:GTG>TTT\", \"Ref\": \"GTG\", \"Alt\": \"TTT\"}, # Substitution\n",
    "\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603028, \"End\": 56603032, \"ID\": \"chr18:56603028-56603032:AGTG>A\", \"Ref\": \"AGTG\", \"Alt\": \"A\"}, # Deletion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603028, \"End\": 56603031, \"ID\": \"chr18:56603028-56603031:AGT>A\", \"Ref\": \"AGT\", \"Alt\": \"A\"}, # Deletion\n",
    "\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603029, \"End\": 56603032, \"ID\": \"chr18:56603029-56603032:GTG>GTGATG\", \"Ref\": \"GTG\", \"Alt\": \"GTGATG\"}, # Insertion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603028, \"End\": 56603032, \"ID\": \"chr18:56603028-56603032:AGTG>AGTTTATG\", \"Ref\": \"AGTG\", \"Alt\": \"AGTTTATG\"}, # Insertion\n",
    "    \n",
    "    # End of CDS (which is the beginning of the CDS on the minus strand)\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603053, \"End\": 56603056, \"ID\": \"chr18:56603053-56603056:TAC>GGG\", \"Ref\": \"TAC\", \"Alt\": \"GGG\"}, # Substitution\n",
    "\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603053, \"End\": 56603056, \"ID\": \"chr18:56603053-56603056:TAC>TTTTT\", \"Ref\": \"TAC\", \"Alt\": \"TTTTT\"}, # Insertion + Substitution\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603055, \"End\": 56603056, \"ID\": \"chr18:56603055-56603056:C>CTTTTT\", \"Ref\": \"C\", \"Alt\": \"CTTTTT\"}, # Insertion\n",
    "\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603053, \"End\": 56603056, \"ID\": \"chr18:56603053-56603056:TAC>T\", \"Ref\": \"TAC\", \"Alt\": \"T\"}, # Deletion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603052, \"End\": 56603056, \"ID\": \"chr18:56603052-56603056:CTAC>C\", \"Ref\": \"CTAC\", \"Alt\": \"C\"}, # Deletion über CDS hinaus\n",
    "\n",
    "    # Within CD\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603040, \"End\": 56603041, \"ID\": \"chr18:56603040-56603041:C>T\", \"Ref\": \"C\", \"Alt\": \"T\"}, # SNV\n",
    "\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603040, \"End\": 56603046, \"ID\": \"chr18:56603040-56603046:CTTTTT>AAAAAA\", \"Ref\": \"CTTTTT\", \"Alt\": \"AAAAAA\"}, # Substitution\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603040, \"End\": 56603043, \"ID\": \"chr18:56603040-56603043:CTT>AAA\", \"Ref\": \"CTT\", \"Alt\": \"AAA\"}, # Substitution\n",
    "\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603040, \"End\": 56603041, \"ID\": \"chr18:56603040-56603041:C>CT\", \"Ref\": \"C\", \"Alt\": \"CT\"}, # Insertion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603040, \"End\": 56603041, \"ID\": \"chr18:56603040-56603041:C>CTTT\", \"Ref\": \"C\", \"Alt\": \"CTTT\"}, # Insertion\n",
    "    \n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603040, \"End\": 56603044, \"ID\": \"chr18:56603040-56603044:CTTT>A\", \"Ref\": \"CTTT\", \"Alt\": \"A\"}, # Deletion\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603040, \"End\": 56603043, \"ID\": \"chr18:56603040-56603043:CTT>A\", \"Ref\": \"CTT\", \"Alt\": \"A\"}, # Deletion\n",
    "\n",
    "    # Variants Starting before Start of CDS\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603037, \"End\": 56603043, \"ID\": \"chr18:56603037-56603043:CTCCTT>TTTAGGGG\", \"Ref\": \"CTCCTT\", \"Alt\": \"TTTAGGGG\"}, # CDS starts with AGGGG instead of CTT\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603037, \"End\": 56603043, \"ID\": \"chr18:56603037-56603043:CTCCTT>T\", \"Ref\": \"CTCCTT\", \"Alt\": \"T\"}, # Shorter CDS, -3\n",
    "\n",
    "    # Variants going over End of CDS\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603053, \"End\": 56603059, \"ID\": \"chr18:56603053-56603059:TACCTA>GGG\", \"Ref\": \"TACCTA\", \"Alt\": \"GGG\"}, # CDS same length\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603053, \"End\": 56603059, \"ID\": \"chr18:56603053-56603059:TACCTA>G\", \"Ref\": \"TACCTA\", \"Alt\": \"G\"}, # shorter CDS\n",
    "    {\"Chromosome\": \"chr18\", \"Start\": 56603053, \"End\": 56603059, \"ID\": \"chr18:56603053-56603059:TACCTA>GGGTTTT\", \"Ref\": \"TACCTA\", \"Alt\": \"GGGTTTT\"} # longer CDS\n",
    "\n",
    "\n",
    "]\n",
    "\n",
    "minus_strand_variants_df = pd.DataFrame(minus_strand_variants_data)\n",
    "minus_strand_variants_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a4afeed5-4476-49be-8833-f4530b2ff9b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Match at chr18:56603029: VCF=GTG, FASTA=GTG, ALT=TTT\n",
      "✅ Match at chr18:56603028: VCF=AGTG, FASTA=AGTG, ALT=A\n",
      "✅ Match at chr18:56603028: VCF=AGT, FASTA=AGT, ALT=A\n",
      "✅ Match at chr18:56603029: VCF=GTG, FASTA=GTG, ALT=GTGATG\n",
      "✅ Match at chr18:56603028: VCF=AGTG, FASTA=AGTG, ALT=AGTTTATG\n",
      "✅ Match at chr18:56603053: VCF=TAC, FASTA=TAC, ALT=GGG\n",
      "✅ Match at chr18:56603053: VCF=TAC, FASTA=TAC, ALT=TTTTT\n",
      "✅ Match at chr18:56603055: VCF=C, FASTA=C, ALT=CTTTTT\n",
      "✅ Match at chr18:56603053: VCF=TAC, FASTA=TAC, ALT=T\n",
      "✅ Match at chr18:56603052: VCF=CTAC, FASTA=CTAC, ALT=C\n",
      "✅ Match at chr18:56603040: VCF=C, FASTA=C, ALT=T\n",
      "✅ Match at chr18:56603040: VCF=CTTTTT, FASTA=CTTTTT, ALT=AAAAAA\n",
      "✅ Match at chr18:56603040: VCF=CTT, FASTA=CTT, ALT=AAA\n",
      "✅ Match at chr18:56603040: VCF=C, FASTA=C, ALT=CT\n",
      "✅ Match at chr18:56603040: VCF=C, FASTA=C, ALT=CTTT\n",
      "✅ Match at chr18:56603040: VCF=CTTT, FASTA=CTTT, ALT=A\n",
      "✅ Match at chr18:56603040: VCF=CTT, FASTA=CTT, ALT=A\n",
      "✅ Match at chr18:56603037: VCF=CTCCTT, FASTA=CTCCTT, ALT=TTTAGGGG\n",
      "✅ Match at chr18:56603037: VCF=CTCCTT, FASTA=CTCCTT, ALT=T\n",
      "✅ Match at chr18:56603053: VCF=TACCTA, FASTA=TACCTA, ALT=GGG\n",
      "✅ Match at chr18:56603053: VCF=TACCTA, FASTA=TACCTA, ALT=G\n",
      "✅ Match at chr18:56603053: VCF=TACCTA, FASTA=TACCTA, ALT=GGGTTTT\n"
     ]
    }
   ],
   "source": [
    "from pyfaidx import Fasta\n",
    "\n",
    "fasta_path = \"/s/genomes/Gencode/Gencode_human/release_40/GRCh38.p13.genome.fa\"\n",
    "fasta = Fasta(fasta_path)\n",
    "\n",
    "# for idx, row in test_variants_df.iterrows():\n",
    "for idx, row in minus_strand_variants_df.iterrows():\n",
    "    \n",
    "    chrom = row[\"Chromosome\"]\n",
    "    pos = int(row[\"Start\"])  # VCF is 1-based\n",
    "    ref = row[\"Ref\"]\n",
    "    ref_len = len(ref)\n",
    "    alt = row[\"Alt\"]\n",
    "\n",
    "    # Normalize chromosome name\n",
    "    if chrom not in fasta:\n",
    "        chrom = chrom.replace(\"chr\", \"\")  # fallback if needed\n",
    "\n",
    "    if chrom not in fasta:\n",
    "        print(f\"❌ Chromosome {row['Chromosome']} not found in FASTA\")\n",
    "        continue\n",
    "\n",
    "    # pyfaidx slicing is 1-based and end-inclusive\n",
    "    fasta_seq = fasta[chrom][pos:pos + ref_len].seq.upper()\n",
    "\n",
    "    if fasta_seq != ref.upper():\n",
    "        print(f\"⚠️ Ref mismatch at {row['Chromosome']}:{pos}: VCF={ref}, FASTA={fasta_seq}, ALT={alt}\")\n",
    "    else:\n",
    "        print(f\"✅ Match at {row['Chromosome']}:{pos}: VCF={ref}, FASTA={fasta_seq}, ALT={alt}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fe155e53-061b-49af-b94c-bf9be358acef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">chr18\n",
      "GATC\n"
     ]
    }
   ],
   "source": [
    "print(f\">chr18\\n{fasta['chr18'][21383514:21383518]}\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7124489-d7e2-4e95-839a-22dfe1b43382",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "19337597-9bea-44e9-9832-dcb6bbae9b8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_header = [\n",
    "    \"##fileformat=VCFv4.2\",\n",
    "    '##INFO=<ID=.,Number=0,Type=String,Description=\"Placeholder INFO\">',\n",
    "    \"#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\"\n",
    "]\n",
    "\n",
    "# Prepare the VCF lines\n",
    "# vcf_body = test_variants_df.apply(\n",
    "vcf_body = minus_strand_variants_df.apply(\n",
    "    lambda row: f\"{row['Chromosome']}\\t{row['Start']}\\t{row['ID']}\\t{row['Ref']}\\t{row['Alt']}\\t.\\t.\\t.\",\n",
    "    axis=1\n",
    ").tolist()\n",
    "\n",
    "# Combine header and body\n",
    "vcf_lines = vcf_header + vcf_body\n",
    "\n",
    "# Save to file (can be found in /resources/test_files/)\n",
    "# with open(\"test_variants.vcf\", \"w\") as f:\n",
    "with open(\"test_variants_minus.vcf\", \"w\") as f:\n",
    "    for line in vcf_lines:\n",
    "        f.write(line + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc29a8fa-6cd7-4eeb-a7de-57f4718f1a21",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "NMD_test",
   "language": "python",
   "name": "nmd_test"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
