{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a30338c8-f9bb-4123-95b7-01da02c8c548",
   "metadata": {},
   "source": [
    "File to go step by step thorugh the NMD-Scanner steps, easier for testing, debugging and adjusting of the code because you can test it here first"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acd81ad1-d86a-4c9c-8a2f-e2189117d5b6",
   "metadata": {},
   "source": [
    "# Import Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "03ae0062-752b-496b-8c30-5f4f37c21f0e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /opt/modules/i12g/anaconda/envs/NMD_test:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_libgcc_mutex             0.1                 conda_forge    conda-forge\n",
      "_openmp_mutex             4.5                       2_gnu    conda-forge\n",
      "_python_abi3_support      1.0                  hd8ed1ab_2    conda-forge\n",
      "alsa-lib                  1.2.13               hb9d3cd8_0    conda-forge\n",
      "anyio                     4.9.0              pyh29332c3_0    conda-forge\n",
      "argcomplete               3.6.0                    pypi_0    pypi\n",
      "argh                      0.31.3                   pypi_0    pypi\n",
      "argon2-cffi               23.1.0             pyhd8ed1ab_1    conda-forge\n",
      "argon2-cffi-bindings      21.2.0          py311h9ecbd09_5    conda-forge\n",
      "arrow                     1.3.0              pyhd8ed1ab_1    conda-forge\n",
      "asttokens                 3.0.0              pyhd8ed1ab_1    conda-forge\n",
      "async-lru                 2.0.5              pyh29332c3_0    conda-forge\n",
      "attrs                     25.3.0             pyh71513ae_0    conda-forge\n",
      "aws-c-auth                0.9.0                h0a147a0_3    conda-forge\n",
      "aws-c-cal                 0.9.0                hada3f3f_0    conda-forge\n",
      "aws-c-common              0.12.2               hb9d3cd8_0    conda-forge\n",
      "aws-c-compression         0.3.1                hc2d532b_4    conda-forge\n",
      "aws-c-event-stream        0.5.4                hc5e5e9e_7    conda-forge\n",
      "aws-c-http                0.9.7                h6884c39_1    conda-forge\n",
      "aws-c-io                  0.18.1               h1a9f769_2    conda-forge\n",
      "aws-c-mqtt                0.12.3               h27aa219_3    conda-forge\n",
      "aws-c-s3                  0.7.15               hea6d4b9_2    conda-forge\n",
      "aws-c-sdkutils            0.2.3                hc2d532b_4    conda-forge\n",
      "aws-checksums             0.2.7                hc2d532b_0    conda-forge\n",
      "aws-crt-cpp               0.32.4               h9a0fb62_1    conda-forge\n",
      "aws-sdk-cpp               1.11.510             h5b777a2_6    conda-forge\n",
      "azure-core-cpp            1.14.0               h5cfcd09_0    conda-forge\n",
      "azure-identity-cpp        1.10.0               h113e628_0    conda-forge\n",
      "azure-storage-blobs-cpp   12.13.0              h3cf044e_1    conda-forge\n",
      "azure-storage-common-cpp  12.8.0               h736e048_1    conda-forge\n",
      "azure-storage-files-datalake-cpp 12.12.0              ha633028_1    conda-forge\n",
      "babel                     2.17.0             pyhd8ed1ab_0    conda-forge\n",
      "beautifulsoup4            4.13.3             pyha770c72_0    conda-forge\n",
      "biopython                 1.85            py311h9ecbd09_1    conda-forge\n",
      "bleach                    6.2.0              pyh29332c3_4    conda-forge\n",
      "bleach-with-css           6.2.0                h82add2a_4    conda-forge\n",
      "boto3                     1.38.4             pyhd8ed1ab_0    conda-forge\n",
      "botocore                  1.38.4          pyge310_1234567_0    conda-forge\n",
      "brotli                    1.1.0                hb9d3cd8_2    conda-forge\n",
      "brotli-bin                1.1.0                hb9d3cd8_2    conda-forge\n",
      "brotli-python             1.1.0           py311hfdbb021_2    conda-forge\n",
      "bzip2                     1.0.8                h4bc722e_7    conda-forge\n",
      "c-ares                    1.34.5               hb9d3cd8_0    conda-forge\n",
      "ca-certificates           2025.8.3             hbd8a1cb_0    conda-forge\n",
      "cached-property           1.5.2                hd8ed1ab_1    conda-forge\n",
      "cached_property           1.5.2              pyha770c72_1    conda-forge\n",
      "cairo                     1.18.4               h3394656_0    conda-forge\n",
      "certifi                   2025.8.3           pyhd8ed1ab_0    conda-forge\n",
      "cffi                      1.17.1          py311hf29c0ef_0    conda-forge\n",
      "charset-normalizer        3.4.1              pyhd8ed1ab_0    conda-forge\n",
      "click                     8.1.8                    pypi_0    pypi\n",
      "colorama                  0.4.6              pyhd8ed1ab_1    conda-forge\n",
      "coloredlogs               15.0.1                   pypi_0    pypi\n",
      "comm                      0.2.2              pyhd8ed1ab_1    conda-forge\n",
      "contourpy                 1.3.1           py311hd18a35c_0    conda-forge\n",
      "cpython                   3.11.13         py311hd8ed1ab_0    conda-forge\n",
      "cycler                    0.12.1             pyhd8ed1ab_1    conda-forge\n",
      "cyrus-sasl                2.1.27               h54b06d7_7    conda-forge\n",
      "cyvcf2                    0.31.1                   pypi_0    pypi\n",
      "dbus                      1.13.6               h5008d03_3    conda-forge\n",
      "debugpy                   1.8.13          py311hfdbb021_0    conda-forge\n",
      "decorator                 5.2.1              pyhd8ed1ab_0    conda-forge\n",
      "defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge\n",
      "double-conversion         3.3.1                h5888daf_0    conda-forge\n",
      "exceptiongroup            1.2.2              pyhd8ed1ab_1    conda-forge\n",
      "executing                 2.1.0              pyhd8ed1ab_1    conda-forge\n",
      "expat                     2.6.4                h5888daf_0    conda-forge\n",
      "font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge\n",
      "font-ttf-inconsolata      3.000                h77eed37_0    conda-forge\n",
      "font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge\n",
      "font-ttf-ubuntu           0.83                 h77eed37_3    conda-forge\n",
      "fontconfig                2.15.0               h7e30c49_1    conda-forge\n",
      "fonts-conda-ecosystem     1                             0    conda-forge\n",
      "fonts-conda-forge         1                             0    conda-forge\n",
      "fonttools                 4.56.0          py311h2dc5d0c_0    conda-forge\n",
      "fqdn                      1.5.1              pyhd8ed1ab_1    conda-forge\n",
      "freetype                  2.13.3               h48d6fc4_0    conda-forge\n",
      "gffutils                  0.13                     pypi_0    pypi\n",
      "gflags                    2.2.2             h5888daf_1005    conda-forge\n",
      "glog                      0.7.1                hbabe93e_0    conda-forge\n",
      "graphite2                 1.3.13            h59595ed_1003    conda-forge\n",
      "h11                       0.14.0             pyhd8ed1ab_1    conda-forge\n",
      "h2                        4.2.0              pyhd8ed1ab_0    conda-forge\n",
      "halo                      0.0.31             pyhd8ed1ab_1    conda-forge\n",
      "harfbuzz                  10.4.0               h76408a6_0    conda-forge\n",
      "hpack                     4.1.0              pyhd8ed1ab_0    conda-forge\n",
      "htslib                    1.22.1               h566b1c6_0    bioconda\n",
      "httpcore                  1.0.7              pyh29332c3_1    conda-forge\n",
      "httpx                     0.28.1             pyhd8ed1ab_0    conda-forge\n",
      "humanfriendly             10.0                     pypi_0    pypi\n",
      "hyperframe                6.1.0              pyhd8ed1ab_0    conda-forge\n",
      "icu                       75.1                 he02047a_0    conda-forge\n",
      "idna                      3.10               pyhd8ed1ab_1    conda-forge\n",
      "importlib-metadata        8.6.1              pyha770c72_0    conda-forge\n",
      "importlib_resources       6.5.2              pyhd8ed1ab_0    conda-forge\n",
      "ipykernel                 6.29.5             pyh3099207_0    conda-forge\n",
      "ipython                   9.0.2              pyhfb0248b_0    conda-forge\n",
      "ipython_pygments_lexers   1.1.1              pyhd8ed1ab_0    conda-forge\n",
      "isoduration               20.11.0            pyhd8ed1ab_1    conda-forge\n",
      "jedi                      0.19.2             pyhd8ed1ab_1    conda-forge\n",
      "jinja2                    3.1.6              pyhd8ed1ab_0    conda-forge\n",
      "jmespath                  1.0.1              pyhd8ed1ab_1    conda-forge\n",
      "joblib                    1.4.2              pyhd8ed1ab_1    conda-forge\n",
      "json5                     0.10.0             pyhd8ed1ab_1    conda-forge\n",
      "jsonpointer               3.0.0           py311h38be061_1    conda-forge\n",
      "jsonschema                4.23.0             pyhd8ed1ab_1    conda-forge\n",
      "jsonschema-specifications 2024.10.1          pyhd8ed1ab_1    conda-forge\n",
      "jsonschema-with-format-nongpl 4.23.0               hd8ed1ab_1    conda-forge\n",
      "jupyter-lsp               2.2.5              pyhd8ed1ab_1    conda-forge\n",
      "jupyter_client            8.6.3              pyhd8ed1ab_1    conda-forge\n",
      "jupyter_core              5.7.2              pyh31011fe_1    conda-forge\n",
      "jupyter_events            0.12.0             pyh29332c3_0    conda-forge\n",
      "jupyter_server            2.15.0             pyhd8ed1ab_0    conda-forge\n",
      "jupyter_server_terminals  0.5.3              pyhd8ed1ab_1    conda-forge\n",
      "jupyterlab                4.3.6              pyhd8ed1ab_0    conda-forge\n",
      "jupyterlab_pygments       0.3.0              pyhd8ed1ab_2    conda-forge\n",
      "jupyterlab_server         2.27.3             pyhd8ed1ab_1    conda-forge\n",
      "keyutils                  1.6.1                h166bdaf_0    conda-forge\n",
      "kiwisolver                1.4.7           py311hd18a35c_0    conda-forge\n",
      "krb5                      1.21.3               h659f571_0    conda-forge\n",
      "lcms2                     2.17                 h717163a_0    conda-forge\n",
      "ld_impl_linux-64          2.43                 h712a8e2_4    conda-forge\n",
      "lerc                      4.0.0                h27087fc_0    conda-forge\n",
      "libabseil                 20250127.1      cxx17_hbbce691_0    conda-forge\n",
      "libarrow                  19.0.1           h27f8bab_8_cpu    conda-forge\n",
      "libarrow-acero            19.0.1           hcb10f89_8_cpu    conda-forge\n",
      "libarrow-dataset          19.0.1           hcb10f89_8_cpu    conda-forge\n",
      "libarrow-substrait        19.0.1           h1bed206_8_cpu    conda-forge\n",
      "libblas                   3.9.0           31_h59b9bed_openblas    conda-forge\n",
      "libboost                  1.88.0               h6c02f8c_0    conda-forge\n",
      "libbrotlicommon           1.1.0                hb9d3cd8_2    conda-forge\n",
      "libbrotlidec              1.1.0                hb9d3cd8_2    conda-forge\n",
      "libbrotlienc              1.1.0                hb9d3cd8_2    conda-forge\n",
      "libcblas                  3.9.0           31_he106b2a_openblas    conda-forge\n",
      "libclang-cpp19.1          19.1.7          default_hb5137d0_2    conda-forge\n",
      "libclang13                19.1.7          default_h9c6a7e4_2    conda-forge\n",
      "libcrc32c                 1.1.2                h9c3ff4c_0    conda-forge\n",
      "libcups                   2.3.3                h4637d8d_4    conda-forge\n",
      "libcurl                   8.14.1               h332b0f4_0    conda-forge\n",
      "libdeflate                1.22                 hb9d3cd8_0    conda-forge\n",
      "libdrm                    2.4.124              hb9d3cd8_0    conda-forge\n",
      "libedit                   3.1.20250104    pl5321h7949ede_0    conda-forge\n",
      "libegl                    1.7.0                ha4b6fd6_2    conda-forge\n",
      "libev                     4.33                 hd590300_2    conda-forge\n",
      "libevent                  2.1.12               hf998b51_1    conda-forge\n",
      "libexpat                  2.6.4                h5888daf_0    conda-forge\n",
      "libffi                    3.4.6                h2dba641_0    conda-forge\n",
      "libgcc                    14.2.0               h767d61c_2    conda-forge\n",
      "libgcc-ng                 14.2.0               h69a702a_2    conda-forge\n",
      "libgfortran               14.2.0               h69a702a_2    conda-forge\n",
      "libgfortran5              14.2.0               hf1ad2bd_2    conda-forge\n",
      "libgl                     1.7.0                ha4b6fd6_2    conda-forge\n",
      "libglib                   2.82.2               h2ff4ddf_1    conda-forge\n",
      "libglvnd                  1.7.0                ha4b6fd6_2    conda-forge\n",
      "libglx                    1.7.0                ha4b6fd6_2    conda-forge\n",
      "libgomp                   14.2.0               h767d61c_2    conda-forge\n",
      "libgoogle-cloud           2.36.0               hc4361e1_1    conda-forge\n",
      "libgoogle-cloud-storage   2.36.0               h0121fbd_1    conda-forge\n",
      "libgrpc                   1.71.0               h8e591d7_1    conda-forge\n",
      "libiconv                  1.18                 h4ce23a2_1    conda-forge\n",
      "libjpeg-turbo             3.0.0                hd590300_1    conda-forge\n",
      "liblapack                 3.9.0           31_h7ac8fdf_openblas    conda-forge\n",
      "liblightgbm               4.6.0            cpu_h1ca3010_3    conda-forge\n",
      "libllvm19                 19.1.7               ha7bfdaf_1    conda-forge\n",
      "liblzma                   5.8.1                hb9d3cd8_2    conda-forge\n",
      "libnghttp2                1.64.0               h161d5f1_0    conda-forge\n",
      "libnsl                    2.0.1                hd590300_0    conda-forge\n",
      "libntlm                   1.8                  hb9d3cd8_0    conda-forge\n",
      "libopenblas               0.3.29          pthreads_h94d23a6_0    conda-forge\n",
      "libopengl                 1.7.0                ha4b6fd6_2    conda-forge\n",
      "libopentelemetry-cpp      1.20.0               hd1b1c89_0    conda-forge\n",
      "libopentelemetry-cpp-headers 1.20.0               ha770c72_0    conda-forge\n",
      "libparquet                19.0.1           h081d1f1_8_cpu    conda-forge\n",
      "libpciaccess              0.18                 hd590300_0    conda-forge\n",
      "libpng                    1.6.47               h943b412_0    conda-forge\n",
      "libpq                     17.4                 h27ae623_0    conda-forge\n",
      "libprotobuf               5.29.3               h501fc15_1    conda-forge\n",
      "libre2-11                 2024.07.02           hba17884_3    conda-forge\n",
      "libsodium                 1.0.20               h4ab18f5_0    conda-forge\n",
      "libsqlite                 3.49.1               hee588c1_2    conda-forge\n",
      "libssh2                   1.11.1               hcf80075_0    conda-forge\n",
      "libstdcxx                 14.2.0               h8f9b012_2    conda-forge\n",
      "libstdcxx-ng              14.2.0               h4852527_2    conda-forge\n",
      "libthrift                 0.21.0               h0e7cc3e_0    conda-forge\n",
      "libtiff                   4.7.0                hc4654cb_2    conda-forge\n",
      "libutf8proc               2.10.0               h4c51ac1_0    conda-forge\n",
      "libuuid                   2.38.1               h0b41bf4_0    conda-forge\n",
      "libwebp-base              1.5.0                h851e524_0    conda-forge\n",
      "libxcb                    1.17.0               h8a09558_0    conda-forge\n",
      "libxcrypt                 4.4.36               hd590300_1    conda-forge\n",
      "libxgboost                3.0.3            cpu_h2ebb00f_0    conda-forge\n",
      "libxkbcommon              1.8.1                hc4a0caf_0    conda-forge\n",
      "libxml2                   2.13.6               h8d12d68_0    conda-forge\n",
      "libxslt                   1.1.39               h76b75d6_0    conda-forge\n",
      "libzlib                   1.3.1                hb9d3cd8_2    conda-forge\n",
      "lightgbm                  4.6.0                  cpu_py_3    conda-forge\n",
      "log_symbols               0.0.14             pyhd8ed1ab_2    conda-forge\n",
      "lz4-c                     1.10.0               h5888daf_1    conda-forge\n",
      "markupsafe                3.0.2           py311h2dc5d0c_1    conda-forge\n",
      "matplotlib                3.10.1          py311h38be061_0    conda-forge\n",
      "matplotlib-base           3.10.1          py311h2b939e6_0    conda-forge\n",
      "matplotlib-inline         0.1.7              pyhd8ed1ab_1    conda-forge\n",
      "matplotlib-venn           1.1.2                    pypi_0    pypi\n",
      "mistune                   3.1.2              pyhd8ed1ab_0    conda-forge\n",
      "mizani                    0.13.3                   pypi_0    pypi\n",
      "munkres                   1.1.4              pyh9f0ad1d_0    conda-forge\n",
      "mysql-common              9.0.1                h266115a_5    conda-forge\n",
      "mysql-libs                9.0.1                he0572af_5    conda-forge\n",
      "natsort                   8.4.0                    pypi_0    pypi\n",
      "nbclient                  0.10.2             pyhd8ed1ab_0    conda-forge\n",
      "nbconvert-core            7.16.6             pyh29332c3_0    conda-forge\n",
      "nbformat                  5.10.4             pyhd8ed1ab_1    conda-forge\n",
      "ncls                      0.0.68                   pypi_0    pypi\n",
      "ncurses                   6.5                  h2d0b736_3    conda-forge\n",
      "nest-asyncio              1.6.0              pyhd8ed1ab_1    conda-forge\n",
      "nlohmann_json             3.12.0               h3f2d84a_0    conda-forge\n",
      "nmd-scanner               0.1.0                    pypi_0    pypi\n",
      "notebook-shim             0.2.4              pyhd8ed1ab_1    conda-forge\n",
      "numpy                     2.2.4           py311h5d046bc_0    conda-forge\n",
      "ocl-icd                   2.3.3                hb9d3cd8_0    conda-forge\n",
      "opencl-headers            2025.06.13           h5888daf_0    conda-forge\n",
      "openjpeg                  2.5.3                h5fbd93e_0    conda-forge\n",
      "openldap                  2.6.9                he970967_0    conda-forge\n",
      "openssl                   3.5.2                h26f9b46_0    conda-forge\n",
      "orc                       2.1.1                h17f744e_1    conda-forge\n",
      "overrides                 7.7.0              pyhd8ed1ab_1    conda-forge\n",
      "packaging                 24.2               pyhd8ed1ab_2    conda-forge\n",
      "pandas                    2.2.3           py311h7db5c69_1    conda-forge\n",
      "pandocfilters             1.5.0              pyhd8ed1ab_0    conda-forge\n",
      "parquet-tools             0.2.16             pyhd8ed1ab_1    conda-forge\n",
      "parso                     0.8.4              pyhd8ed1ab_1    conda-forge\n",
      "patsy                     1.0.1              pyhd8ed1ab_1    conda-forge\n",
      "pcre2                     10.44                hba22ea6_2    conda-forge\n",
      "pexpect                   4.9.0              pyhd8ed1ab_1    conda-forge\n",
      "pickleshare               0.7.5           pyhd8ed1ab_1004    conda-forge\n",
      "pillow                    11.1.0          py311h1322bbf_0    conda-forge\n",
      "pip                       25.0.1             pyh8b19718_0    conda-forge\n",
      "pixman                    0.44.2               h29eaf8c_0    conda-forge\n",
      "pkgutil-resolve-name      1.3.10             pyhd8ed1ab_2    conda-forge\n",
      "platformdirs              4.3.6              pyhd8ed1ab_1    conda-forge\n",
      "plotnine                  0.14.5                   pypi_0    pypi\n",
      "polars                    1.31.0          default_h70f2ef1_1    conda-forge\n",
      "polars-default            1.31.0           py39hf521cc8_1    conda-forge\n",
      "prometheus-cpp            1.3.0                ha5d0236_0    conda-forge\n",
      "prometheus_client         0.21.1             pyhd8ed1ab_0    conda-forge\n",
      "prompt-toolkit            3.0.50             pyha770c72_0    conda-forge\n",
      "psutil                    7.0.0           py311h9ecbd09_0    conda-forge\n",
      "pthread-stubs             0.4               hb9d3cd8_1002    conda-forge\n",
      "ptyprocess                0.7.0              pyhd8ed1ab_1    conda-forge\n",
      "pure_eval                 0.2.3              pyhd8ed1ab_1    conda-forge\n",
      "py-xgboost                3.0.3           cpu_pyh718b53a_0    conda-forge\n",
      "py4j                      0.10.9.7                 pypi_0    pypi\n",
      "pyarrow                   19.0.1          py311h38be061_0    conda-forge\n",
      "pyarrow-core              19.0.1          py311h4854187_0_cpu    conda-forge\n",
      "pycparser                 2.22               pyh29332c3_1    conda-forge\n",
      "pyfaidx                   0.8.1.3            pyhdfd78af_0    bioconda\n",
      "pygments                  2.19.1             pyhd8ed1ab_0    conda-forge\n",
      "pyparsing                 3.2.1              pyhd8ed1ab_0    conda-forge\n",
      "pyranges                  0.1.4                    pypi_0    pypi\n",
      "pyside6                   6.8.2           py311h9053184_1    conda-forge\n",
      "pysocks                   1.7.1              pyha55dd90_7    conda-forge\n",
      "pyspark                   3.5.5                    pypi_0    pypi\n",
      "python                    3.11.11         h9e4cc4f_2_cpython    conda-forge\n",
      "python-dateutil           2.9.0.post0        pyhff2d567_1    conda-forge\n",
      "python-fastjsonschema     2.21.1             pyhd8ed1ab_0    conda-forge\n",
      "python-gil                3.11.13              hd8ed1ab_0    conda-forge\n",
      "python-json-logger        2.0.7              pyhd8ed1ab_0    conda-forge\n",
      "python-tzdata             2025.1             pyhd8ed1ab_0    conda-forge\n",
      "python_abi                3.11                    5_cp311    conda-forge\n",
      "pytz                      2024.1             pyhd8ed1ab_0    conda-forge\n",
      "pyvcf3                    1.0.3              pyhdfd78af_0    bioconda\n",
      "pyyaml                    6.0.2           py311h2dc5d0c_2    conda-forge\n",
      "pyzmq                     26.3.0          py311h7deb3e3_0    conda-forge\n",
      "qhull                     2020.2               h434a139_5    conda-forge\n",
      "qt6-main                  6.8.2                h588cce1_0    conda-forge\n",
      "re2                       2024.07.02           h9925aae_3    conda-forge\n",
      "readline                  8.2                  h8c095d6_2    conda-forge\n",
      "referencing               0.36.2             pyh29332c3_0    conda-forge\n",
      "requests                  2.32.3             pyhd8ed1ab_1    conda-forge\n",
      "rfc3339-validator         0.1.4              pyhd8ed1ab_1    conda-forge\n",
      "rfc3986-validator         0.1.1              pyh9f0ad1d_0    conda-forge\n",
      "rpds-py                   0.23.1          py311h687327b_0    conda-forge\n",
      "s2n                       1.5.17               hba75a32_0    conda-forge\n",
      "s3transfer                0.12.0             pyhd8ed1ab_0    conda-forge\n",
      "samtools                  1.22.1               h96c455f_0    bioconda\n",
      "scikit-learn              1.6.1           py311h57cc02b_0    conda-forge\n",
      "scipy                     1.15.2          py311h8f841c2_0    conda-forge\n",
      "seaborn                   0.13.2               hd8ed1ab_3    conda-forge\n",
      "seaborn-base              0.13.2             pyhd8ed1ab_3    conda-forge\n",
      "send2trash                1.8.3              pyh0d859eb_1    conda-forge\n",
      "setuptools                75.8.2             pyhff2d567_0    conda-forge\n",
      "simplejson                3.20.1                   pypi_0    pypi\n",
      "six                       1.17.0             pyhd8ed1ab_0    conda-forge\n",
      "snappy                    1.2.1                h8bd8927_1    conda-forge\n",
      "sniffio                   1.3.1              pyhd8ed1ab_1    conda-forge\n",
      "sorted-nearest            0.0.39                   pypi_0    pypi\n",
      "soupsieve                 2.5                pyhd8ed1ab_1    conda-forge\n",
      "spinners                  0.0.24             pyhd8ed1ab_1    conda-forge\n",
      "stack_data                0.6.3              pyhd8ed1ab_1    conda-forge\n",
      "statsmodels               0.14.4          py311h9f3472d_0    conda-forge\n",
      "tabulate                  0.9.0                    pypi_0    pypi\n",
      "termcolor                 3.0.1              pyhd8ed1ab_0    conda-forge\n",
      "terminado                 0.18.1             pyh0d859eb_0    conda-forge\n",
      "threadpoolctl             3.6.0              pyhecae5ae_0    conda-forge\n",
      "thrift                    0.20.0          py311hfdbb021_1    conda-forge\n",
      "tinycss2                  1.4.0              pyhd8ed1ab_0    conda-forge\n",
      "tk                        8.6.13          noxft_h4845f30_101    conda-forge\n",
      "tomli                     2.2.1              pyhd8ed1ab_1    conda-forge\n",
      "tornado                   6.4.2           py311h9ecbd09_0    conda-forge\n",
      "traitlets                 5.14.3             pyhd8ed1ab_1    conda-forge\n",
      "types-python-dateutil     2.9.0.20241206     pyhd8ed1ab_0    conda-forge\n",
      "typing-extensions         4.12.2               hd8ed1ab_1    conda-forge\n",
      "typing_extensions         4.12.2             pyha770c72_1    conda-forge\n",
      "typing_utils              0.1.0              pyhd8ed1ab_1    conda-forge\n",
      "tzdata                    2025a                h78e105d_0    conda-forge\n",
      "unicodedata2              16.0.0          py311h9ecbd09_0    conda-forge\n",
      "uri-template              1.3.0              pyhd8ed1ab_1    conda-forge\n",
      "urllib3                   2.3.0              pyhd8ed1ab_0    conda-forge\n",
      "wayland                   1.23.1               h3e06ad9_0    conda-forge\n",
      "wcwidth                   0.2.13             pyhd8ed1ab_1    conda-forge\n",
      "webcolors                 24.11.1            pyhd8ed1ab_0    conda-forge\n",
      "webencodings              0.5.1              pyhd8ed1ab_3    conda-forge\n",
      "websocket-client          1.8.0              pyhd8ed1ab_1    conda-forge\n",
      "wheel                     0.45.1             pyhd8ed1ab_1    conda-forge\n",
      "xcb-util                  0.4.1                hb711507_2    conda-forge\n",
      "xcb-util-cursor           0.1.5                hb9d3cd8_0    conda-forge\n",
      "xcb-util-image            0.4.0                hb711507_2    conda-forge\n",
      "xcb-util-keysyms          0.4.1                hb711507_0    conda-forge\n",
      "xcb-util-renderutil       0.3.10               hb711507_0    conda-forge\n",
      "xcb-util-wm               0.4.2                hb711507_0    conda-forge\n",
      "xgboost                   3.0.3           cpu_pyhb39878e_0    conda-forge\n",
      "xkeyboard-config          2.43                 hb9d3cd8_0    conda-forge\n",
      "xorg-libice               1.1.2                hb9d3cd8_0    conda-forge\n",
      "xorg-libsm                1.2.6                he73a12e_0    conda-forge\n",
      "xorg-libx11               1.8.12               h4f16b4b_0    conda-forge\n",
      "xorg-libxau               1.0.12               hb9d3cd8_0    conda-forge\n",
      "xorg-libxcomposite        0.4.6                hb9d3cd8_2    conda-forge\n",
      "xorg-libxcursor           1.2.3                hb9d3cd8_0    conda-forge\n",
      "xorg-libxdamage           1.1.6                hb9d3cd8_0    conda-forge\n",
      "xorg-libxdmcp             1.1.5                hb9d3cd8_0    conda-forge\n",
      "xorg-libxext              1.3.6                hb9d3cd8_0    conda-forge\n",
      "xorg-libxfixes            6.0.1                hb9d3cd8_0    conda-forge\n",
      "xorg-libxi                1.8.2                hb9d3cd8_0    conda-forge\n",
      "xorg-libxrandr            1.5.4                hb9d3cd8_0    conda-forge\n",
      "xorg-libxrender           0.9.12               hb9d3cd8_0    conda-forge\n",
      "xorg-libxtst              1.2.5                hb9d3cd8_3    conda-forge\n",
      "xorg-libxxf86vm           1.1.6                hb9d3cd8_0    conda-forge\n",
      "yaml                      0.2.5                h7f98852_2    conda-forge\n",
      "zeromq                    4.3.5                h3b0a872_7    conda-forge\n",
      "zipp                      3.21.0             pyhd8ed1ab_1    conda-forge\n",
      "zlib                      1.3.1                hb9d3cd8_2    conda-forge\n",
      "zstandard                 0.23.0          py311h9ecbd09_1    conda-forge\n",
      "zstd                      1.5.7                hb8e6e7a_1    conda-forge\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "conda list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "beb51ed1-137e-4176-8e04-e16feddb97cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import os\n",
    "from pyfaidx import Fasta\n",
    "import numpy as np\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "pd.options.display.max_columns = None # to show all columns of the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a218480c-d8aa-431f-b8c6-aea7526f8466",
   "metadata": {},
   "source": [
    "# Read files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d0d5a144-a9ea-40f2-9e3b-24383d7a9b05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 9068\n",
      "-rw-r--r--. 1 l_schroeder users   38865 Jan 11 14:06 create_test_VCF.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users 6252969 Dec 18 14:04 NMD.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users  249345 Jan 11 14:27 nmd-vep.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users    2465 Dec 18 14:34 README.md\n",
      "-rw-r--r--. 1 l_schroeder users 1307856 Dec 18 14:04 train.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users 1307856 Jan 11 13:57 train_new.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users   54222 Dec 18 14:04 validation_MMRF_TARGET.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users   54222 Jan 11 14:05 validation_new.ipynb\n"
     ]
    }
   ],
   "source": [
    "ls -ll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "361f29f1-977f-4e7c-b1b1-389bc16fcb62",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_path = \"~/NMD/nmd-variant-effect-prediction/resources/TCGA_benchmark/tcga_dataset.vcf\"\n",
    "# vcf_path = \"resources/part-00241-61a0abbf-fbf9-444f-8287-4e46ad4b9b7b-c000.vcf\"\n",
    "gtf_path = \"/s/genomes/Gencode/Gencode_human/release_40/gencode.v40.annotation.gtf.gz\" # new version\n",
    "gtf_path_old = \"/s/genomes/Gencode/Gencode_human/release_33/GRCh37_mapping/gencode.v33lift37.annotation.gtf.gz\" # old version\n",
    "fasta_path = \"/s/genomes/Gencode/Gencode_human/release_40/GRCh38.p13.genome.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "104bdcb9-d1f2-4ae8-97f9-88f56ba49ccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_path = \"~/NMD/nmd-variant-effect-prediction/test_variant/variant.vcf\"\n",
    "gtf_path = \"/s/genomes/Gencode/Gencode_human/release_34/GRCh37_mapping/gencode.v34lift37.annotation.gtf.gz\"\n",
    "fasta_path = \"/s/genomes/Gencode/Gencode_human/release_34/GRCh37_mapping/GRCh37.primary_assembly.genome.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "224c1151-be37-4387-820e-8755482991f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf(vcf_path):\n",
    "    \"\"\"\n",
    "    Read a single VCF file into a PyRanges object with adjusted coordinates.\n",
    "    \"\"\"\n",
    "    df = pd.read_csv(\n",
    "        vcf_path,\n",
    "        comment='#',\n",
    "        sep='\\t',\n",
    "        header=None,\n",
    "        names=['Chromosome', 'Start', 'ID', 'Ref', 'Alt', 'Qual', 'Filter', 'Info']\n",
    "    )\n",
    "\n",
    "    # Adjust coordinates to 0-based, half-open (BED-like)\n",
    "    df['Start'] = df['Start'] - 1\n",
    "    df['End'] = df['Start'] + df['Ref'].str.len()\n",
    "\n",
    "    # Keep only relevant columns\n",
    "    gr = pr.PyRanges(df[['Chromosome', 'Start', 'End', 'ID', 'Ref', 'Alt', 'Qual', 'Filter', 'Info']])\n",
    "    return gr\n",
    "\n",
    "def read_gtf(gtf_path):\n",
    "    \"\"\"\n",
    "    Reads a GTF file into a PyRanges object.\n",
    "    \"\"\"\n",
    "    if not os.path.exists(gtf_path):\n",
    "        raise FileNotFoundError(f\"GTF file not found: {gtf_path}\")\n",
    "    return pr.read_gtf(gtf_path)\n",
    "\n",
    "def read_fasta(fasta_path):\n",
    "    \"\"\"\n",
    "    Reads a genome FASTA file using pyfaidx.Fasta and returns a pyfaidx.Fasta object.\n",
    "    \"\"\"\n",
    "    if not os.path.exists(fasta_path):\n",
    "        raise FileNotFoundError(f\"FASTA file not found: {fasta_path}\")\n",
    "    return Fasta(fasta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "28d9a776-41b2-40e3-b16a-c91e7e40bc5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF shape: (3, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ID</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Qual</th>\n",
       "      <th>Filter</th>\n",
       "      <th>Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-------+\n",
       "| Chromosome   |     Start |       End | ID         | +5    |\n",
       "| (category)   |   (int64) |   (int64) | (object)   | ...   |\n",
       "|--------------+-----------+-----------+------------+-------|\n",
       "| chr8         |  42720558 |  42720559 | .          | ...   |\n",
       "| chr8         |  42706139 |  42706140 | .          | ...   |\n",
       "| chr8         |  42720557 |  42720632 | .          | ...   |\n",
       "+--------------+-----------+-----------+------------+-------+\n",
       "Unstranded PyRanges object has 3 rows and 9 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "5 hidden columns: Ref, Alt, Qual, Filter, Info"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf = read_vcf(vcf_path)\n",
    "print(f\"VCF shape: {vcf.df.shape}\")\n",
    "vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7f981d80-57f4-453b-93f0-fefce3902a16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTF File shape: (2919180, 32)\n"
     ]
    }
   ],
   "source": [
    "gtf = read_gtf(gtf_path)\n",
    "print(f\"GTF File shape: {gtf.df.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c596e044-8708-4c4f-81c8-c9ec5aeab2ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>remap_original_location</th>\n",
       "      <th>remap_status</th>\n",
       "      <th>remap_num_mappings</th>\n",
       "      <th>remap_target_status</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>remap_substituted_missing_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5_3</td>\n",
       "      <td>ENST00000456328.2_1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1_1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.1_3</td>\n",
       "      <td>OTTHUMT00000362751.1_1</td>\n",
       "      <td>chr1:+:11869-12227</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5_3</td>\n",
       "      <td>ENST00000456328.2_1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.1_3</td>\n",
       "      <td>OTTHUMT00000362751.1_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>1</td>\n",
       "      <td>overlap</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5_3</td>\n",
       "      <td>ENST00000450305.2_1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1_1</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.1_3</td>\n",
       "      <td>OTTHUMT00000002844.1_1</td>\n",
       "      <td>chr1:+:12010-12057</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12009</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5_3</td>\n",
       "      <td>ENST00000450305.2_1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.1_3</td>\n",
       "      <td>OTTHUMT00000002844.1_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>1</td>\n",
       "      <td>overlap</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5_3</td>\n",
       "      <td>ENST00000450305.2_1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2_1</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.1_3</td>\n",
       "      <td>OTTHUMT00000002844.1_1</td>\n",
       "      <td>chr1:+:12179-12227</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375420</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>59358334</td>\n",
       "      <td>59360548</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_3_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_2_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMG00000022678.1_3</td>\n",
       "      <td>OTTHUMT00000058841.1_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>1</td>\n",
       "      <td>overlap</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375421</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>59359354</td>\n",
       "      <td>59359508</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_3_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_2_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002036959.1_1</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMG00000022678.1_3</td>\n",
       "      <td>OTTHUMT00000058841.1_2</td>\n",
       "      <td>chrY:-:57213204-57213357</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375422</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>59359676</td>\n",
       "      <td>59359753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_3_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_2_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002021169.1_1</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMG00000022678.1_3</td>\n",
       "      <td>OTTHUMT00000058841.1_2</td>\n",
       "      <td>chrY:-:57213526-57213602</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375423</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>59360030</td>\n",
       "      <td>59360115</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_3_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_2_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00002046926.1_1</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMG00000022678.1_3</td>\n",
       "      <td>OTTHUMT00000058841.1_2</td>\n",
       "      <td>chrY:-:57213880-57213964</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375424</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>59360500</td>\n",
       "      <td>59360548</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_3_PAR_Y</td>\n",
       "      <td>ENST00000507418.6_2_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002072208.1_1</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMG00000022678.1_3</td>\n",
       "      <td>OTTHUMT00000058841.1_2</td>\n",
       "      <td>chrY:-:57214350-57214397</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2375425 rows  32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | +27   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | exon       | 11868     | 12227     | ...   |\n",
       "| chr1         | HAVANA     | transcript | 11868     | 14409     | ...   |\n",
       "| chr1         | HAVANA     | exon       | 12009     | 12057     | ...   |\n",
       "| chr1         | HAVANA     | transcript | 12009     | 13670     | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| chrY         | HAVANA     | exon       | 59359354  | 59359508  | ...   |\n",
       "| chrY         | HAVANA     | exon       | 59359676  | 59359753  | ...   |\n",
       "| chrY         | HAVANA     | exon       | 59360030  | 59360115  | ...   |\n",
       "| chrY         | HAVANA     | exon       | 59360500  | 59360548  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 2,375,425 rows and 32 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "27 hidden columns: Score, Strand, Frame, gene_id, transcript_id, gene_type, gene_name, ... (+ 20 more.)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_test = gtf[gtf.Feature.isin([\"CDS\", \"transcript\", \"exon\"])]\n",
    "gtf_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "17c0a8e0-4d85-488f-841d-b9be1986bd04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fasta = read_fasta(fasta_path)\n",
    "# or\n",
    "fasta = Fasta(fasta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3a4e0519-ffcf-4130-a4c4-515212aa8a9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adjust exon number in GTF (recommended for hg19)\n",
    "\n",
    "def compute_exon_numbers(gtf_df):\n",
    "    \"\"\"\n",
    "    Compute exon numbers for all transcripts in a GTF PyRanges object.\n",
    "    Exon numbers are assigned based on genomic order per transcript and strand.\n",
    "    CDS features inherit the exon number of the exon they overlap.\n",
    "\n",
    "    :param gtf: PyRanges object of the GTF\n",
    "    :return: PyRanges object with new column 'exon_number_computed'\n",
    "    \"\"\"\n",
    "    gtf_df = gtf.df.copy()\n",
    "\n",
    "    # Step 1: Compute exon numbers for exon features\n",
    "    exons = gtf_df[gtf_df.Feature == \"exon\"].copy()\n",
    "    for tx, group in exons.groupby(\"transcript_id\"):\n",
    "        strand = group[\"Strand\"].iloc[0]\n",
    "        if strand == \"+\":\n",
    "            sorted_group = group.sort_values(\"Start\")\n",
    "        else:\n",
    "            sorted_group = group.sort_values(\"Start\", ascending=False)\n",
    "        exons.loc[sorted_group.index, \"exon_number\"] = range(1, len(sorted_group) + 1)\n",
    "\n",
    "    # Step 2: Assign exon numbers to CDS features\n",
    "    cds = gtf_df[gtf_df.Feature == \"CDS\"].copy()\n",
    "    for tx, exon_group in exons.groupby(\"transcript_id\"):\n",
    "        cds_group = cds[cds.transcript_id == tx]\n",
    "        for idx, cds_row in cds_group.iterrows():\n",
    "            overlaps = exon_group[(exon_group[\"Start\"] <= cds_row[\"End\"]) & (exon_group[\"End\"] >= cds_row[\"Start\"])]\n",
    "            if not overlaps.empty:\n",
    "                # choose exon with maximum overlap\n",
    "                overlap_idx = overlaps.apply(lambda row: min(row[\"End\"], cds_row[\"End\"]) - max(row[\"Start\"], cds_row[\"Start\"]), axis=1).idxmax()\n",
    "                gtf_df.loc[idx, \"exon_number\"] = exon_group.loc[overlap_idx, \"exon_number\"]\n",
    "\n",
    "    # Step 3: Update exon features\n",
    "    gtf_df.loc[exons.index, \"exon_number\"] = exons[\"exon_number\"]\n",
    "\n",
    "    return pr.PyRanges(gtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "011aac16-e30c-439f-8390-d3e23fd74e0a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# test compute_exon_numbers() only with a few rows of gtf so its faster\n",
    "\n",
    "gtf_df_subset = gtf.df.iloc[:500,:]\n",
    "gtf_df_subset = pr.PyRanges(gtf_df_subset)\n",
    "\n",
    "gtf_df_subset = compute_exon_numbers(gtf_df_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f25a0efe-28e6-44fa-9a77-53e0fe47e5d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract CDS regions from the GTF file\n",
    "cds = gtf[gtf.Feature == \"CDS\"]\n",
    "cds_df = cds.df\n",
    "\n",
    "# extract exon regions from the GTF file and compute exon related metrics:\n",
    "# exon length & number of exons contained in each transcript\n",
    "exons = gtf[gtf.Feature == \"exon\"]\n",
    "exons_df = exons.df\n",
    "exons_df[\"exon_length\"] = exons_df[\"End\"] - exons_df[\"Start\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a22e3725-f408-4bef-bbe4-03d071b215e5",
   "metadata": {},
   "source": [
    "# Extract PTC\n",
    "\n",
    "Create reference and alternative CDS and transcript sequences (+ metadata) and analyze for start and stop codons & -loss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d0153533-2d6f-41eb-a1c1-1d8338c1ce6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>remap_original_location</th>\n",
       "      <th>remap_status</th>\n",
       "      <th>remap_num_mappings</th>\n",
       "      <th>remap_target_status</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>remap_substituted_missing_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65564</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.6_5</td>\n",
       "      <td>ENST00000641515.2_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-202</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000001094.1_5</td>\n",
       "      <td>OTTHUMT00000003223.1_2</td>\n",
       "      <td>chr1:+:65565-65573</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69036</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.6_5</td>\n",
       "      <td>ENST00000641515.2_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-202</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003813949.1</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000001094.1_5</td>\n",
       "      <td>OTTHUMT00000003223.1_2</td>\n",
       "      <td>chr1:+:69037-70005</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.6_5</td>\n",
       "      <td>ENST00000335137.4_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002319515.2</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000001094.1_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1:+:69091-70005</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000334393.3</td>\n",
       "      <td>CCDS30547.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>367658</td>\n",
       "      <td>368597</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000235249.1</td>\n",
       "      <td>ENST00000426406.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29-001</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002316283.1</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000002860.1</td>\n",
       "      <td>OTTHUMT00000007999.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000409316.1</td>\n",
       "      <td>CCDS41220.1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>V19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>859811</td>\n",
       "      <td>860328</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.12_9</td>\n",
       "      <td>ENST00000420190.6_5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-203</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001637883.2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:28706</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000040719.1_9</td>\n",
       "      <td>OTTHUMT00000316521.1_5</td>\n",
       "      <td>chr1:+:924432-924948</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000411579.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763876</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27187915</td>\n",
       "      <td>27188033</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000185894.8_7</td>\n",
       "      <td>ENST00000618574.1_5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C-204</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001711311.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:18225</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000045199.1_7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:-:25041769-25041886</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000480751.1</td>\n",
       "      <td>CCDS44030.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763877</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27187915</td>\n",
       "      <td>27188033</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000185894.8_7</td>\n",
       "      <td>ENST00000382287.5_6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C-201</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001711311.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:18225</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000045199.1_7</td>\n",
       "      <td>OTTHUMT00000104944.1_6</td>\n",
       "      <td>chrY:-:25041769-25041886</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000371724.1</td>\n",
       "      <td>CCDS44030.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763878</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27190092</td>\n",
       "      <td>27190170</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000185894.8_7</td>\n",
       "      <td>ENST00000618574.1_5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C-204</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003722907.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:18225</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000045199.1_7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY:-:25043946-25044023</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000480751.1</td>\n",
       "      <td>CCDS44030.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763879</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27190092</td>\n",
       "      <td>27190170</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000185894.8_7</td>\n",
       "      <td>ENST00000382287.5_6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C-201</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00003604811.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:18225</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000045199.1_7</td>\n",
       "      <td>OTTHUMT00000104944.1_6</td>\n",
       "      <td>chrY:-:25043946-25044023</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000371724.1</td>\n",
       "      <td>CCDS44030.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763880</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27190134</td>\n",
       "      <td>27190170</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000185894.8_7</td>\n",
       "      <td>ENST00000602680.1_5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C-203</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003266869.1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>HGNC:18225</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000045199.1_7</td>\n",
       "      <td>OTTHUMT00000467695.1_5</td>\n",
       "      <td>chrY:-:25043991-25044023</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000473549.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>763881 rows  32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "0            chr1   HAVANA     CDS     65564     65573     .      +     0   \n",
       "1            chr1   HAVANA     CDS     69036     70008     .      +     0   \n",
       "2            chr1  ENSEMBL     CDS     69090     70008     .      +     0   \n",
       "3            chr1   HAVANA     CDS    367658    368597     .      +     0   \n",
       "4            chr1   HAVANA     CDS    859811    860328     .      +     0   \n",
       "...           ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "763876       chrY  ENSEMBL     CDS  27187915  27188033     .      -     0   \n",
       "763877       chrY   HAVANA     CDS  27187915  27188033     .      -     0   \n",
       "763878       chrY  ENSEMBL     CDS  27190092  27190170     .      -     0   \n",
       "763879       chrY   HAVANA     CDS  27190092  27190170     .      -     0   \n",
       "763880       chrY   HAVANA     CDS  27190134  27190170     .      -     0   \n",
       "\n",
       "                     gene_id        transcript_id       gene_type gene_name  \\\n",
       "0        ENSG00000186092.6_5  ENST00000641515.2_2  protein_coding     OR4F5   \n",
       "1        ENSG00000186092.6_5  ENST00000641515.2_2  protein_coding     OR4F5   \n",
       "2        ENSG00000186092.6_5  ENST00000335137.4_2  protein_coding     OR4F5   \n",
       "3          ENSG00000235249.1    ENST00000426406.1  protein_coding    OR4F29   \n",
       "4       ENSG00000187634.12_9  ENST00000420190.6_5  protein_coding    SAMD11   \n",
       "...                      ...                  ...             ...       ...   \n",
       "763876   ENSG00000185894.8_7  ENST00000618574.1_5  protein_coding     BPY2C   \n",
       "763877   ENSG00000185894.8_7  ENST00000382287.5_6  protein_coding     BPY2C   \n",
       "763878   ENSG00000185894.8_7  ENST00000618574.1_5  protein_coding     BPY2C   \n",
       "763879   ENSG00000185894.8_7  ENST00000382287.5_6  protein_coding     BPY2C   \n",
       "763880   ENSG00000185894.8_7  ENST00000602680.1_5  protein_coding     BPY2C   \n",
       "\n",
       "       transcript_type transcript_name exon_number            exon_id level  \\\n",
       "0       protein_coding       OR4F5-202           2  ENSE00003813641.1     2   \n",
       "1       protein_coding       OR4F5-202           3  ENSE00003813949.1     2   \n",
       "2       protein_coding       OR4F5-201           1  ENSE00002319515.2     3   \n",
       "3       protein_coding      OR4F29-001           1  ENSE00002316283.1     2   \n",
       "4       protein_coding      SAMD11-203           1  ENSE00001637883.2     2   \n",
       "...                ...             ...         ...                ...   ...   \n",
       "763876  protein_coding       BPY2C-204           3  ENSE00001711311.1     3   \n",
       "763877  protein_coding       BPY2C-201           5  ENSE00001711311.1     2   \n",
       "763878  protein_coding       BPY2C-204           4  ENSE00003722907.1     3   \n",
       "763879  protein_coding       BPY2C-201           6  ENSE00003604811.1     2   \n",
       "763880  protein_coding       BPY2C-203           1  ENSE00003266869.1     2   \n",
       "\n",
       "       transcript_support_level     hgnc_id         tag  \\\n",
       "0                           NaN  HGNC:14825       basic   \n",
       "1                           NaN  HGNC:14825       basic   \n",
       "2                            NA  HGNC:14825        CCDS   \n",
       "3                           NaN         NaN        CCDS   \n",
       "4                             3  HGNC:28706  cds_end_NF   \n",
       "...                         ...         ...         ...   \n",
       "763876                        1  HGNC:18225        CCDS   \n",
       "763877                        1  HGNC:18225        CCDS   \n",
       "763878                        1  HGNC:18225        CCDS   \n",
       "763879                        1  HGNC:18225        CCDS   \n",
       "763880                        4  HGNC:18225  cds_end_NF   \n",
       "\n",
       "                   havana_gene       havana_transcript  \\\n",
       "0       OTTHUMG00000001094.1_5  OTTHUMT00000003223.1_2   \n",
       "1       OTTHUMG00000001094.1_5  OTTHUMT00000003223.1_2   \n",
       "2       OTTHUMG00000001094.1_5                     NaN   \n",
       "3         OTTHUMG00000002860.1    OTTHUMT00000007999.1   \n",
       "4       OTTHUMG00000040719.1_9  OTTHUMT00000316521.1_5   \n",
       "...                        ...                     ...   \n",
       "763876  OTTHUMG00000045199.1_7                     NaN   \n",
       "763877  OTTHUMG00000045199.1_7  OTTHUMT00000104944.1_6   \n",
       "763878  OTTHUMG00000045199.1_7                     NaN   \n",
       "763879  OTTHUMG00000045199.1_7  OTTHUMT00000104944.1_6   \n",
       "763880  OTTHUMG00000045199.1_7  OTTHUMT00000467695.1_5   \n",
       "\n",
       "         remap_original_location remap_status remap_num_mappings  \\\n",
       "0             chr1:+:65565-65573  full_contig                NaN   \n",
       "1             chr1:+:69037-70005  full_contig                NaN   \n",
       "2             chr1:+:69091-70005  full_contig                NaN   \n",
       "3                            NaN          NaN                NaN   \n",
       "4           chr1:+:924432-924948  full_contig                NaN   \n",
       "...                          ...          ...                ...   \n",
       "763876  chrY:-:25041769-25041886  full_contig                NaN   \n",
       "763877  chrY:-:25041769-25041886  full_contig                NaN   \n",
       "763878  chrY:-:25043946-25044023  full_contig                NaN   \n",
       "763879  chrY:-:25043946-25044023  full_contig                NaN   \n",
       "763880  chrY:-:25043991-25044023  full_contig                NaN   \n",
       "\n",
       "       remap_target_status  ont         protein_id       ccdsid gene_status  \\\n",
       "0                      NaN  NaN  ENSP00000493376.2          NaN         NaN   \n",
       "1                      NaN  NaN  ENSP00000493376.2          NaN         NaN   \n",
       "2                      NaN  NaN  ENSP00000334393.3  CCDS30547.1         NaN   \n",
       "3                      NaN  NaN  ENSP00000409316.1  CCDS41220.1       KNOWN   \n",
       "4                      NaN  NaN  ENSP00000411579.2          NaN         NaN   \n",
       "...                    ...  ...                ...          ...         ...   \n",
       "763876                 NaN  NaN  ENSP00000480751.1  CCDS44030.1         NaN   \n",
       "763877                 NaN  NaN  ENSP00000371724.1  CCDS44030.1         NaN   \n",
       "763878                 NaN  NaN  ENSP00000480751.1  CCDS44030.1         NaN   \n",
       "763879                 NaN  NaN  ENSP00000371724.1  CCDS44030.1         NaN   \n",
       "763880                 NaN  NaN  ENSP00000473549.1          NaN         NaN   \n",
       "\n",
       "       transcript_status remap_substituted_missing_target  \n",
       "0                    NaN                              NaN  \n",
       "1                    NaN                              NaN  \n",
       "2                    NaN                              NaN  \n",
       "3                  KNOWN                              V19  \n",
       "4                    NaN                              NaN  \n",
       "...                  ...                              ...  \n",
       "763876               NaN                              NaN  \n",
       "763877               NaN                              NaN  \n",
       "763878               NaN                              NaN  \n",
       "763879               NaN                              NaN  \n",
       "763880               NaN                              NaN  \n",
       "\n",
       "[763881 rows x 32 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def adjust_last_cds_for_stop_codon(df, transcript_col=\"transcript_id\"):\n",
    "\n",
    "    \"\"\"\n",
    "    Adjusts the genomic coordinates of the last CDS exon in each transcript by adding 3 positions,\n",
    "    thus to include the stop codon.\n",
    "    Plus strand: extend last exon (largest start position) at the END (+3 to End)\n",
    "    Minus strand: extend last exon (smallest start position) at the START (-3 from Start)\n",
    "    :param df: Dataframe containing CDS annotation\n",
    "    :param exon_col: The name of the column that indicated the exon number, so we can find out which is the last CDS snippet\n",
    "    :param transcript_col: The name of the column that indicates the transcript ID\n",
    "    :return: Modified pandas DataFrame where the last exon of each transcript is extended by 3 bases to include the stop codon.\n",
    "    \"\"\"\n",
    "\n",
    "    df = df.copy()\n",
    "\n",
    "    adjusted_idx = []\n",
    "    for tx, group in df.groupby(transcript_col):\n",
    "        strand = group[\"Strand\"].iloc[0]\n",
    "\n",
    "        if strand == \"+\":\n",
    "            # last exon has max Start position\n",
    "            idx = group[\"Start\"].idxmax()\n",
    "            df.at[idx, \"End\"] += 3\n",
    "        elif strand == \"-\":\n",
    "            # last exon has min Start position\n",
    "            idx = group[\"Start\"].idxmin()\n",
    "            df.at[idx, \"Start\"] -= 3\n",
    "\n",
    "        adjusted_idx.append(idx)\n",
    "\n",
    "    return df\n",
    "\n",
    "cds_df_adj = adjust_last_cds_for_stop_codon(cds_df)\n",
    "cds_df_adj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "66e1b50d-e3e2-4401-b06e-17bd4953e423",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_df_adj[\"exon_number\"] = cds_df_adj[\"exon_number\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8648e817-da50-4643-9bf3-c320926232d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>remap_original_location</th>\n",
       "      <th>remap_status</th>\n",
       "      <th>remap_num_mappings</th>\n",
       "      <th>remap_target_status</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>remap_substituted_missing_target</th>\n",
       "      <th>Start_variant</th>\n",
       "      <th>End_variant</th>\n",
       "      <th>ID</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Qual</th>\n",
       "      <th>Filter</th>\n",
       "      <th>Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001239742.5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42850794-42850997</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-208</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002147066.1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383170.1_2</td>\n",
       "      <td>chr8:-:42865466-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000436416.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "0       chr8   HAVANA     CDS  42705933  42706140     .      -     0   \n",
       "1       chr8  ENSEMBL     CDS  42720558  42720632     .      -     2   \n",
       "2       chr8  ENSEMBL     CDS  42720558  42720632     .      -     2   \n",
       "3       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "4       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "5       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "6       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "7       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "8       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "9       chr8   HAVANA     CDS  42720605  42720632     .      -     2   \n",
       "\n",
       "                gene_id        transcript_id       gene_type gene_name  \\\n",
       "0  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "1  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "2  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "3  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "4  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "5  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "6  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "7  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "8  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "9  ENSG00000120925.16_6  ENST00000531440.5_2  protein_coding    RNF170   \n",
       "\n",
       "  transcript_type transcript_name  exon_number            exon_id level  \\\n",
       "0  protein_coding      RNF170-203            1  ENSE00001239742.5     2   \n",
       "1  protein_coding      RNF170-209            3  ENSE00003678305.1     3   \n",
       "2  protein_coding      RNF170-209            3  ENSE00003678305.1     3   \n",
       "3  protein_coding      RNF170-203            2  ENSE00003678305.1     2   \n",
       "4  protein_coding      RNF170-203            2  ENSE00003678305.1     2   \n",
       "5  protein_coding      RNF170-205            3  ENSE00003678305.1     2   \n",
       "6  protein_coding      RNF170-205            3  ENSE00003678305.1     2   \n",
       "7  protein_coding      RNF170-206            3  ENSE00003678305.1     2   \n",
       "8  protein_coding      RNF170-206            3  ENSE00003678305.1     2   \n",
       "9  protein_coding      RNF170-208            1  ENSE00002147066.1     2   \n",
       "\n",
       "  transcript_support_level     hgnc_id         tag             havana_gene  \\\n",
       "0                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "1                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "2                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "3                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "4                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "5                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "6                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "7                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "8                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "9                        4  HGNC:25358  cds_end_NF  OTTHUMG00000165277.1_6   \n",
       "\n",
       "        havana_transcript   remap_original_location remap_status  \\\n",
       "0  OTTHUMT00000383165.1_2  chr8:-:42850794-42850997  full_contig   \n",
       "1                     NaN  chr8:-:42865416-42865489  full_contig   \n",
       "2                     NaN  chr8:-:42865416-42865489  full_contig   \n",
       "3  OTTHUMT00000383165.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "4  OTTHUMT00000383165.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "5  OTTHUMT00000383168.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "6  OTTHUMT00000383168.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "7  OTTHUMT00000383166.1_3  chr8:-:42865416-42865489  full_contig   \n",
       "8  OTTHUMT00000383166.1_3  chr8:-:42865416-42865489  full_contig   \n",
       "9  OTTHUMT00000383170.1_2  chr8:-:42865466-42865489  full_contig   \n",
       "\n",
       "  remap_num_mappings remap_target_status  ont         protein_id       ccdsid  \\\n",
       "0                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "1                NaN                 NaN  NaN  ENSP00000445725.1   CCDS6138.1   \n",
       "2                NaN                 NaN  NaN  ENSP00000445725.1   CCDS6138.1   \n",
       "3                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "4                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "5                NaN                 NaN  NaN  ENSP00000435782.1  CCDS55229.1   \n",
       "6                NaN                 NaN  NaN  ENSP00000435782.1  CCDS55229.1   \n",
       "7                NaN                 NaN  NaN  ENSP00000434797.1   CCDS6138.1   \n",
       "8                NaN                 NaN  NaN  ENSP00000434797.1   CCDS6138.1   \n",
       "9                NaN                 NaN  NaN  ENSP00000436416.1          NaN   \n",
       "\n",
       "  gene_status transcript_status remap_substituted_missing_target  \\\n",
       "0         NaN               NaN                              NaN   \n",
       "1         NaN               NaN                              NaN   \n",
       "2         NaN               NaN                              NaN   \n",
       "3         NaN               NaN                              NaN   \n",
       "4         NaN               NaN                              NaN   \n",
       "5         NaN               NaN                              NaN   \n",
       "6         NaN               NaN                              NaN   \n",
       "7         NaN               NaN                              NaN   \n",
       "8         NaN               NaN                              NaN   \n",
       "9         NaN               NaN                              NaN   \n",
       "\n",
       "   Start_variant  End_variant ID  \\\n",
       "0       42706139     42706140  .   \n",
       "1       42720557     42720632  .   \n",
       "2       42720558     42720559  .   \n",
       "3       42720557     42720632  .   \n",
       "4       42720558     42720559  .   \n",
       "5       42720557     42720632  .   \n",
       "6       42720558     42720559  .   \n",
       "7       42720557     42720632  .   \n",
       "8       42720558     42720559  .   \n",
       "9       42720557     42720632  .   \n",
       "\n",
       "                                                 Ref Alt Qual Filter Info  \n",
       "0                                                  C   C    .      .    .  \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .  \n",
       "2                                                  C   T    .      .    .  \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .  \n",
       "4                                                  C   T    .      .    .  \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .  \n",
       "6                                                  C   T    .      .    .  \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .  \n",
       "8                                                  C   T    .      .    .  \n",
       "9  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersection_cds_vcf = pr.PyRanges(cds_df_adj).join(vcf, how=None, suffix=\"_variant\").df\n",
    "intersection_cds_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e117e249-d3ad-4836-b907-c97efb161b27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix minus strand variants (only for TCGA / MMREF VCF!)\n",
    "\n",
    "# Fix REF and ALT for minus-strand CDSs\n",
    "mask_minus_strand = intersection_cds_vcf[\"Strand\"] == \"-\"\n",
    "intersection_cds_vcf.loc[mask_minus_strand, \"Ref\"] = intersection_cds_vcf.loc[mask_minus_strand, \"Ref\"].apply(lambda seq: str(Seq(seq).reverse_complement()))\n",
    "intersection_cds_vcf.loc[mask_minus_strand, \"Alt\"] = intersection_cds_vcf.loc[mask_minus_strand, \"Alt\"].apply(lambda seq: str(Seq(seq).reverse_complement()))\n",
    "\n",
    "intersection_cds_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ea2f4b4c-60b4-4510-9ca2-3cb637944aaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>remap_original_location</th>\n",
       "      <th>remap_status</th>\n",
       "      <th>remap_num_mappings</th>\n",
       "      <th>remap_target_status</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>remap_substituted_missing_target</th>\n",
       "      <th>Start_variant</th>\n",
       "      <th>End_variant</th>\n",
       "      <th>ID</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Qual</th>\n",
       "      <th>Filter</th>\n",
       "      <th>Info</th>\n",
       "      <th>Exon_CDS_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001239742.5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42850794-42850997</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-208</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002147066.1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383170.1_2</td>\n",
       "      <td>chr8:-:42865466-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000436416.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ATCGCCAGTAAGCAATAATGCAGGCAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "0       chr8   HAVANA     CDS  42705933  42706140     .      -     0   \n",
       "1       chr8  ENSEMBL     CDS  42720558  42720632     .      -     2   \n",
       "2       chr8  ENSEMBL     CDS  42720558  42720632     .      -     2   \n",
       "3       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "4       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "5       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "6       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "7       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "8       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "9       chr8   HAVANA     CDS  42720605  42720632     .      -     2   \n",
       "\n",
       "                gene_id        transcript_id       gene_type gene_name  \\\n",
       "0  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "1  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "2  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "3  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "4  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "5  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "6  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "7  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "8  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "9  ENSG00000120925.16_6  ENST00000531440.5_2  protein_coding    RNF170   \n",
       "\n",
       "  transcript_type transcript_name  exon_number            exon_id level  \\\n",
       "0  protein_coding      RNF170-203            1  ENSE00001239742.5     2   \n",
       "1  protein_coding      RNF170-209            3  ENSE00003678305.1     3   \n",
       "2  protein_coding      RNF170-209            3  ENSE00003678305.1     3   \n",
       "3  protein_coding      RNF170-203            2  ENSE00003678305.1     2   \n",
       "4  protein_coding      RNF170-203            2  ENSE00003678305.1     2   \n",
       "5  protein_coding      RNF170-205            3  ENSE00003678305.1     2   \n",
       "6  protein_coding      RNF170-205            3  ENSE00003678305.1     2   \n",
       "7  protein_coding      RNF170-206            3  ENSE00003678305.1     2   \n",
       "8  protein_coding      RNF170-206            3  ENSE00003678305.1     2   \n",
       "9  protein_coding      RNF170-208            1  ENSE00002147066.1     2   \n",
       "\n",
       "  transcript_support_level     hgnc_id         tag             havana_gene  \\\n",
       "0                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "1                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "2                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "3                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "4                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "5                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "6                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "7                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "8                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "9                        4  HGNC:25358  cds_end_NF  OTTHUMG00000165277.1_6   \n",
       "\n",
       "        havana_transcript   remap_original_location remap_status  \\\n",
       "0  OTTHUMT00000383165.1_2  chr8:-:42850794-42850997  full_contig   \n",
       "1                     NaN  chr8:-:42865416-42865489  full_contig   \n",
       "2                     NaN  chr8:-:42865416-42865489  full_contig   \n",
       "3  OTTHUMT00000383165.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "4  OTTHUMT00000383165.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "5  OTTHUMT00000383168.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "6  OTTHUMT00000383168.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "7  OTTHUMT00000383166.1_3  chr8:-:42865416-42865489  full_contig   \n",
       "8  OTTHUMT00000383166.1_3  chr8:-:42865416-42865489  full_contig   \n",
       "9  OTTHUMT00000383170.1_2  chr8:-:42865466-42865489  full_contig   \n",
       "\n",
       "  remap_num_mappings remap_target_status  ont         protein_id       ccdsid  \\\n",
       "0                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "1                NaN                 NaN  NaN  ENSP00000445725.1   CCDS6138.1   \n",
       "2                NaN                 NaN  NaN  ENSP00000445725.1   CCDS6138.1   \n",
       "3                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "4                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "5                NaN                 NaN  NaN  ENSP00000435782.1  CCDS55229.1   \n",
       "6                NaN                 NaN  NaN  ENSP00000435782.1  CCDS55229.1   \n",
       "7                NaN                 NaN  NaN  ENSP00000434797.1   CCDS6138.1   \n",
       "8                NaN                 NaN  NaN  ENSP00000434797.1   CCDS6138.1   \n",
       "9                NaN                 NaN  NaN  ENSP00000436416.1          NaN   \n",
       "\n",
       "  gene_status transcript_status remap_substituted_missing_target  \\\n",
       "0         NaN               NaN                              NaN   \n",
       "1         NaN               NaN                              NaN   \n",
       "2         NaN               NaN                              NaN   \n",
       "3         NaN               NaN                              NaN   \n",
       "4         NaN               NaN                              NaN   \n",
       "5         NaN               NaN                              NaN   \n",
       "6         NaN               NaN                              NaN   \n",
       "7         NaN               NaN                              NaN   \n",
       "8         NaN               NaN                              NaN   \n",
       "9         NaN               NaN                              NaN   \n",
       "\n",
       "   Start_variant  End_variant ID  \\\n",
       "0       42706139     42706140  .   \n",
       "1       42720557     42720632  .   \n",
       "2       42720558     42720559  .   \n",
       "3       42720557     42720632  .   \n",
       "4       42720558     42720559  .   \n",
       "5       42720557     42720632  .   \n",
       "6       42720558     42720559  .   \n",
       "7       42720557     42720632  .   \n",
       "8       42720558     42720559  .   \n",
       "9       42720557     42720632  .   \n",
       "\n",
       "                                                 Ref Alt Qual Filter Info  \\\n",
       "0                                                  C   C    .      .    .   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "2                                                  C   T    .      .    .   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "4                                                  C   T    .      .    .   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "6                                                  C   T    .      .    .   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "8                                                  C   T    .      .    .   \n",
       "9  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "\n",
       "                                        Exon_CDS_seq  \n",
       "0  TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...  \n",
       "1  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "2  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "3  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "4  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "5  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "6  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "7  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "8  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "9                        ATCGCCAGTAAGCAATAATGCAGGCAC  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import Dependencies\n",
    "from functools import lru_cache\n",
    "\n",
    "# Global cache to map id --> actual fasta object\n",
    "_fasta_cache = {}\n",
    "\n",
    "@lru_cache(maxsize=None)\n",
    "def fetch_seq_cached(chrom, start, end, fasta_id):\n",
    "    fasta = _fasta_cache[fasta_id]  # lookup the actual fasta object\n",
    "    return fasta[chrom][start:end].seq.upper()\n",
    "\n",
    "def add_exon_cds_sequence(df, fasta, chrom_col=\"Chromosome\", start_col=\"Start\", end_col=\"End\", new_col=\"Exon_CDS_seq\"):\n",
    "    fasta_id = id(fasta)\n",
    "    _fasta_cache[fasta_id] = fasta  # store for later access\n",
    "\n",
    "    def extract_seq(row):\n",
    "        return fetch_seq_cached(row[chrom_col], row[start_col], row[end_col], fasta_id)\n",
    "\n",
    "    df[new_col] = df.apply(extract_seq, axis=1)\n",
    "    return df\n",
    "\n",
    "intersection_cds_vcf = add_exon_cds_sequence(intersection_cds_vcf, fasta)\n",
    "intersection_cds_vcf"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ea7cd935-3f39-4053-a1b6-68c2519fcc7b",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# old function\n",
    "\n",
    "def apply_variant_edge_aware_with_lengths(row):\n",
    "\n",
    "    \"\"\"\n",
    "    Applies a variant to a CDS exon sequence, taking into account not only SNVs but also partial overlaps at exon\n",
    "    boundaries and computing the alternative sequence.\n",
    "\n",
    "    :param row: A single row from the DataFrame (A pandas.Series) containing among others CDS Start and End, Variant\n",
    "                Start and End, Ref, Alt, Exon_CDS_seq (Original CDS sequence as string)\n",
    "    :return: The input pandas.Series with additional information:\n",
    "             Exon_CDS_length (length of the original CDS),\n",
    "             Exon_Alt_CDS_seq (alternative CDS after applying the variant / None if invalid),\n",
    "             Exon_Alt_CDS_length (length of the alternative CDS / None if invalid).\n",
    "    \"\"\"\n",
    "\n",
    "    cds_seq = list(row[\"Exon_CDS_seq\"])\n",
    "    strand = row[\"Strand\"]\n",
    "    ref = row[\"Ref\"]\n",
    "    alt = row[\"Alt\"]\n",
    "    cds_start = int(row[\"Start\"])\n",
    "    cds_end = int(row[\"End\"])\n",
    "    var_start = int(row[\"Start_variant\"])\n",
    "    var_end = int(row[\"End_variant\"])\n",
    "\n",
    "    # Determine the overlap between variant and this CDS region\n",
    "    overlap_start = max(var_start, cds_start)\n",
    "    overlap_end = min(var_end, cds_end)\n",
    "\n",
    "    # If there is no overlap between the variant and this CDS region\n",
    "    if overlap_start >= overlap_end:\n",
    "        return pd.Series({\n",
    "            \"Exon_CDS_length\": len(cds_seq),\n",
    "            \"Exon_Alt_CDS_seq\": None,\n",
    "            \"Exon_Alt_CDS_length\": None\n",
    "        })\n",
    "\n",
    "    # Position of overlap within the CDS\n",
    "    cds_index = overlap_start - cds_start\n",
    "    overlap_len = overlap_end - overlap_start\n",
    "\n",
    "    # Offset of the overlapping region within the variant\n",
    "    ref_offset = overlap_start - var_start\n",
    "    ref_in_cds = ref[ref_offset:ref_offset + overlap_len]\n",
    "    alt_in_cds = alt[ref_offset:ref_offset + overlap_len]\n",
    "\n",
    "    # Determine if theres leftover alt outside CDS (insertions at end)\n",
    "    extra_alt = \"\"\n",
    "    if len(alt) > len(ref):\n",
    "        # Limit extra_alt to what corresponds to CDS overlap\n",
    "        extra_start = ref_offset + overlap_len\n",
    "        if var_end > cds_end:\n",
    "            # Only include alt bases that map to CDS\n",
    "            remaining_cds_len = cds_end - overlap_end\n",
    "            extra_alt = alt[extra_start:extra_start + remaining_cds_len]\n",
    "        else:\n",
    "            extra_alt = alt[extra_start:]\n",
    "\n",
    "    # Confirm that the reference matches\n",
    "    cds_ref_part = \"\".join(cds_seq[cds_index:cds_index + overlap_len])\n",
    "    if cds_ref_part != ref_in_cds.upper(): # reference mismatch\n",
    "        return pd.Series({\n",
    "            \"Exon_CDS_length\": len(cds_seq),\n",
    "            \"Exon_Alt_CDS_seq\": None,\n",
    "            \"Exon_Alt_CDS_length\": None\n",
    "        })\n",
    "\n",
    "    # Build alternative sequence\n",
    "    alt_seq = []\n",
    "    alt_seq.extend(cds_seq[:cds_index]) # Copy the CDS up to the variant position\n",
    "\n",
    "    if len(ref) == len(alt): # Substitution\n",
    "        alt_seq.extend(list(alt_in_cds))\n",
    "    elif len(alt) > len(ref): # Insertion\n",
    "        alt_seq.extend(list(alt_in_cds))\n",
    "        alt_seq.extend(list(extra_alt))\n",
    "    elif len(ref) > len(alt): # Deletion\n",
    "        alt_seq.extend(list(alt_in_cds))\n",
    "\n",
    "    # Add remaining CDS sequence after variant\n",
    "    alt_seq.extend(cds_seq[cds_index + overlap_len:])\n",
    "\n",
    "    return pd.Series({\n",
    "        \"Exon_CDS_length\": len(cds_seq),\n",
    "        \"Exon_Alt_CDS_seq\": \"\".join(alt_seq),\n",
    "        \"Exon_Alt_CDS_length\": len(alt_seq)\n",
    "    })\n",
    "\n",
    "\n",
    "intersection_cds_vcf[[\"Exon_CDS_length\", \"Exon_Alt_CDS_seq\", \"Exon_Alt_CDS_length\"]] = intersection_cds_vcf.apply(\n",
    "        apply_variant_edge_aware_with_lengths,\n",
    "        axis=1\n",
    "    )\n",
    "\n",
    "intersection_cds_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "143c31e1-20eb-42e4-8e3b-3579296ed082",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>remap_original_location</th>\n",
       "      <th>remap_status</th>\n",
       "      <th>remap_num_mappings</th>\n",
       "      <th>remap_target_status</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>remap_substituted_missing_target</th>\n",
       "      <th>Start_variant</th>\n",
       "      <th>End_variant</th>\n",
       "      <th>ID</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Qual</th>\n",
       "      <th>Filter</th>\n",
       "      <th>Info</th>\n",
       "      <th>Exon_CDS_seq</th>\n",
       "      <th>Exon_CDS_length</th>\n",
       "      <th>Exon_Alt_CDS_seq</th>\n",
       "      <th>Exon_Alt_CDS_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001239742.5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42850794-42850997</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...</td>\n",
       "      <td>207</td>\n",
       "      <td>TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...</td>\n",
       "      <td>207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td>TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td>TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td>TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td>TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-208</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002147066.1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383170.1_2</td>\n",
       "      <td>chr8:-:42865466-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000436416.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ATCGCCAGTAAGCAATAATGCAGGCAC</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "0       chr8   HAVANA     CDS  42705933  42706140     .      -     0   \n",
       "1       chr8  ENSEMBL     CDS  42720558  42720632     .      -     2   \n",
       "2       chr8  ENSEMBL     CDS  42720558  42720632     .      -     2   \n",
       "3       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "4       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "5       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "6       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "7       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "8       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "9       chr8   HAVANA     CDS  42720605  42720632     .      -     2   \n",
       "\n",
       "                gene_id        transcript_id       gene_type gene_name  \\\n",
       "0  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "1  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "2  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "3  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "4  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "5  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "6  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "7  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "8  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "9  ENSG00000120925.16_6  ENST00000531440.5_2  protein_coding    RNF170   \n",
       "\n",
       "  transcript_type transcript_name  exon_number            exon_id level  \\\n",
       "0  protein_coding      RNF170-203            1  ENSE00001239742.5     2   \n",
       "1  protein_coding      RNF170-209            3  ENSE00003678305.1     3   \n",
       "2  protein_coding      RNF170-209            3  ENSE00003678305.1     3   \n",
       "3  protein_coding      RNF170-203            2  ENSE00003678305.1     2   \n",
       "4  protein_coding      RNF170-203            2  ENSE00003678305.1     2   \n",
       "5  protein_coding      RNF170-205            3  ENSE00003678305.1     2   \n",
       "6  protein_coding      RNF170-205            3  ENSE00003678305.1     2   \n",
       "7  protein_coding      RNF170-206            3  ENSE00003678305.1     2   \n",
       "8  protein_coding      RNF170-206            3  ENSE00003678305.1     2   \n",
       "9  protein_coding      RNF170-208            1  ENSE00002147066.1     2   \n",
       "\n",
       "  transcript_support_level     hgnc_id         tag             havana_gene  \\\n",
       "0                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "1                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "2                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "3                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "4                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "5                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "6                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "7                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "8                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "9                        4  HGNC:25358  cds_end_NF  OTTHUMG00000165277.1_6   \n",
       "\n",
       "        havana_transcript   remap_original_location remap_status  \\\n",
       "0  OTTHUMT00000383165.1_2  chr8:-:42850794-42850997  full_contig   \n",
       "1                     NaN  chr8:-:42865416-42865489  full_contig   \n",
       "2                     NaN  chr8:-:42865416-42865489  full_contig   \n",
       "3  OTTHUMT00000383165.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "4  OTTHUMT00000383165.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "5  OTTHUMT00000383168.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "6  OTTHUMT00000383168.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "7  OTTHUMT00000383166.1_3  chr8:-:42865416-42865489  full_contig   \n",
       "8  OTTHUMT00000383166.1_3  chr8:-:42865416-42865489  full_contig   \n",
       "9  OTTHUMT00000383170.1_2  chr8:-:42865466-42865489  full_contig   \n",
       "\n",
       "  remap_num_mappings remap_target_status  ont         protein_id       ccdsid  \\\n",
       "0                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "1                NaN                 NaN  NaN  ENSP00000445725.1   CCDS6138.1   \n",
       "2                NaN                 NaN  NaN  ENSP00000445725.1   CCDS6138.1   \n",
       "3                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "4                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "5                NaN                 NaN  NaN  ENSP00000435782.1  CCDS55229.1   \n",
       "6                NaN                 NaN  NaN  ENSP00000435782.1  CCDS55229.1   \n",
       "7                NaN                 NaN  NaN  ENSP00000434797.1   CCDS6138.1   \n",
       "8                NaN                 NaN  NaN  ENSP00000434797.1   CCDS6138.1   \n",
       "9                NaN                 NaN  NaN  ENSP00000436416.1          NaN   \n",
       "\n",
       "  gene_status transcript_status remap_substituted_missing_target  \\\n",
       "0         NaN               NaN                              NaN   \n",
       "1         NaN               NaN                              NaN   \n",
       "2         NaN               NaN                              NaN   \n",
       "3         NaN               NaN                              NaN   \n",
       "4         NaN               NaN                              NaN   \n",
       "5         NaN               NaN                              NaN   \n",
       "6         NaN               NaN                              NaN   \n",
       "7         NaN               NaN                              NaN   \n",
       "8         NaN               NaN                              NaN   \n",
       "9         NaN               NaN                              NaN   \n",
       "\n",
       "   Start_variant  End_variant ID  \\\n",
       "0       42706139     42706140  .   \n",
       "1       42720557     42720632  .   \n",
       "2       42720558     42720559  .   \n",
       "3       42720557     42720632  .   \n",
       "4       42720558     42720559  .   \n",
       "5       42720557     42720632  .   \n",
       "6       42720558     42720559  .   \n",
       "7       42720557     42720632  .   \n",
       "8       42720558     42720559  .   \n",
       "9       42720557     42720632  .   \n",
       "\n",
       "                                                 Ref Alt Qual Filter Info  \\\n",
       "0                                                  C   C    .      .    .   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "2                                                  C   T    .      .    .   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "4                                                  C   T    .      .    .   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "6                                                  C   T    .      .    .   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "8                                                  C   T    .      .    .   \n",
       "9  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "\n",
       "                                        Exon_CDS_seq  Exon_CDS_length  \\\n",
       "0  TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...              207   \n",
       "1  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "2  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "3  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "4  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "5  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "6  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "7  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "8  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "9                        ATCGCCAGTAAGCAATAATGCAGGCAC               27   \n",
       "\n",
       "                                    Exon_Alt_CDS_seq  Exon_Alt_CDS_length  \n",
       "0  TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...                  207  \n",
       "1                                                                       0  \n",
       "2  TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...                   74  \n",
       "3                                                                       0  \n",
       "4  TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...                   74  \n",
       "5                                                                       0  \n",
       "6  TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...                   74  \n",
       "7                                                                       0  \n",
       "8  TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...                   74  \n",
       "9                                                                       0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# new function to apply variants with <DEL> and <DUP>\n",
    "\n",
    "def apply_variant_edge_aware_with_lengths(row):\n",
    "\n",
    "    \"\"\"\n",
    "    Applies a variant to a CDS exon sequence, taking into account not only SNVs but also partial overlaps at exon\n",
    "    boundaries and computing the alternative sequence.\n",
    "\n",
    "    :param row: A single row from the DataFrame (A pandas.Series) containing among others CDS Start and End, Variant\n",
    "                Start and End, Ref, Alt, Exon_CDS_seq (Original CDS sequence as string)\n",
    "    :return: The input pandas.Series with additional information:\n",
    "             Exon_CDS_length (length of the original CDS),\n",
    "             Exon_Alt_CDS_seq (alternative CDS after applying the variant / None if invalid),\n",
    "             Exon_Alt_CDS_length (length of the alternative CDS / None if invalid).\n",
    "    \"\"\"\n",
    "\n",
    "    cds_seq = list(row[\"Exon_CDS_seq\"])\n",
    "    strand = row[\"Strand\"]\n",
    "    ref = row[\"Ref\"]\n",
    "    alt = row[\"Alt\"]\n",
    "    cds_start = int(row[\"Start\"])\n",
    "    cds_end = int(row[\"End\"])\n",
    "    var_start = int(row[\"Start_variant\"])\n",
    "    var_end = int(row[\"End_variant\"])\n",
    "\n",
    "    # Special handling for deletions (Ref = N and Alt = <DEL>)\n",
    "    if ref == \"N\" and alt == \"<DEL>\":\n",
    "        # Clip deletion to the CDS region (only remove overlap part)\n",
    "\n",
    "        # Determine the overlap between variant and this CDS region\n",
    "        overlap_start = max(var_start, cds_start)\n",
    "        overlap_end = min(var_end, cds_end)\n",
    "\n",
    "        # If there is no overlap between the variant and this CDS region\n",
    "        if overlap_start >= overlap_end:\n",
    "            return pd.Series({\n",
    "                \"Exon_CDS_length\": len(cds_seq),\n",
    "                \"Exon_Alt_CDS_seq\": None,\n",
    "                \"Exon_Alt_CDS_length\": None\n",
    "            })\n",
    "\n",
    "        cds_index_start = overlap_start - cds_start\n",
    "        cds_index_end = overlap_end - cds_start\n",
    "\n",
    "        alt_seq = []\n",
    "        alt_seq.extend(cds_seq[:cds_index_start])  # keep sequence before deletion\n",
    "        alt_seq.extend(cds_seq[cds_index_end:])  # keep sequence after deletion\n",
    "\n",
    "        return pd.Series({\n",
    "            \"Exon_CDS_length\": len(cds_seq),\n",
    "            \"Exon_Alt_CDS_seq\": \"\".join(alt_seq),\n",
    "            \"Exon_Alt_CDS_length\": len(alt_seq)\n",
    "        })\n",
    "\n",
    "    # Special handling for duplications (Ref = N and Alt = <DUP>)\n",
    "    if ref == \"N\" and alt == \"<DUP>\":\n",
    "\n",
    "        # Determine the overlap between variant and this CDS region\n",
    "        overlap_start = max(var_start, cds_start)\n",
    "        overlap_end = min(var_end, cds_end)\n",
    "\n",
    "        # If there is no overlap between the variant and this CDS region\n",
    "        if overlap_start >= overlap_end:\n",
    "            return pd.Series({\n",
    "                \"Exon_CDS_length\": len(cds_seq),\n",
    "                \"Exon_Alt_CDS_seq\": None,\n",
    "                \"Exon_Alt_CDS_length\": None\n",
    "            })\n",
    "\n",
    "        cds_index_start = overlap_start - cds_start\n",
    "        cds_index_end = overlap_end - cds_start\n",
    "\n",
    "        alt_seq = []\n",
    "        alt_seq.extend(cds_seq[:cds_index_start])  # sequence up to the end of the overlap\n",
    "        alt_seq.extend(cds_seq[cds_index_start:cds_index_end]) # overlap-region (original, in CDS)\n",
    "        alt_seq.extend(cds_seq[cds_index_start:cds_index_end])  # duplicate the overlapped region\n",
    "        alt_seq.extend(cds_seq[cds_index_end:])  # keep sequence after duplication\n",
    "\n",
    "        return pd.Series({\n",
    "            \"Exon_CDS_length\": len(cds_seq),\n",
    "            \"Exon_Alt_CDS_seq\": \"\".join(alt_seq),\n",
    "            \"Exon_Alt_CDS_length\": len(alt_seq)\n",
    "        })\n",
    "\n",
    "    # Determine the overlap between variant and this CDS region\n",
    "    overlap_start = max(var_start, cds_start)\n",
    "    overlap_end = min(var_end, cds_end)\n",
    "\n",
    "    # If there is no overlap between the variant and this CDS region\n",
    "    if overlap_start >= overlap_end:\n",
    "        return pd.Series({\n",
    "            \"Exon_CDS_length\": len(cds_seq),\n",
    "            \"Exon_Alt_CDS_seq\": None,\n",
    "            \"Exon_Alt_CDS_length\": None\n",
    "        })\n",
    "\n",
    "    # Position of overlap within the CDS\n",
    "    cds_index = overlap_start - cds_start\n",
    "    overlap_len = overlap_end - overlap_start\n",
    "\n",
    "    # Offset of the overlapping region within the variant\n",
    "    ref_offset = overlap_start - var_start\n",
    "    ref_in_cds = ref[ref_offset:ref_offset + overlap_len]\n",
    "    alt_in_cds = alt[ref_offset:ref_offset + overlap_len]\n",
    "\n",
    "    # Determine if theres leftover alt outside CDS (insertions at end)\n",
    "    extra_alt = \"\"\n",
    "    if len(alt) > len(ref):\n",
    "        # Limit extra_alt to what corresponds to CDS overlap\n",
    "        extra_start = ref_offset + overlap_len\n",
    "        if var_end > cds_end:\n",
    "            # Only include alt bases that map to CDS\n",
    "            remaining_cds_len = cds_end - overlap_end\n",
    "            extra_alt = alt[extra_start:extra_start + remaining_cds_len]\n",
    "        else:\n",
    "            extra_alt = alt[extra_start:]\n",
    "\n",
    "    # Confirm that the reference matches\n",
    "    cds_ref_part = \"\".join(cds_seq[cds_index:cds_index + overlap_len])\n",
    "    if cds_ref_part != ref_in_cds.upper(): # reference mismatch\n",
    "        return pd.Series({\n",
    "            \"Exon_CDS_length\": len(cds_seq),\n",
    "            \"Exon_Alt_CDS_seq\": None,\n",
    "            \"Exon_Alt_CDS_length\": None\n",
    "        })\n",
    "\n",
    "    # Build alternative sequence\n",
    "    alt_seq = []\n",
    "    alt_seq.extend(cds_seq[:cds_index]) # Copy the CDS up to the variant position\n",
    "\n",
    "    if len(ref) == len(alt): # Substitution\n",
    "        alt_seq.extend(list(alt_in_cds))\n",
    "    elif len(alt) > len(ref): # Insertion\n",
    "        alt_seq.extend(list(alt_in_cds))\n",
    "        alt_seq.extend(list(extra_alt))\n",
    "    elif len(ref) > len(alt): # Deletion\n",
    "        alt_seq.extend(list(alt_in_cds))\n",
    "\n",
    "    # Add remaining CDS sequence after variant\n",
    "    alt_seq.extend(cds_seq[cds_index + overlap_len:])\n",
    "\n",
    "    return pd.Series({\n",
    "        \"Exon_CDS_length\": len(cds_seq),\n",
    "        \"Exon_Alt_CDS_seq\": \"\".join(alt_seq),\n",
    "        \"Exon_Alt_CDS_length\": len(alt_seq)\n",
    "    })\n",
    "\n",
    "intersection_cds_vcf[[\"Exon_CDS_length\", \"Exon_Alt_CDS_seq\", \"Exon_Alt_CDS_length\"]] = intersection_cds_vcf.apply(\n",
    "        apply_variant_edge_aware_with_lengths,\n",
    "        axis=1\n",
    "    )\n",
    "\n",
    "intersection_cds_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5facaca3-58d6-42cc-a6b6-7545a4331495",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[Warning] Skipping 0 variants due to reference mismatches:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>remap_original_location</th>\n",
       "      <th>remap_status</th>\n",
       "      <th>remap_num_mappings</th>\n",
       "      <th>remap_target_status</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>remap_substituted_missing_target</th>\n",
       "      <th>Start_variant</th>\n",
       "      <th>End_variant</th>\n",
       "      <th>ID</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Qual</th>\n",
       "      <th>Filter</th>\n",
       "      <th>Info</th>\n",
       "      <th>Exon_CDS_seq</th>\n",
       "      <th>Exon_CDS_length</th>\n",
       "      <th>Exon_Alt_CDS_seq</th>\n",
       "      <th>Exon_Alt_CDS_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001239742.5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42850794-42850997</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...</td>\n",
       "      <td>207</td>\n",
       "      <td>TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...</td>\n",
       "      <td>207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td>TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td>TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td>TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "      <td>TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-208</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002147066.1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383170.1_2</td>\n",
       "      <td>chr8:-:42865466-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000436416.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ATCGCCAGTAAGCAATAATGCAGGCAC</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "0       chr8   HAVANA     CDS  42705933  42706140     .      -     0   \n",
       "1       chr8  ENSEMBL     CDS  42720558  42720632     .      -     2   \n",
       "2       chr8  ENSEMBL     CDS  42720558  42720632     .      -     2   \n",
       "3       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "4       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "5       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "6       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "7       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "8       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "9       chr8   HAVANA     CDS  42720605  42720632     .      -     2   \n",
       "\n",
       "                gene_id        transcript_id       gene_type gene_name  \\\n",
       "0  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "1  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "2  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "3  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "4  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "5  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "6  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "7  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "8  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "9  ENSG00000120925.16_6  ENST00000531440.5_2  protein_coding    RNF170   \n",
       "\n",
       "  transcript_type transcript_name  exon_number            exon_id level  \\\n",
       "0  protein_coding      RNF170-203            1  ENSE00001239742.5     2   \n",
       "1  protein_coding      RNF170-209            3  ENSE00003678305.1     3   \n",
       "2  protein_coding      RNF170-209            3  ENSE00003678305.1     3   \n",
       "3  protein_coding      RNF170-203            2  ENSE00003678305.1     2   \n",
       "4  protein_coding      RNF170-203            2  ENSE00003678305.1     2   \n",
       "5  protein_coding      RNF170-205            3  ENSE00003678305.1     2   \n",
       "6  protein_coding      RNF170-205            3  ENSE00003678305.1     2   \n",
       "7  protein_coding      RNF170-206            3  ENSE00003678305.1     2   \n",
       "8  protein_coding      RNF170-206            3  ENSE00003678305.1     2   \n",
       "9  protein_coding      RNF170-208            1  ENSE00002147066.1     2   \n",
       "\n",
       "  transcript_support_level     hgnc_id         tag             havana_gene  \\\n",
       "0                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "1                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "2                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "3                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "4                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "5                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "6                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "7                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "8                        1  HGNC:25358        CCDS  OTTHUMG00000165277.1_6   \n",
       "9                        4  HGNC:25358  cds_end_NF  OTTHUMG00000165277.1_6   \n",
       "\n",
       "        havana_transcript   remap_original_location remap_status  \\\n",
       "0  OTTHUMT00000383165.1_2  chr8:-:42850794-42850997  full_contig   \n",
       "1                     NaN  chr8:-:42865416-42865489  full_contig   \n",
       "2                     NaN  chr8:-:42865416-42865489  full_contig   \n",
       "3  OTTHUMT00000383165.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "4  OTTHUMT00000383165.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "5  OTTHUMT00000383168.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "6  OTTHUMT00000383168.1_2  chr8:-:42865416-42865489  full_contig   \n",
       "7  OTTHUMT00000383166.1_3  chr8:-:42865416-42865489  full_contig   \n",
       "8  OTTHUMT00000383166.1_3  chr8:-:42865416-42865489  full_contig   \n",
       "9  OTTHUMT00000383170.1_2  chr8:-:42865466-42865489  full_contig   \n",
       "\n",
       "  remap_num_mappings remap_target_status  ont         protein_id       ccdsid  \\\n",
       "0                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "1                NaN                 NaN  NaN  ENSP00000445725.1   CCDS6138.1   \n",
       "2                NaN                 NaN  NaN  ENSP00000445725.1   CCDS6138.1   \n",
       "3                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "4                NaN                 NaN  NaN  ENSP00000326138.3  CCDS55230.1   \n",
       "5                NaN                 NaN  NaN  ENSP00000435782.1  CCDS55229.1   \n",
       "6                NaN                 NaN  NaN  ENSP00000435782.1  CCDS55229.1   \n",
       "7                NaN                 NaN  NaN  ENSP00000434797.1   CCDS6138.1   \n",
       "8                NaN                 NaN  NaN  ENSP00000434797.1   CCDS6138.1   \n",
       "9                NaN                 NaN  NaN  ENSP00000436416.1          NaN   \n",
       "\n",
       "  gene_status transcript_status remap_substituted_missing_target  \\\n",
       "0         NaN               NaN                              NaN   \n",
       "1         NaN               NaN                              NaN   \n",
       "2         NaN               NaN                              NaN   \n",
       "3         NaN               NaN                              NaN   \n",
       "4         NaN               NaN                              NaN   \n",
       "5         NaN               NaN                              NaN   \n",
       "6         NaN               NaN                              NaN   \n",
       "7         NaN               NaN                              NaN   \n",
       "8         NaN               NaN                              NaN   \n",
       "9         NaN               NaN                              NaN   \n",
       "\n",
       "   Start_variant  End_variant ID  \\\n",
       "0       42706139     42706140  .   \n",
       "1       42720557     42720632  .   \n",
       "2       42720558     42720559  .   \n",
       "3       42720557     42720632  .   \n",
       "4       42720558     42720559  .   \n",
       "5       42720557     42720632  .   \n",
       "6       42720558     42720559  .   \n",
       "7       42720557     42720632  .   \n",
       "8       42720558     42720559  .   \n",
       "9       42720557     42720632  .   \n",
       "\n",
       "                                                 Ref Alt Qual Filter Info  \\\n",
       "0                                                  C   C    .      .    .   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "2                                                  C   T    .      .    .   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "4                                                  C   T    .      .    .   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "6                                                  C   T    .      .    .   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "8                                                  C   T    .      .    .   \n",
       "9  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C    .      .    .   \n",
       "\n",
       "                                        Exon_CDS_seq  Exon_CDS_length  \\\n",
       "0  TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...              207   \n",
       "1  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "2  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "3  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "4  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "5  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "6  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "7  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "8  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...               74   \n",
       "9                        ATCGCCAGTAAGCAATAATGCAGGCAC               27   \n",
       "\n",
       "                                    Exon_Alt_CDS_seq  Exon_Alt_CDS_length  \n",
       "0  TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...                  207  \n",
       "1                                                                       0  \n",
       "2  TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...                   74  \n",
       "3                                                                       0  \n",
       "4  TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...                   74  \n",
       "5                                                                       0  \n",
       "6  TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...                   74  \n",
       "7                                                                       0  \n",
       "8  TGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...                   74  \n",
       "9                                                                       0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter out Variants with a reference mismatch\n",
    "mismatched_rows = intersection_cds_vcf[intersection_cds_vcf[\"Exon_Alt_CDS_seq\"].isna()]\n",
    "print(f\"\\n[Warning] Skipping {len(mismatched_rows)} variants due to reference mismatches\")\n",
    "if not mismatched_rows.empty:\n",
    "    print(mismatched_rows[[\"transcript_id\", \"Chromosome\", \"Start_variant\", \"End_variant\", \"Ref\", \"Alt\"]].to_string(index=False))\n",
    "    \n",
    "intersection_cds_vcf = intersection_cds_vcf[intersection_cds_vcf[\"Exon_Alt_CDS_seq\"].notna()].copy()\n",
    "intersection_cds_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9f86a7e2-73b4-48ee-92a9-778ac322a942",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Limit to relevant transcript (to save time)\n",
    "relevant_transcripts = intersection_cds_vcf[\"transcript_id\"].unique()\n",
    "cds_df_adj = cds_df_adj[cds_df_adj[\"transcript_id\"].isin(relevant_transcripts)].copy()"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b6bad810-36a4-46d2-af77-3de37387e24c",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "relevant_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "284c4961-4504-4e68-81da-3f065b21cb24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>remap_original_location</th>\n",
       "      <th>remap_status</th>\n",
       "      <th>remap_num_mappings</th>\n",
       "      <th>remap_target_status</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>remap_substituted_missing_target</th>\n",
       "      <th>Exon_CDS_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>330801</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42706140</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001239742.5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42850794-42850997</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330802</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42711571</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001239728.3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42856162-42856428</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCATCTAGTTAGCCTTTGGGTTATCACTTCTCGATACATAATAGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330803</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42711571</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002152697.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42856162-42856428</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCATCTAGTTAGCCTTTGGGTTATCACTTCTCGATACATAATAGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330804</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42711571</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002184629.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42856162-42856428</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCATCTAGTTAGCCTTTGGGTTATCACTTCTCGATACATAATAGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330805</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42716887</td>\n",
       "      <td>42716998</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003614933.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42861745-42861855</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGATCTGGGTTGCCCTGAGAATCTCCGGTTATAATCATTAATATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330806</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42716887</td>\n",
       "      <td>42716998</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003614933.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42861745-42861855</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGATCTGGGTTGCCCTGAGAATCTCCGGTTATAATCATTAATATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330807</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42716887</td>\n",
       "      <td>42716998</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003614933.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42861745-42861855</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGATCTGGGTTGCCCTGAGAATCTCCGGTTATAATCATTAATATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330810</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330811</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330812</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330813</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003678305.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42865416-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330814</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42720632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-208</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002147066.1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383170.1_2</td>\n",
       "      <td>chr8:-:42865466-42865489</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000436416.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATCGCCAGTAAGCAATAATGCAGGCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330816</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42725146</td>\n",
       "      <td>42725255</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003714369.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42870004-42870112</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330817</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42725146</td>\n",
       "      <td>42725216</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-205</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003672739.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383168.1_2</td>\n",
       "      <td>chr8:-:42870004-42870073</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000435782.1</td>\n",
       "      <td>CCDS55229.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330819</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42725146</td>\n",
       "      <td>42725255</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003714369.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42870004-42870112</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330820</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42725146</td>\n",
       "      <td>42725255</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003714369.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42870004-42870112</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330821</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42725146</td>\n",
       "      <td>42725255</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-208</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003714369.1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383170.1_2</td>\n",
       "      <td>chr8:-:42870004-42870112</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000436416.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330823</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42729073</td>\n",
       "      <td>42729149</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003643269.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42873931-42874006</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGTTCTGTTTGAAGCTGTTCTCGAAGTACCCTTACTAGCTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330825</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42729073</td>\n",
       "      <td>42729149</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003643269.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42873931-42874006</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGTTCTGTTTGAAGCTGTTCTCGAAGTACCCTTACTAGCTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330826</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42729073</td>\n",
       "      <td>42729149</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003643269.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42873931-42874006</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGTTCTGTTTGAAGCTGTTCTCGAAGTACCCTTACTAGCTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330827</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42729073</td>\n",
       "      <td>42729149</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-208</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003643269.1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383170.1_2</td>\n",
       "      <td>chr8:-:42873931-42874006</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000436416.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGTTCTGTTTGAAGCTGTTCTCGAAGTACCCTTACTAGCTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330829</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42742870</td>\n",
       "      <td>42743007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-209</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00003567977.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr8:-:42887728-42887864</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000445725.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGAAAAGTGCATATACCAGGGTAGCAATCAAAGCGAAACTGACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330831</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42742870</td>\n",
       "      <td>42743007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-203</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003567977.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383165.1_2</td>\n",
       "      <td>chr8:-:42887728-42887864</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000326138.3</td>\n",
       "      <td>CCDS55230.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGAAAAGTGCATATACCAGGGTAGCAATCAAAGCGAAACTGACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330833</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42742870</td>\n",
       "      <td>42743007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-206</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00003567977.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383166.1_3</td>\n",
       "      <td>chr8:-:42887728-42887864</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000434797.1</td>\n",
       "      <td>CCDS6138.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGAAAAGTGCATATACCAGGGTAGCAATCAAAGCGAAACTGACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330834</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42742870</td>\n",
       "      <td>42743007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNF170-208</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003567977.1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>HGNC:25358</td>\n",
       "      <td>cds_end_NF</td>\n",
       "      <td>OTTHUMG00000165277.1_6</td>\n",
       "      <td>OTTHUMT00000383170.1_2</td>\n",
       "      <td>chr8:-:42887728-42887864</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000436416.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGAAAAGTGCATATACCAGGGTAGCAATCAAAGCGAAACTGACCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "330801       chr8   HAVANA     CDS  42705933  42706140     .      -     0   \n",
       "330802       chr8  ENSEMBL     CDS  42711301  42711571     .      -     0   \n",
       "330803       chr8   HAVANA     CDS  42711301  42711571     .      -     0   \n",
       "330804       chr8   HAVANA     CDS  42711301  42711571     .      -     0   \n",
       "330805       chr8  ENSEMBL     CDS  42716887  42716998     .      -     0   \n",
       "330806       chr8   HAVANA     CDS  42716887  42716998     .      -     0   \n",
       "330807       chr8   HAVANA     CDS  42716887  42716998     .      -     0   \n",
       "330810       chr8  ENSEMBL     CDS  42720558  42720632     .      -     2   \n",
       "330811       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "330812       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "330813       chr8   HAVANA     CDS  42720558  42720632     .      -     2   \n",
       "330814       chr8   HAVANA     CDS  42720605  42720632     .      -     2   \n",
       "330816       chr8  ENSEMBL     CDS  42725146  42725255     .      -     0   \n",
       "330817       chr8   HAVANA     CDS  42725146  42725216     .      -     0   \n",
       "330819       chr8   HAVANA     CDS  42725146  42725255     .      -     0   \n",
       "330820       chr8   HAVANA     CDS  42725146  42725255     .      -     0   \n",
       "330821       chr8   HAVANA     CDS  42725146  42725255     .      -     0   \n",
       "330823       chr8  ENSEMBL     CDS  42729073  42729149     .      -     1   \n",
       "330825       chr8   HAVANA     CDS  42729073  42729149     .      -     1   \n",
       "330826       chr8   HAVANA     CDS  42729073  42729149     .      -     1   \n",
       "330827       chr8   HAVANA     CDS  42729073  42729149     .      -     1   \n",
       "330829       chr8  ENSEMBL     CDS  42742870  42743007     .      -     0   \n",
       "330831       chr8   HAVANA     CDS  42742870  42743007     .      -     0   \n",
       "330833       chr8   HAVANA     CDS  42742870  42743007     .      -     0   \n",
       "330834       chr8   HAVANA     CDS  42742870  42743007     .      -     0   \n",
       "\n",
       "                     gene_id        transcript_id       gene_type gene_name  \\\n",
       "330801  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "330802  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "330803  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "330804  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "330805  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "330806  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "330807  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "330810  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "330811  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "330812  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "330813  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "330814  ENSG00000120925.16_6  ENST00000531440.5_2  protein_coding    RNF170   \n",
       "330816  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "330817  ENSG00000120925.16_6  ENST00000526349.5_2  protein_coding    RNF170   \n",
       "330819  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "330820  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "330821  ENSG00000120925.16_6  ENST00000531440.5_2  protein_coding    RNF170   \n",
       "330823  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "330825  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "330826  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "330827  ENSG00000120925.16_6  ENST00000531440.5_2  protein_coding    RNF170   \n",
       "330829  ENSG00000120925.16_6  ENST00000534961.5_2  protein_coding    RNF170   \n",
       "330831  ENSG00000120925.16_6  ENST00000319104.7_2  protein_coding    RNF170   \n",
       "330833  ENSG00000120925.16_6  ENST00000527424.6_3  protein_coding    RNF170   \n",
       "330834  ENSG00000120925.16_6  ENST00000531440.5_2  protein_coding    RNF170   \n",
       "\n",
       "       transcript_type transcript_name  exon_number            exon_id level  \\\n",
       "330801  protein_coding      RNF170-203            1  ENSE00001239742.5     2   \n",
       "330802  protein_coding      RNF170-209            1  ENSE00001239728.3     3   \n",
       "330803  protein_coding      RNF170-205            1  ENSE00002152697.1     2   \n",
       "330804  protein_coding      RNF170-206            1  ENSE00002184629.1     2   \n",
       "330805  protein_coding      RNF170-209            2  ENSE00003614933.1     3   \n",
       "330806  protein_coding      RNF170-205            2  ENSE00003614933.1     2   \n",
       "330807  protein_coding      RNF170-206            2  ENSE00003614933.1     2   \n",
       "330810  protein_coding      RNF170-209            3  ENSE00003678305.1     3   \n",
       "330811  protein_coding      RNF170-203            2  ENSE00003678305.1     2   \n",
       "330812  protein_coding      RNF170-205            3  ENSE00003678305.1     2   \n",
       "330813  protein_coding      RNF170-206            3  ENSE00003678305.1     2   \n",
       "330814  protein_coding      RNF170-208            1  ENSE00002147066.1     2   \n",
       "330816  protein_coding      RNF170-209            4  ENSE00003714369.1     3   \n",
       "330817  protein_coding      RNF170-205            4  ENSE00003672739.1     2   \n",
       "330819  protein_coding      RNF170-203            3  ENSE00003714369.1     2   \n",
       "330820  protein_coding      RNF170-206            4  ENSE00003714369.1     2   \n",
       "330821  protein_coding      RNF170-208            2  ENSE00003714369.1     2   \n",
       "330823  protein_coding      RNF170-209            5  ENSE00003643269.1     3   \n",
       "330825  protein_coding      RNF170-203            4  ENSE00003643269.1     2   \n",
       "330826  protein_coding      RNF170-206            5  ENSE00003643269.1     2   \n",
       "330827  protein_coding      RNF170-208            3  ENSE00003643269.1     2   \n",
       "330829  protein_coding      RNF170-209            6  ENSE00003567977.1     3   \n",
       "330831  protein_coding      RNF170-203            5  ENSE00003567977.1     2   \n",
       "330833  protein_coding      RNF170-206            6  ENSE00003567977.1     2   \n",
       "330834  protein_coding      RNF170-208            4  ENSE00003567977.1     2   \n",
       "\n",
       "       transcript_support_level     hgnc_id         tag  \\\n",
       "330801                        1  HGNC:25358        CCDS   \n",
       "330802                        1  HGNC:25358        CCDS   \n",
       "330803                        1  HGNC:25358        CCDS   \n",
       "330804                        1  HGNC:25358        CCDS   \n",
       "330805                        1  HGNC:25358        CCDS   \n",
       "330806                        1  HGNC:25358        CCDS   \n",
       "330807                        1  HGNC:25358        CCDS   \n",
       "330810                        1  HGNC:25358        CCDS   \n",
       "330811                        1  HGNC:25358        CCDS   \n",
       "330812                        1  HGNC:25358        CCDS   \n",
       "330813                        1  HGNC:25358        CCDS   \n",
       "330814                        4  HGNC:25358  cds_end_NF   \n",
       "330816                        1  HGNC:25358        CCDS   \n",
       "330817                        1  HGNC:25358        CCDS   \n",
       "330819                        1  HGNC:25358        CCDS   \n",
       "330820                        1  HGNC:25358        CCDS   \n",
       "330821                        4  HGNC:25358  cds_end_NF   \n",
       "330823                        1  HGNC:25358        CCDS   \n",
       "330825                        1  HGNC:25358        CCDS   \n",
       "330826                        1  HGNC:25358        CCDS   \n",
       "330827                        4  HGNC:25358  cds_end_NF   \n",
       "330829                        1  HGNC:25358        CCDS   \n",
       "330831                        1  HGNC:25358        CCDS   \n",
       "330833                        1  HGNC:25358        CCDS   \n",
       "330834                        4  HGNC:25358  cds_end_NF   \n",
       "\n",
       "                   havana_gene       havana_transcript  \\\n",
       "330801  OTTHUMG00000165277.1_6  OTTHUMT00000383165.1_2   \n",
       "330802  OTTHUMG00000165277.1_6                     NaN   \n",
       "330803  OTTHUMG00000165277.1_6  OTTHUMT00000383168.1_2   \n",
       "330804  OTTHUMG00000165277.1_6  OTTHUMT00000383166.1_3   \n",
       "330805  OTTHUMG00000165277.1_6                     NaN   \n",
       "330806  OTTHUMG00000165277.1_6  OTTHUMT00000383168.1_2   \n",
       "330807  OTTHUMG00000165277.1_6  OTTHUMT00000383166.1_3   \n",
       "330810  OTTHUMG00000165277.1_6                     NaN   \n",
       "330811  OTTHUMG00000165277.1_6  OTTHUMT00000383165.1_2   \n",
       "330812  OTTHUMG00000165277.1_6  OTTHUMT00000383168.1_2   \n",
       "330813  OTTHUMG00000165277.1_6  OTTHUMT00000383166.1_3   \n",
       "330814  OTTHUMG00000165277.1_6  OTTHUMT00000383170.1_2   \n",
       "330816  OTTHUMG00000165277.1_6                     NaN   \n",
       "330817  OTTHUMG00000165277.1_6  OTTHUMT00000383168.1_2   \n",
       "330819  OTTHUMG00000165277.1_6  OTTHUMT00000383165.1_2   \n",
       "330820  OTTHUMG00000165277.1_6  OTTHUMT00000383166.1_3   \n",
       "330821  OTTHUMG00000165277.1_6  OTTHUMT00000383170.1_2   \n",
       "330823  OTTHUMG00000165277.1_6                     NaN   \n",
       "330825  OTTHUMG00000165277.1_6  OTTHUMT00000383165.1_2   \n",
       "330826  OTTHUMG00000165277.1_6  OTTHUMT00000383166.1_3   \n",
       "330827  OTTHUMG00000165277.1_6  OTTHUMT00000383170.1_2   \n",
       "330829  OTTHUMG00000165277.1_6                     NaN   \n",
       "330831  OTTHUMG00000165277.1_6  OTTHUMT00000383165.1_2   \n",
       "330833  OTTHUMG00000165277.1_6  OTTHUMT00000383166.1_3   \n",
       "330834  OTTHUMG00000165277.1_6  OTTHUMT00000383170.1_2   \n",
       "\n",
       "         remap_original_location remap_status remap_num_mappings  \\\n",
       "330801  chr8:-:42850794-42850997  full_contig                NaN   \n",
       "330802  chr8:-:42856162-42856428  full_contig                NaN   \n",
       "330803  chr8:-:42856162-42856428  full_contig                NaN   \n",
       "330804  chr8:-:42856162-42856428  full_contig                NaN   \n",
       "330805  chr8:-:42861745-42861855  full_contig                NaN   \n",
       "330806  chr8:-:42861745-42861855  full_contig                NaN   \n",
       "330807  chr8:-:42861745-42861855  full_contig                NaN   \n",
       "330810  chr8:-:42865416-42865489  full_contig                NaN   \n",
       "330811  chr8:-:42865416-42865489  full_contig                NaN   \n",
       "330812  chr8:-:42865416-42865489  full_contig                NaN   \n",
       "330813  chr8:-:42865416-42865489  full_contig                NaN   \n",
       "330814  chr8:-:42865466-42865489  full_contig                NaN   \n",
       "330816  chr8:-:42870004-42870112  full_contig                NaN   \n",
       "330817  chr8:-:42870004-42870073  full_contig                NaN   \n",
       "330819  chr8:-:42870004-42870112  full_contig                NaN   \n",
       "330820  chr8:-:42870004-42870112  full_contig                NaN   \n",
       "330821  chr8:-:42870004-42870112  full_contig                NaN   \n",
       "330823  chr8:-:42873931-42874006  full_contig                NaN   \n",
       "330825  chr8:-:42873931-42874006  full_contig                NaN   \n",
       "330826  chr8:-:42873931-42874006  full_contig                NaN   \n",
       "330827  chr8:-:42873931-42874006  full_contig                NaN   \n",
       "330829  chr8:-:42887728-42887864  full_contig                NaN   \n",
       "330831  chr8:-:42887728-42887864  full_contig                NaN   \n",
       "330833  chr8:-:42887728-42887864  full_contig                NaN   \n",
       "330834  chr8:-:42887728-42887864  full_contig                NaN   \n",
       "\n",
       "       remap_target_status  ont         protein_id       ccdsid gene_status  \\\n",
       "330801                 NaN  NaN  ENSP00000326138.3  CCDS55230.1         NaN   \n",
       "330802                 NaN  NaN  ENSP00000445725.1   CCDS6138.1         NaN   \n",
       "330803                 NaN  NaN  ENSP00000435782.1  CCDS55229.1         NaN   \n",
       "330804                 NaN  NaN  ENSP00000434797.1   CCDS6138.1         NaN   \n",
       "330805                 NaN  NaN  ENSP00000445725.1   CCDS6138.1         NaN   \n",
       "330806                 NaN  NaN  ENSP00000435782.1  CCDS55229.1         NaN   \n",
       "330807                 NaN  NaN  ENSP00000434797.1   CCDS6138.1         NaN   \n",
       "330810                 NaN  NaN  ENSP00000445725.1   CCDS6138.1         NaN   \n",
       "330811                 NaN  NaN  ENSP00000326138.3  CCDS55230.1         NaN   \n",
       "330812                 NaN  NaN  ENSP00000435782.1  CCDS55229.1         NaN   \n",
       "330813                 NaN  NaN  ENSP00000434797.1   CCDS6138.1         NaN   \n",
       "330814                 NaN  NaN  ENSP00000436416.1          NaN         NaN   \n",
       "330816                 NaN  NaN  ENSP00000445725.1   CCDS6138.1         NaN   \n",
       "330817                 NaN  NaN  ENSP00000435782.1  CCDS55229.1         NaN   \n",
       "330819                 NaN  NaN  ENSP00000326138.3  CCDS55230.1         NaN   \n",
       "330820                 NaN  NaN  ENSP00000434797.1   CCDS6138.1         NaN   \n",
       "330821                 NaN  NaN  ENSP00000436416.1          NaN         NaN   \n",
       "330823                 NaN  NaN  ENSP00000445725.1   CCDS6138.1         NaN   \n",
       "330825                 NaN  NaN  ENSP00000326138.3  CCDS55230.1         NaN   \n",
       "330826                 NaN  NaN  ENSP00000434797.1   CCDS6138.1         NaN   \n",
       "330827                 NaN  NaN  ENSP00000436416.1          NaN         NaN   \n",
       "330829                 NaN  NaN  ENSP00000445725.1   CCDS6138.1         NaN   \n",
       "330831                 NaN  NaN  ENSP00000326138.3  CCDS55230.1         NaN   \n",
       "330833                 NaN  NaN  ENSP00000434797.1   CCDS6138.1         NaN   \n",
       "330834                 NaN  NaN  ENSP00000436416.1          NaN         NaN   \n",
       "\n",
       "       transcript_status remap_substituted_missing_target  \\\n",
       "330801               NaN                              NaN   \n",
       "330802               NaN                              NaN   \n",
       "330803               NaN                              NaN   \n",
       "330804               NaN                              NaN   \n",
       "330805               NaN                              NaN   \n",
       "330806               NaN                              NaN   \n",
       "330807               NaN                              NaN   \n",
       "330810               NaN                              NaN   \n",
       "330811               NaN                              NaN   \n",
       "330812               NaN                              NaN   \n",
       "330813               NaN                              NaN   \n",
       "330814               NaN                              NaN   \n",
       "330816               NaN                              NaN   \n",
       "330817               NaN                              NaN   \n",
       "330819               NaN                              NaN   \n",
       "330820               NaN                              NaN   \n",
       "330821               NaN                              NaN   \n",
       "330823               NaN                              NaN   \n",
       "330825               NaN                              NaN   \n",
       "330826               NaN                              NaN   \n",
       "330827               NaN                              NaN   \n",
       "330829               NaN                              NaN   \n",
       "330831               NaN                              NaN   \n",
       "330833               NaN                              NaN   \n",
       "330834               NaN                              NaN   \n",
       "\n",
       "                                             Exon_CDS_seq  \n",
       "330801  TTAACCTGAGAACCAAAGGATGGAAACACAGTCTAGTAACGTGAAG...  \n",
       "330802  TCATCTAGTTAGCCTTTGGGTTATCACTTCTCGATACATAATAGAG...  \n",
       "330803  TCATCTAGTTAGCCTTTGGGTTATCACTTCTCGATACATAATAGAG...  \n",
       "330804  TCATCTAGTTAGCCTTTGGGTTATCACTTCTCGATACATAATAGAG...  \n",
       "330805  AGATCTGGGTTGCCCTGAGAATCTCCGGTTATAATCATTAATATCC...  \n",
       "330806  AGATCTGGGTTGCCCTGAGAATCTCCGGTTATAATCATTAATATCC...  \n",
       "330807  AGATCTGGGTTGCCCTGAGAATCTCCGGTTATAATCATTAATATCC...  \n",
       "330810  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "330811  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "330812  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "330813  CGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACCA...  \n",
       "330814                        ATCGCCAGTAAGCAATAATGCAGGCAC  \n",
       "330816  CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...  \n",
       "330817  CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...  \n",
       "330819  CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...  \n",
       "330820  CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...  \n",
       "330821  CACAAAAAAGATGTCCACAGTTGGTCTCCACCGGGAAGGAGGCTTG...  \n",
       "330823  CTGTTCTGTTTGAAGCTGTTCTCGAAGTACCCTTACTAGCTCCTGG...  \n",
       "330825  CTGTTCTGTTTGAAGCTGTTCTCGAAGTACCCTTACTAGCTCCTGG...  \n",
       "330826  CTGTTCTGTTTGAAGCTGTTCTCGAAGTACCCTTACTAGCTCCTGG...  \n",
       "330827  CTGTTCTGTTTGAAGCTGTTCTCGAAGTACCCTTACTAGCTCCTGG...  \n",
       "330829  CTGAAAAGTGCATATACCAGGGTAGCAATCAAAGCGAAACTGACCA...  \n",
       "330831  CTGAAAAGTGCATATACCAGGGTAGCAATCAAAGCGAAACTGACCA...  \n",
       "330833  CTGAAAAGTGCATATACCAGGGTAGCAATCAAAGCGAAACTGACCA...  \n",
       "330834  CTGAAAAGTGCATATACCAGGGTAGCAATCAAAGCGAAACTGACCA...  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_df_adj = add_exon_cds_sequence(cds_df_adj, fasta)\n",
    "cds_df_adj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fe23b222-41d9-46c3-a6a8-6f93f1f1aad8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  \n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]  \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]  \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]  \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]  \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]  \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...  \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...  \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]  \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...  \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def create_reference_cds(intersection_cds_vcf, cds_df_test):\n",
    "\n",
    "    \"\"\"\n",
    "    Constructs the whole CDS sequence (multiple exons) for transcripts affected by a variant, both in their reference\n",
    "    and alternative form.\n",
    "    :param intersection_cds_vcf: DataFrame containing variant-CDS intersection and corresponding alternative CDS sequences\n",
    "                                 includes: transcript_id, Exon_CDS_seq + length, Exon_ALT_CDS_seq + length\n",
    "    :param cds_df_test: Reference exon-level CDS data for all transcripts with exon_number\n",
    "                        includes: transcript_id, exon_number, Start, End, Strand, Exon_CDS_seq\n",
    "    :return: DataFrame with one row per variant-transcript pair, containing full reference and alternative CDS + lengths,\n",
    "             exon-wise CDS information as tuple (exon number, exon-wise CDS length)\n",
    "    \"\"\"\n",
    "\n",
    "    results = []\n",
    "\n",
    "    for transcript_id, var_df in intersection_cds_vcf.groupby(\"transcript_id\"):  # Only transcripts with a variant\n",
    "\n",
    "        # 1. Get reference exons\n",
    "        ref_exons = cds_df_test[cds_df_test[\"transcript_id\"] == transcript_id].copy()\n",
    "        ref_exons = ref_exons.sort_values(\"Start\")\n",
    "\n",
    "        # Get reference CDS sequence start und stop position for finding position in transcript sequence\n",
    "        ref_cds_start = ref_exons[\"Start\"].min()\n",
    "        ref_cds_stop = ref_exons[\"End\"].max()\n",
    "\n",
    "        # Join reference exon sequences to form full CDS sequence\n",
    "        ref_seq = \"\".join(ref_exons[\"Exon_CDS_seq\"].tolist())\n",
    "\n",
    "        ######\n",
    "        # Since I sometimes get errors in the following code snippet because of NaN values,\n",
    "        # we print them but leave them in our dataframe for now\n",
    "        #nan_rows = ref_exons[ref_exons[\"Exon_CDS_seq\"].isna()]\n",
    "        #if not nan_rows.empty:\n",
    "        #    print(f\"\\n[Warning] Found {len(nan_rows)} NaN Exon_CDS_seq entries in transcript: {transcript_id}\")\n",
    "        #    print(nan_rows.to_string(index=False))\n",
    "\n",
    "        # Collect exon numbers and lengths (for tracking exon contribution later on)\n",
    "        # ref_cds_lengths = [len(seq) for seq in ref_exons[\"Exon_CDS_seq\"].tolist()]\n",
    "        \n",
    "        ## old option:\n",
    "        #ref_cds_info = [\n",
    "        #    (row[\"exon_number\"], len(row[\"Exon_CDS_seq\"]))\n",
    "        #    for _, row in ref_exons.iterrows()\n",
    "        #]\n",
    "\n",
    "        ## new option:\n",
    "        ref_cds_info = sorted([\n",
    "            (row[\"exon_number\"], len(row[\"Exon_CDS_seq\"]))\n",
    "            for _, row in ref_exons.iterrows()\n",
    "        ], key=lambda x: x[0])\n",
    "\n",
    "        ######\n",
    "\n",
    "        # Get strand info (all should be the same within transcript)\n",
    "        strand = ref_exons[\"Strand\"].iloc[0]\n",
    "\n",
    "        for variant, cds_df in var_df.groupby([\"Chromosome\", \"Start_variant\", \"End_variant\", \"Ref\", \"Alt\"],\n",
    "                                              observed=True):\n",
    "\n",
    "            # Sort variant exons\n",
    "            cds_df = cds_df.sort_values(\"Start\")\n",
    "\n",
    "            # Copy ref exons for modification\n",
    "            alt_exons = ref_exons.copy()\n",
    "\n",
    "            # Replace affected exon sequences with variant versions\n",
    "            for _, var_row in cds_df.iterrows():\n",
    "                exon_nr = var_row[\"exon_number\"]\n",
    "                alt_exons.loc[alt_exons[\"exon_number\"] == exon_nr, \"Exon_CDS_seq\"] = var_row[\"Exon_Alt_CDS_seq\"]\n",
    "\n",
    "            # Join and sort alt CDS\n",
    "            alt_exons = alt_exons.sort_values(\"Start\")\n",
    "\n",
    "            ######\n",
    "            # Since I sometimes get errors in the following code snippet because of NaN values,\n",
    "            # we print them but leave them in our dataframe for now (same as before)\n",
    "            #nan_alt_rows = alt_exons[alt_exons[\"Exon_CDS_seq\"].isna()]\n",
    "            #if not nan_alt_rows.empty:\n",
    "            #    print(f\"\\n[Warning] NaN Exon_CDS_seq values found in alt_exons for variant in transcript: {transcript_id}\")\n",
    "            #    print(nan_alt_rows.to_string(index=False))\n",
    "\n",
    "            # alt_cds_lengths = [len(seq) for seq in alt_exons[\"Exon_CDS_seq\"].tolist()]\n",
    "            \n",
    "            ## old option:\n",
    "            #alt_cds_info = [\n",
    "            #    (row[\"exon_number\"], len(row[\"Exon_CDS_seq\"]))\n",
    "            #    for _, row in alt_exons.iterrows()\n",
    "            #]\n",
    "\n",
    "            ## new option:\n",
    "            alt_cds_info = sorted([\n",
    "                (row[\"exon_number\"], len(row[\"Exon_CDS_seq\"]))\n",
    "                for _, row in alt_exons.iterrows()\n",
    "            ], key=lambda x: x[0])\n",
    "            ######\n",
    "\n",
    "            # Get alternative CDS sequence start and stop position for finding position in transcript sequence\n",
    "            alt_cds_start = alt_exons[\"Start\"].min()\n",
    "            alt_cds_stop = alt_exons[\"End\"].max()\n",
    "\n",
    "            alt_seq = \"\".join(alt_exons[\"Exon_CDS_seq\"].tolist())\n",
    "\n",
    "            # Apply reverse complement if on minus strand\n",
    "            if strand == \"-\":\n",
    "                ref_seq_final = str(Seq(ref_seq).reverse_complement())\n",
    "                alt_seq_final = str(Seq(alt_seq).reverse_complement())\n",
    "            else:\n",
    "                ref_seq_final = ref_seq\n",
    "                alt_seq_final = alt_seq\n",
    "\n",
    "            # Append to results\n",
    "            results.append({\n",
    "                \"transcript_id\": transcript_id,\n",
    "                \"variant_id\": var_row[\"ID\"],\n",
    "\n",
    "                \"ref_cds_start\": ref_cds_start,\n",
    "                \"ref_cds_stop\": ref_cds_stop,\n",
    "                \"ref_cds_seq\": ref_seq_final,\n",
    "                \"ref_cds_len\": len(ref_seq_final),\n",
    "\n",
    "                \"alt_cds_start\": alt_cds_start,\n",
    "                \"alt_cds_stop\": alt_cds_stop,\n",
    "                \"alt_cds_seq\": alt_seq_final,\n",
    "                \"alt_cds_len\": len(alt_seq_final),\n",
    "\n",
    "                \"chromosome\": var_row[\"Chromosome\"],\n",
    "                \"gene_id\": var_row[\"gene_id\"],\n",
    "                \"strand\": strand,\n",
    "                \"ref\": var_row[\"Ref\"],\n",
    "                \"alt\": var_row[\"Alt\"],\n",
    "                \"start_variant\": var_row[\"Start_variant\"],\n",
    "                \"end_variant\": var_row[\"End_variant\"],\n",
    "\n",
    "                \"ref_cds_info\": ref_cds_info,\n",
    "                \"alt_cds_info\": alt_cds_info,\n",
    "\n",
    "                # \"ref_cds_lengths\": [length for exon_num, length in ref_cds_info],\n",
    "                # \"alt_cds_lengths\": [length for exon_num, length in alt_cds_info]\n",
    "            })\n",
    "\n",
    "    results_df = pd.DataFrame(results)\n",
    "    return(results_df)\n",
    "\n",
    "results_df = create_reference_cds(intersection_cds_vcf, cds_df_adj)\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f4c34570-4b6f-4efb-9515-f8f25b352123",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript_sequence(exons_df, fasta):\n",
    "\n",
    "    \"\"\"\n",
    "    Construct full transcript sequences by concatenating the exon sequences from the FASTA genome reference, grouped by transcript.\n",
    "    Get transcript length and transcript information as well.\n",
    "    :param exons_df: DataFrame containing exon-level annotations from the GTF file.\n",
    "                     Must include: transcript_id, strand, chromosome, start, end, exon_number\n",
    "    :param fasta: Fasta file, reference genome object\n",
    "    :return: DataFrame with one row per transcript with full transcript sequence, start, end, strand, transcript sequence length, and\n",
    "             per exon sequence length information for that transcript\n",
    "    \"\"\"\n",
    "\n",
    "    exon_data = []\n",
    "\n",
    "    # Process each transcript individually\n",
    "    for transcript_id, group in exons_df.groupby(\"transcript_id\"):\n",
    "        strand = group.iloc[0][\"Strand\"]\n",
    "\n",
    "        if strand not in [\"+\", \"-\"]:\n",
    "            print(f\"Unknown strand for {transcript_id}\")\n",
    "            continue\n",
    "\n",
    "        # Sort by exon start coordinate (strand not considered here yet)\n",
    "        group_sorted = group.sort_values(by=\"Start\").copy()\n",
    "\n",
    "        seq_parts = [] # to accumulate exon sequences\n",
    "        starts = [] # for overall transcript start\n",
    "        ends = [] # for overall transcript end\n",
    "        exon_info = [] # for tracking exon_number and length\n",
    "\n",
    "        # fetch exon sequence and metadata\n",
    "        for _, row in group_sorted.iterrows():\n",
    "            chrom = row[\"Chromosome\"]\n",
    "            start = int(row[\"Start\"])\n",
    "            end = int(row[\"End\"])\n",
    "            exon_number = row[\"exon_number\"]\n",
    "\n",
    "            starts.append(start)\n",
    "            ends.append(end)\n",
    "\n",
    "            # Fetch exon sequence from fasta reference genome\n",
    "            exon_seq = fasta[chrom][start:end] #.seq\n",
    "            exon_seq_str = str(exon_seq).upper()\n",
    "            seq_parts.append(exon_seq_str)\n",
    "\n",
    "            exon_info.append((exon_number, len(exon_seq_str)))\n",
    "\n",
    "        # join exon sequences into a full transcript sequence\n",
    "        joined_seq = \"\".join(seq_parts)\n",
    "\n",
    "        # Apply reverse complement for minus strand transcripts\n",
    "        if strand == \"-\":\n",
    "            joined_seq = str(Seq(joined_seq).reverse_complement())\n",
    "            exon_info = exon_info[::-1]\n",
    "\n",
    "        exon_data.append({\n",
    "            \"Chromosome\": chrom,\n",
    "            \"transcript_id\": transcript_id,\n",
    "            \"start\": min(starts),\n",
    "            \"end\": max(ends),\n",
    "            \"strand\": strand,\n",
    "            \"transcript_sequence\": joined_seq,\n",
    "            \"transcript_length\": len(joined_seq),\n",
    "            \"transcript_exon_info\": exon_info\n",
    "        })\n",
    "\n",
    "    exon_seqs = pd.DataFrame(exon_data)\n",
    "    return(exon_seqs)\n",
    "\n",
    "\n",
    "exons_df = exons_df[exons_df[\"transcript_id\"].isin(relevant_transcripts)].copy()\n",
    "exon_seqs = get_transcript_sequence(exons_df, fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4873800c-7c23-41dc-8b5f-1d160bb3a153",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_sequence</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>transcript_exon_info</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000319104.7_2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>-</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000526349.5_2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000527424.6_3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>-</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000531440.5_2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>42720608</td>\n",
       "      <td>42751741</td>\n",
       "      <td>-</td>\n",
       "      <td>AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...</td>\n",
       "      <td>580</td>\n",
       "      <td>[(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000534961.5_2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>-</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Chromosome     start       end strand  \\\n",
       "transcript_id                                               \n",
       "ENST00000319104.7_2       chr8  42704779  42751728      -   \n",
       "ENST00000526349.5_2       chr8  42710849  42751849      -   \n",
       "ENST00000527424.6_3       chr8  42708443  42751748      -   \n",
       "ENST00000531440.5_2       chr8  42720608  42751741      -   \n",
       "ENST00000534961.5_2       chr8  42708439  42751866      -   \n",
       "\n",
       "                                                   transcript_sequence  \\\n",
       "transcript_id                                                            \n",
       "ENST00000319104.7_2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...   \n",
       "ENST00000526349.5_2  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...   \n",
       "ENST00000527424.6_3  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...   \n",
       "ENST00000531440.5_2  AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...   \n",
       "ENST00000534961.5_2  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...   \n",
       "\n",
       "                     transcript_length  \\\n",
       "transcript_id                            \n",
       "ENST00000319104.7_2               1866   \n",
       "ENST00000526349.5_2               1338   \n",
       "ENST00000527424.6_3               3764   \n",
       "ENST00000531440.5_2                580   \n",
       "ENST00000534961.5_2               4116   \n",
       "\n",
       "                                                  transcript_exon_info  \n",
       "transcript_id                                                           \n",
       "ENST00000319104.7_2  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...  \n",
       "ENST00000526349.5_2  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...  \n",
       "ENST00000527424.6_3  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...  \n",
       "ENST00000531440.5_2  [(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...  \n",
       "ENST00000534961.5_2  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_seqs_indexed = exon_seqs.set_index(\"transcript_id\")\n",
    "exon_seqs_indexed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7282c5a2-45e8-4249-aa2b-5e416c585931",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "      <th>cds_in_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  cds_in_transcript  \n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True  \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]               True  \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True  \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]               True  \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]               True  \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True  \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True  \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]              False  \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True  \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def check_cds_in_transcript(row):\n",
    "    transcript_id = row[\"transcript_id\"]\n",
    "\n",
    "    # Skip if transcript_id not found\n",
    "    if transcript_id not in exon_seqs_indexed.index:\n",
    "        return False\n",
    "\n",
    "    transcript_seq = exon_seqs_indexed.loc[transcript_id, \"transcript_sequence\"]\n",
    "    ref_cds_seq = row[\"ref_cds_seq\"]\n",
    "\n",
    "    # Check if CDS is a substring of the transcript\n",
    "    return ref_cds_seq in transcript_seq\n",
    "\n",
    "results_df[\"cds_in_transcript\"] = results_df.apply(check_cds_in_transcript, axis=1)\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "df5ba5c7-673b-4fef-aac7-390dff05b347",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_exon(cds_pos, exon_info):\n",
    "    \"\"\"\n",
    "    Map a CDS-relative position to the corresponding exon number using exon_info,\n",
    "    which is a list of (exon_number, exon_length) tuples in CDS order.\n",
    "    \"\"\"\n",
    "    pos_counter = 0\n",
    "    for exon_number, exon_length in exon_info:\n",
    "        if cds_pos < pos_counter + exon_length:\n",
    "            return exon_number\n",
    "        pos_counter += exon_length\n",
    "    return exon_info[-1][0]  # fallback\n",
    "\n",
    "def analyze_sequence(results_df):\n",
    "\n",
    "    \"\"\"\n",
    "    Analyzes reference and alternative CDS for start and stop codons, their positions, and potential premature termination codons (PTCs)\n",
    "\n",
    "    :param results_df: DataFrame containing CDS sequences and exon information for both reference and alternative sequences, per variant\n",
    "    :return: DataFrame with added annotation columns for reference and alternative sequence separately:\n",
    "             such as start codon position / exon, last codon and its validity as stop codon, first in-frame stop codon + position,\n",
    "             number and information of all available stop codons, premature stop codon flag\n",
    "    \"\"\"\n",
    "\n",
    "    valid_stop_codons = {\"TAA\", \"TAG\", \"TGA\"}\n",
    "    start_codon = \"ATG\"\n",
    "\n",
    "    df = results_df.copy()\n",
    "\n",
    "    # Initialize result columns for both reference and alternative sequence\n",
    "    for label in [\"ref\", \"alt\"]:\n",
    "        df[f\"{label}_start_codon_pos\"] = None\n",
    "        df[f\"{label}_start_codon_exon\"] = None  # exon number\n",
    "        df[f\"{label}_last_codon\"] = None\n",
    "        df[f\"{label}_valid_stop\"] = None\n",
    "        df[f\"{label}_first_stop_codon\"] = None\n",
    "        df[f\"{label}_first_stop_pos\"] = None\n",
    "        df[f\"{label}_num_stop_codons\"] = None\n",
    "        df[f\"{label}_all_stop_codons\"] = None\n",
    "        df[f\"{label}_stop_codon_exons\"] = None  # exon number\n",
    "        df[f\"{label}_is_premature\"] = None\n",
    "\n",
    "    # Row-wise codon scanning\n",
    "    for idx, row in df.iterrows():\n",
    "        for label in [\"ref\", \"alt\"]:\n",
    "            seq = row[f\"{label}_cds_seq\"]\n",
    "\n",
    "            exon_info = row[f\"{label}_cds_info\"]  # for exon number\n",
    "\n",
    "            # Skip invalid or too-short sequences\n",
    "            if not isinstance(seq, str) or len(seq) < 3:\n",
    "                continue\n",
    "\n",
    "            start_pos = None\n",
    "            stop_codons = []\n",
    "            stop_exons = []  # for exon number\n",
    "\n",
    "            # Scan in codons (step=3)\n",
    "            for i in range(0, len(seq) - 2, 3):\n",
    "                codon = seq[i:i + 3]\n",
    "                if codon == start_codon and start_pos is None: # first start codon position\n",
    "                    start_pos = i\n",
    "                if codon in valid_stop_codons: # record all stop codons with their positions and exons\n",
    "                    stop_codons.append((i, codon))\n",
    "                    stop_exons.append(get_exon(i, exon_info))  # for exon number\n",
    "\n",
    "            last_codon = seq[-3:]\n",
    "            is_valid_stop = last_codon in valid_stop_codons\n",
    "            first_stop_pos = stop_codons[0][0] if stop_codons else None\n",
    "            first_stop = stop_codons[0][1] if stop_codons else None\n",
    "            is_premature = first_stop_pos is not None and first_stop_pos < len(seq) - 3\n",
    "            start_exon = get_exon(start_pos, exon_info) if start_pos is not None else None  # for exon number\n",
    "\n",
    "            # Store results\n",
    "            df.at[idx, f\"{label}_start_codon_pos\"] = start_pos\n",
    "            df.at[idx, f\"{label}_start_codon_exon\"] = start_exon  # exon number\n",
    "            df.at[idx, f\"{label}_last_codon\"] = last_codon\n",
    "            df.at[idx, f\"{label}_valid_stop\"] = is_valid_stop\n",
    "            df.at[idx, f\"{label}_first_stop_codon\"] = first_stop\n",
    "            df.at[idx, f\"{label}_first_stop_pos\"] = first_stop_pos\n",
    "            df.at[idx, f\"{label}_num_stop_codons\"] = len(stop_codons)\n",
    "            df.at[idx, f\"{label}_all_stop_codons\"] = stop_codons\n",
    "            df.at[idx, f\"{label}_stop_codon_exons\"] = stop_exons  # exon number\n",
    "            df.at[idx, f\"{label}_is_premature\"] = is_premature\n",
    "\n",
    "    return df\n",
    "\n",
    "analysis_df = analyze_sequence(results_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fda3495b-1e7a-4068-856a-5d4f78014e4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "      <th>cds_in_transcript</th>\n",
       "      <th>ref_start_codon_pos</th>\n",
       "      <th>ref_start_codon_exon</th>\n",
       "      <th>ref_last_codon</th>\n",
       "      <th>ref_valid_stop</th>\n",
       "      <th>ref_first_stop_codon</th>\n",
       "      <th>ref_first_stop_pos</th>\n",
       "      <th>ref_num_stop_codons</th>\n",
       "      <th>ref_all_stop_codons</th>\n",
       "      <th>ref_stop_codon_exons</th>\n",
       "      <th>ref_is_premature</th>\n",
       "      <th>alt_start_codon_pos</th>\n",
       "      <th>alt_start_codon_exon</th>\n",
       "      <th>alt_last_codon</th>\n",
       "      <th>alt_valid_stop</th>\n",
       "      <th>alt_first_stop_codon</th>\n",
       "      <th>alt_first_stop_pos</th>\n",
       "      <th>alt_num_stop_codons</th>\n",
       "      <th>alt_all_stop_codons</th>\n",
       "      <th>alt_stop_codon_exons</th>\n",
       "      <th>alt_is_premature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>351</td>\n",
       "      <td>2</td>\n",
       "      <td>[(351, TGA), (447, TGA)]</td>\n",
       "      <td>[4, 5]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>96</td>\n",
       "      <td>8</td>\n",
       "      <td>[(96, TGA), (102, TGA), (141, TAA), (144, TGA)...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 2, 2, 4]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GTG</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  cds_in_transcript  \\\n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]               True   \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]               True   \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]               True   \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]              False   \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "\n",
       "  ref_start_codon_pos ref_start_codon_exon ref_last_codon ref_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    1            GAT          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  ref_first_stop_codon ref_first_stop_pos ref_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TAA                600                   1   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                522                   1   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                774                   1   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                774                   1   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "  ref_all_stop_codons ref_stop_codon_exons ref_is_premature  \\\n",
       "0        [(600, TAA)]                  [5]            False   \n",
       "1        [(600, TAA)]                  [5]            False   \n",
       "2        [(600, TAA)]                  [5]            False   \n",
       "3        [(522, TGA)]                  [4]            False   \n",
       "4        [(522, TGA)]                  [4]            False   \n",
       "5        [(774, TGA)]                  [6]            False   \n",
       "6        [(774, TGA)]                  [6]            False   \n",
       "7                  []                   []            False   \n",
       "8        [(774, TGA)]                  [6]            False   \n",
       "9        [(774, TGA)]                  [6]            False   \n",
       "\n",
       "  alt_start_codon_pos alt_start_codon_exon alt_last_codon alt_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    2            GTG          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  alt_first_stop_codon alt_first_stop_pos alt_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TGA                351                   2   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                 96                   8   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                348                   8   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                348                   8   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "                                 alt_all_stop_codons  \\\n",
       "0                                       [(600, TAA)]   \n",
       "1                           [(351, TGA), (447, TGA)]   \n",
       "2                                       [(600, TAA)]   \n",
       "3  [(96, TGA), (102, TGA), (141, TAA), (144, TGA)...   \n",
       "4                                       [(522, TGA)]   \n",
       "5  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "6                                       [(774, TGA)]   \n",
       "7                                                 []   \n",
       "8  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "9                                       [(774, TGA)]   \n",
       "\n",
       "       alt_stop_codon_exons alt_is_premature  \n",
       "0                       [5]            False  \n",
       "1                    [4, 5]             True  \n",
       "2                       [5]            False  \n",
       "3  [1, 1, 1, 1, 1, 2, 2, 4]             True  \n",
       "4                       [4]            False  \n",
       "5  [2, 2, 4, 4, 4, 6, 6, 6]             True  \n",
       "6                       [6]            False  \n",
       "7                        []            False  \n",
       "8  [2, 2, 4, 4, 4, 6, 6, 6]             True  \n",
       "9                       [6]            False  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2d6172fe-fac2-4569-baff-16ec4d724529",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "      <th>cds_in_transcript</th>\n",
       "      <th>ref_start_codon_pos</th>\n",
       "      <th>ref_start_codon_exon</th>\n",
       "      <th>ref_last_codon</th>\n",
       "      <th>ref_valid_stop</th>\n",
       "      <th>ref_first_stop_codon</th>\n",
       "      <th>ref_first_stop_pos</th>\n",
       "      <th>ref_num_stop_codons</th>\n",
       "      <th>ref_all_stop_codons</th>\n",
       "      <th>ref_stop_codon_exons</th>\n",
       "      <th>ref_is_premature</th>\n",
       "      <th>alt_start_codon_pos</th>\n",
       "      <th>alt_start_codon_exon</th>\n",
       "      <th>alt_last_codon</th>\n",
       "      <th>alt_valid_stop</th>\n",
       "      <th>alt_first_stop_codon</th>\n",
       "      <th>alt_first_stop_pos</th>\n",
       "      <th>alt_num_stop_codons</th>\n",
       "      <th>alt_all_stop_codons</th>\n",
       "      <th>alt_stop_codon_exons</th>\n",
       "      <th>alt_is_premature</th>\n",
       "      <th>start_loss</th>\n",
       "      <th>stop_loss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>351</td>\n",
       "      <td>2</td>\n",
       "      <td>[(351, TGA), (447, TGA)]</td>\n",
       "      <td>[4, 5]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>96</td>\n",
       "      <td>8</td>\n",
       "      <td>[(96, TGA), (102, TGA), (141, TAA), (144, TGA)...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 2, 2, 4]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GTG</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  cds_in_transcript  \\\n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]               True   \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]               True   \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]               True   \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]              False   \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "\n",
       "  ref_start_codon_pos ref_start_codon_exon ref_last_codon ref_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    1            GAT          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  ref_first_stop_codon ref_first_stop_pos ref_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TAA                600                   1   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                522                   1   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                774                   1   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                774                   1   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "  ref_all_stop_codons ref_stop_codon_exons ref_is_premature  \\\n",
       "0        [(600, TAA)]                  [5]            False   \n",
       "1        [(600, TAA)]                  [5]            False   \n",
       "2        [(600, TAA)]                  [5]            False   \n",
       "3        [(522, TGA)]                  [4]            False   \n",
       "4        [(522, TGA)]                  [4]            False   \n",
       "5        [(774, TGA)]                  [6]            False   \n",
       "6        [(774, TGA)]                  [6]            False   \n",
       "7                  []                   []            False   \n",
       "8        [(774, TGA)]                  [6]            False   \n",
       "9        [(774, TGA)]                  [6]            False   \n",
       "\n",
       "  alt_start_codon_pos alt_start_codon_exon alt_last_codon alt_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    2            GTG          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  alt_first_stop_codon alt_first_stop_pos alt_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TGA                351                   2   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                 96                   8   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                348                   8   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                348                   8   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "                                 alt_all_stop_codons  \\\n",
       "0                                       [(600, TAA)]   \n",
       "1                           [(351, TGA), (447, TGA)]   \n",
       "2                                       [(600, TAA)]   \n",
       "3  [(96, TGA), (102, TGA), (141, TAA), (144, TGA)...   \n",
       "4                                       [(522, TGA)]   \n",
       "5  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "6                                       [(774, TGA)]   \n",
       "7                                                 []   \n",
       "8  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "9                                       [(774, TGA)]   \n",
       "\n",
       "       alt_stop_codon_exons alt_is_premature  start_loss  stop_loss  \n",
       "0                       [5]            False       False      False  \n",
       "1                    [4, 5]             True       False      False  \n",
       "2                       [5]            False       False      False  \n",
       "3  [1, 1, 1, 1, 1, 2, 2, 4]             True       False      False  \n",
       "4                       [4]            False       False      False  \n",
       "5  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False  \n",
       "6                       [6]            False       False      False  \n",
       "7                        []            False       False       True  \n",
       "8  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False  \n",
       "9                       [6]            False       False      False  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def start_stop_loss(df):\n",
    "    df = df.copy()\n",
    "\n",
    "    # Start codon loss: reference sequence has a start codon, alternative sequence does not or the position is changed\n",
    "    df[\"start_loss\"] = (\n",
    "                               (df[\"ref_start_codon_pos\"].notna()) & df[\"alt_start_codon_pos\"].isna()\n",
    "                       ) | (\n",
    "                               df[\"ref_start_codon_pos\"] != df[\"alt_start_codon_pos\"]\n",
    "                       )\n",
    "\n",
    "    # Stop codon loss: reference sequence had a valid stop codon, the alternative sequence does not or the position is changed\n",
    "    df[\"stop_loss\"] = (\n",
    "                              (df[\"ref_valid_stop\"] == True) & (df[\"alt_valid_stop\"] != True)\n",
    "                      ) | (\n",
    "                              df[\"ref_last_codon\"] != df[\"alt_last_codon\"]  # Or take this out?\n",
    "                      )\n",
    "\n",
    "    return df\n",
    "\n",
    "loss_df = start_stop_loss(analysis_df)\n",
    "loss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1435cfdf-aa5b-442f-91fa-a2b30af4db9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotate transcript information (transcript start, end, sequence, length, exon info) in case of start or stop loss\n",
    "# transcript sequences are in: exon_seqs_subset\n",
    "\n",
    "transcript_starts = exon_seqs.set_index(\"transcript_id\")[\"start\"].to_dict()\n",
    "loss_df[\"transcript_start\"] = loss_df[\"transcript_id\"].map(transcript_starts)\n",
    "transcript_ends = exon_seqs.set_index(\"transcript_id\")[\"end\"].to_dict()\n",
    "loss_df[\"transcript_end\"] = loss_df[\"transcript_id\"].map(transcript_ends)\n",
    "transcript_sequences = exon_seqs.set_index(\"transcript_id\")[\"transcript_sequence\"].to_dict()  # create map of transcript-id to transcript sequence\n",
    "loss_df[\"transcript_seq\"] = loss_df[\"transcript_id\"].map(transcript_sequences)\n",
    "transcript_lengths = exon_seqs.set_index(\"transcript_id\")[\"transcript_length\"].to_dict()\n",
    "loss_df[\"transcript_length\"] = loss_df[\"transcript_id\"].map(transcript_lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8bdab4fa-09ce-4565-936e-754c96f7cfbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# In case of start or stop loss:\n",
    "# Splice alternative CDS into reference transcript sequence to create alternative transcript sequence and measure new length    \n",
    "\n",
    "def splice_alt_cds_into_transcript(row, transcript_seq):\n",
    "\n",
    "    # Step 1: search for ref_cds_seq match in the transcript, and replace that with the alt_cds_seq\n",
    "    ref_cds_seq = row[\"ref_cds_seq\"].upper()\n",
    "    alt_cds_seq = row[\"alt_cds_seq\"].upper()\n",
    "\n",
    "    # Find the ref CDS in the transcript sequence\n",
    "    ref_start_idx = transcript_seq.find(ref_cds_seq)\n",
    "\n",
    "    if ref_start_idx == -1:\n",
    "        return None  # Cannot find ref CDS, alignment problem\n",
    "\n",
    "    ref_end_idx = ref_start_idx + len(ref_cds_seq)\n",
    "\n",
    "    # Replace the reference CDS with the variant-modified / alternative one\n",
    "    new_transcript_seq = (\n",
    "        transcript_seq[:ref_start_idx] +\n",
    "        alt_cds_seq +\n",
    "        transcript_seq[ref_end_idx:]\n",
    "    )\n",
    "\n",
    "    return new_transcript_seq\n",
    "\n",
    "loss_df[\"alt_transcript_seq\"] = loss_df.apply(\n",
    "    lambda row: splice_alt_cds_into_transcript(row, row[\"transcript_seq\"])\n",
    "    if pd.notnull(row[\"transcript_seq\"]) else None,\n",
    "    axis=1\n",
    ")\n",
    "loss_df[\"alt_transcript_length\"] = loss_df[\"alt_transcript_seq\"].apply(\n",
    "    lambda x: len(x) if pd.notnull(x) else None\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "16c0d3a5-2ab4-4112-b6cc-b5aaeda9669f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "      <th>cds_in_transcript</th>\n",
       "      <th>ref_start_codon_pos</th>\n",
       "      <th>ref_start_codon_exon</th>\n",
       "      <th>ref_last_codon</th>\n",
       "      <th>ref_valid_stop</th>\n",
       "      <th>ref_first_stop_codon</th>\n",
       "      <th>ref_first_stop_pos</th>\n",
       "      <th>ref_num_stop_codons</th>\n",
       "      <th>ref_all_stop_codons</th>\n",
       "      <th>ref_stop_codon_exons</th>\n",
       "      <th>ref_is_premature</th>\n",
       "      <th>alt_start_codon_pos</th>\n",
       "      <th>alt_start_codon_exon</th>\n",
       "      <th>alt_last_codon</th>\n",
       "      <th>alt_valid_stop</th>\n",
       "      <th>alt_first_stop_codon</th>\n",
       "      <th>alt_first_stop_pos</th>\n",
       "      <th>alt_num_stop_codons</th>\n",
       "      <th>alt_all_stop_codons</th>\n",
       "      <th>alt_stop_codon_exons</th>\n",
       "      <th>alt_is_premature</th>\n",
       "      <th>start_loss</th>\n",
       "      <th>stop_loss</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>alt_transcript_seq</th>\n",
       "      <th>alt_transcript_length</th>\n",
       "      <th>transcript_exon_info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>351</td>\n",
       "      <td>2</td>\n",
       "      <td>[(351, TGA), (447, TGA)]</td>\n",
       "      <td>[4, 5]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1792.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>96</td>\n",
       "      <td>8</td>\n",
       "      <td>[(96, TGA), (102, TGA), (141, TAA), (144, TGA)...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 2, 2, 4]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1264.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3690.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GTG</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>42720608</td>\n",
       "      <td>42751741</td>\n",
       "      <td>AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...</td>\n",
       "      <td>580</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4042.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  cds_in_transcript  \\\n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]               True   \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]               True   \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]               True   \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]              False   \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "\n",
       "  ref_start_codon_pos ref_start_codon_exon ref_last_codon ref_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    1            GAT          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  ref_first_stop_codon ref_first_stop_pos ref_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TAA                600                   1   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                522                   1   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                774                   1   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                774                   1   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "  ref_all_stop_codons ref_stop_codon_exons ref_is_premature  \\\n",
       "0        [(600, TAA)]                  [5]            False   \n",
       "1        [(600, TAA)]                  [5]            False   \n",
       "2        [(600, TAA)]                  [5]            False   \n",
       "3        [(522, TGA)]                  [4]            False   \n",
       "4        [(522, TGA)]                  [4]            False   \n",
       "5        [(774, TGA)]                  [6]            False   \n",
       "6        [(774, TGA)]                  [6]            False   \n",
       "7                  []                   []            False   \n",
       "8        [(774, TGA)]                  [6]            False   \n",
       "9        [(774, TGA)]                  [6]            False   \n",
       "\n",
       "  alt_start_codon_pos alt_start_codon_exon alt_last_codon alt_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    2            GTG          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  alt_first_stop_codon alt_first_stop_pos alt_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TGA                351                   2   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                 96                   8   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                348                   8   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                348                   8   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "                                 alt_all_stop_codons  \\\n",
       "0                                       [(600, TAA)]   \n",
       "1                           [(351, TGA), (447, TGA)]   \n",
       "2                                       [(600, TAA)]   \n",
       "3  [(96, TGA), (102, TGA), (141, TAA), (144, TGA)...   \n",
       "4                                       [(522, TGA)]   \n",
       "5  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "6                                       [(774, TGA)]   \n",
       "7                                                 []   \n",
       "8  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "9                                       [(774, TGA)]   \n",
       "\n",
       "       alt_stop_codon_exons alt_is_premature  start_loss  stop_loss  \\\n",
       "0                       [5]            False       False      False   \n",
       "1                    [4, 5]             True       False      False   \n",
       "2                       [5]            False       False      False   \n",
       "3  [1, 1, 1, 1, 1, 2, 2, 4]             True       False      False   \n",
       "4                       [4]            False       False      False   \n",
       "5  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "6                       [6]            False       False      False   \n",
       "7                        []            False       False       True   \n",
       "8  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "9                       [6]            False       False      False   \n",
       "\n",
       "   transcript_start  transcript_end  \\\n",
       "0          42704779        42751728   \n",
       "1          42704779        42751728   \n",
       "2          42704779        42751728   \n",
       "3          42710849        42751849   \n",
       "4          42710849        42751849   \n",
       "5          42708443        42751748   \n",
       "6          42708443        42751748   \n",
       "7          42720608        42751741   \n",
       "8          42708439        42751866   \n",
       "9          42708439        42751866   \n",
       "\n",
       "                                      transcript_seq  transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "7  AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...                580   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "\n",
       "                                  alt_transcript_seq  alt_transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1792.0   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1264.0   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1338.0   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3690.0   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3764.0   \n",
       "7                                               None                    NaN   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4042.0   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4116.0   \n",
       "\n",
       "                                transcript_exon_info  \n",
       "0  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...  \n",
       "1  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...  \n",
       "2  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...  \n",
       "3  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...  \n",
       "4  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...  \n",
       "5  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...  \n",
       "6  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...  \n",
       "7  [(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...  \n",
       "8  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...  \n",
       "9  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add exon info to dataframe\n",
    "transcript_exon_info = exon_seqs.set_index(\"transcript_id\")[\"transcript_exon_info\"].to_dict()\n",
    "loss_df[\"transcript_exon_info\"] = loss_df[\"transcript_id\"].map(transcript_exon_info)\n",
    "\n",
    "loss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6ef1a469-45c9-485b-8a4d-832b41a3e034",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "      <th>cds_in_transcript</th>\n",
       "      <th>ref_start_codon_pos</th>\n",
       "      <th>ref_start_codon_exon</th>\n",
       "      <th>ref_last_codon</th>\n",
       "      <th>ref_valid_stop</th>\n",
       "      <th>ref_first_stop_codon</th>\n",
       "      <th>ref_first_stop_pos</th>\n",
       "      <th>ref_num_stop_codons</th>\n",
       "      <th>ref_all_stop_codons</th>\n",
       "      <th>ref_stop_codon_exons</th>\n",
       "      <th>ref_is_premature</th>\n",
       "      <th>alt_start_codon_pos</th>\n",
       "      <th>alt_start_codon_exon</th>\n",
       "      <th>alt_last_codon</th>\n",
       "      <th>alt_valid_stop</th>\n",
       "      <th>alt_first_stop_codon</th>\n",
       "      <th>alt_first_stop_pos</th>\n",
       "      <th>alt_num_stop_codons</th>\n",
       "      <th>alt_all_stop_codons</th>\n",
       "      <th>alt_stop_codon_exons</th>\n",
       "      <th>alt_is_premature</th>\n",
       "      <th>start_loss</th>\n",
       "      <th>stop_loss</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>alt_transcript_seq</th>\n",
       "      <th>alt_transcript_length</th>\n",
       "      <th>transcript_exon_info</th>\n",
       "      <th>transcript_start_codon_pos</th>\n",
       "      <th>transcript_start_codon_exon</th>\n",
       "      <th>transcript_last_codon</th>\n",
       "      <th>transcript_valid_stop</th>\n",
       "      <th>transcript_first_stop_codon</th>\n",
       "      <th>transcript_first_stop_pos</th>\n",
       "      <th>transcript_num_stop_codons</th>\n",
       "      <th>transcript_all_stop_codons</th>\n",
       "      <th>transcript_stop_codon_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>351</td>\n",
       "      <td>2</td>\n",
       "      <td>[(351, TGA), (447, TGA)]</td>\n",
       "      <td>[4, 5]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1792.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>96</td>\n",
       "      <td>8</td>\n",
       "      <td>[(96, TGA), (102, TGA), (141, TAA), (144, TGA)...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 2, 2, 4]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1264.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3690.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GTG</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>42720608</td>\n",
       "      <td>42751741</td>\n",
       "      <td>AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...</td>\n",
       "      <td>580</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4042.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  cds_in_transcript  \\\n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]               True   \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]               True   \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]               True   \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]              False   \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "\n",
       "  ref_start_codon_pos ref_start_codon_exon ref_last_codon ref_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    1            GAT          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  ref_first_stop_codon ref_first_stop_pos ref_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TAA                600                   1   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                522                   1   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                774                   1   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                774                   1   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "  ref_all_stop_codons ref_stop_codon_exons ref_is_premature  \\\n",
       "0        [(600, TAA)]                  [5]            False   \n",
       "1        [(600, TAA)]                  [5]            False   \n",
       "2        [(600, TAA)]                  [5]            False   \n",
       "3        [(522, TGA)]                  [4]            False   \n",
       "4        [(522, TGA)]                  [4]            False   \n",
       "5        [(774, TGA)]                  [6]            False   \n",
       "6        [(774, TGA)]                  [6]            False   \n",
       "7                  []                   []            False   \n",
       "8        [(774, TGA)]                  [6]            False   \n",
       "9        [(774, TGA)]                  [6]            False   \n",
       "\n",
       "  alt_start_codon_pos alt_start_codon_exon alt_last_codon alt_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    2            GTG          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  alt_first_stop_codon alt_first_stop_pos alt_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TGA                351                   2   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                 96                   8   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                348                   8   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                348                   8   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "                                 alt_all_stop_codons  \\\n",
       "0                                       [(600, TAA)]   \n",
       "1                           [(351, TGA), (447, TGA)]   \n",
       "2                                       [(600, TAA)]   \n",
       "3  [(96, TGA), (102, TGA), (141, TAA), (144, TGA)...   \n",
       "4                                       [(522, TGA)]   \n",
       "5  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "6                                       [(774, TGA)]   \n",
       "7                                                 []   \n",
       "8  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "9                                       [(774, TGA)]   \n",
       "\n",
       "       alt_stop_codon_exons alt_is_premature  start_loss  stop_loss  \\\n",
       "0                       [5]            False       False      False   \n",
       "1                    [4, 5]             True       False      False   \n",
       "2                       [5]            False       False      False   \n",
       "3  [1, 1, 1, 1, 1, 2, 2, 4]             True       False      False   \n",
       "4                       [4]            False       False      False   \n",
       "5  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "6                       [6]            False       False      False   \n",
       "7                        []            False       False       True   \n",
       "8  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "9                       [6]            False       False      False   \n",
       "\n",
       "   transcript_start  transcript_end  \\\n",
       "0          42704779        42751728   \n",
       "1          42704779        42751728   \n",
       "2          42704779        42751728   \n",
       "3          42710849        42751849   \n",
       "4          42710849        42751849   \n",
       "5          42708443        42751748   \n",
       "6          42708443        42751748   \n",
       "7          42720608        42751741   \n",
       "8          42708439        42751866   \n",
       "9          42708439        42751866   \n",
       "\n",
       "                                      transcript_seq  transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "7  AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...                580   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "\n",
       "                                  alt_transcript_seq  alt_transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1792.0   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1264.0   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1338.0   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3690.0   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3764.0   \n",
       "7                                               None                    NaN   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4042.0   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4116.0   \n",
       "\n",
       "                                transcript_exon_info  \\\n",
       "0  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "1  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "2  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "3  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "4  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "5  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "6  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "7  [(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...   \n",
       "8  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "9  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "\n",
       "  transcript_start_codon_pos transcript_start_codon_exon  \\\n",
       "0                       None                        None   \n",
       "1                       None                        None   \n",
       "2                       None                        None   \n",
       "3                       None                        None   \n",
       "4                       None                        None   \n",
       "5                       None                        None   \n",
       "6                       None                        None   \n",
       "7                       None                        None   \n",
       "8                       None                        None   \n",
       "9                       None                        None   \n",
       "\n",
       "  transcript_last_codon transcript_valid_stop transcript_first_stop_codon  \\\n",
       "0                  None                  None                        None   \n",
       "1                  None                  None                        None   \n",
       "2                  None                  None                        None   \n",
       "3                  None                  None                        None   \n",
       "4                  None                  None                        None   \n",
       "5                  None                  None                        None   \n",
       "6                  None                  None                        None   \n",
       "7                  None                  None                        None   \n",
       "8                  None                  None                        None   \n",
       "9                  None                  None                        None   \n",
       "\n",
       "  transcript_first_stop_pos transcript_num_stop_codons  \\\n",
       "0                      None                       None   \n",
       "1                      None                       None   \n",
       "2                      None                       None   \n",
       "3                      None                       None   \n",
       "4                      None                       None   \n",
       "5                      None                       None   \n",
       "6                      None                       None   \n",
       "7                      None                       None   \n",
       "8                      None                       None   \n",
       "9                      None                       None   \n",
       "\n",
       "  transcript_all_stop_codons transcript_stop_codon_exons  \n",
       "0                       None                        None  \n",
       "1                       None                        None  \n",
       "2                       None                        None  \n",
       "3                       None                        None  \n",
       "4                       None                        None  \n",
       "5                       None                        None  \n",
       "6                       None                        None  \n",
       "7                       None                        None  \n",
       "8                       None                        None  \n",
       "9                       None                        None  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def analyze_transcript(results_df):\n",
    "\n",
    "    \"\"\"\n",
    "    Analyze the alternative transcript sequence in cases of start or stop codons loss due to mutations.\n",
    "    Scan for new in-frame start or stop codons in the alternative transcript sequence.\n",
    "    :param results_df: DataFrame containing transcript sequence data and annotations, including start_loss and stop_loss flags\n",
    "    :return: pandas DataFrame with additional columns for rescued start / stop codon information\n",
    "    \"\"\"\n",
    "\n",
    "    valid_stop_codons = {\"TAA\", \"TAG\", \"TGA\"}\n",
    "    start_codon = \"ATG\"\n",
    "\n",
    "    df = results_df.copy()\n",
    "\n",
    "    # Add new columns to store results\n",
    "    df[\"transcript_start_codon_pos\"] = None\n",
    "    df[\"transcript_start_codon_exon\"] = None  # for exon number\n",
    "    df[\"transcript_last_codon\"] = None\n",
    "    df[\"transcript_valid_stop\"] = None\n",
    "    df[\"transcript_first_stop_codon\"] = None\n",
    "    df[\"transcript_first_stop_pos\"] = None\n",
    "    df[\"transcript_num_stop_codons\"] = None\n",
    "    df[\"transcript_all_stop_codons\"] = None\n",
    "    df[\"transcript_stop_codon_exons\"] = None  # for exon number\n",
    "\n",
    "    for idx, row in df.iterrows():\n",
    "        seq = row[\"alt_transcript_seq\"]\n",
    "        cds_start = row[\"alt_cds_start\"] - row[\"transcript_start\"]  # start analysis at cds start position\n",
    "        # cds_stop = cds_start + len(seq)\n",
    "\n",
    "        exon_info = row[\"transcript_exon_info\"]  # for exon number\n",
    "\n",
    "        # Skip rows with invalid or too-short sequences\n",
    "        if not isinstance(seq, str) or len(seq) < 3:\n",
    "            continue\n",
    "\n",
    "        start_pos = None\n",
    "        start_exon = None  # for exon number\n",
    "        stop_codons = []\n",
    "        stop_exons = []  # for exon number\n",
    "\n",
    "        # only analyze rows flagged with start or stop codon loss: skip the others and fill with None values\n",
    "        # Skips only if we have both start_loss = FALSE and stop_loss = FALSE. If one is true, then don't skip.\n",
    "        if not (row[\"start_loss\"] or row[\"stop_loss\"]):\n",
    "            continue\n",
    "\n",
    "        # START LOSS rescue search\n",
    "        if row[\"start_loss\"]:\n",
    "            # Walk through sequence starting at CDS start with +1 positions until start codon is found\n",
    "            for i in range(cds_start, len(seq) - 2):\n",
    "                codon = seq[i:i + 3]\n",
    "                if codon == start_codon:\n",
    "                    start_pos = i\n",
    "\n",
    "                    start_exon = get_exon(start_pos, exon_info)  # for exon number\n",
    "\n",
    "                    # From new start codon, scan codons in frame\n",
    "                    for j in range(i, len(seq) - 2, 3):\n",
    "                        codon2 = seq[j:j + 3]\n",
    "                        if codon2 in valid_stop_codons:\n",
    "                            stop_codons.append((j, codon2))\n",
    "                            stop_exons.append(get_exon(j, exon_info))  # for exon number\n",
    "\n",
    "                    break\n",
    "\n",
    "        # STOP LOSS rescue search\n",
    "        elif row[\"stop_loss\"]:\n",
    "            # Start at cds_stop, scan codons in frame to end --> lets start at cds start so we are in frame\n",
    "            for i in range(cds_start, len(seq) - 2, 3):\n",
    "                codon = seq[i:i + 3]\n",
    "                if codon == start_codon and start_pos is None:\n",
    "                    start_pos = i\n",
    "                    start_exon = get_exon(start_pos, exon_info)  # for exon number\n",
    "                if codon in valid_stop_codons:\n",
    "                    stop_codons.append((i, codon))\n",
    "                    stop_exons.append(get_exon(i, exon_info))  # for exon number\n",
    "\n",
    "        # Set last codon in transcript (use last full codon)\n",
    "        last_codon = seq[-3:] if len(seq) >= 3 else None\n",
    "        is_valid_stop = last_codon in valid_stop_codons\n",
    "        first_stop_pos = stop_codons[0][0] if stop_codons else None\n",
    "        first_stop = stop_codons[0][1] if stop_codons else None\n",
    "\n",
    "        # Store results\n",
    "        df.at[idx, \"transcript_start_codon_pos\"] = start_pos\n",
    "        df.at[idx, \"transcript_start_codon_exon\"] = start_exon  # for exon number\n",
    "        df.at[idx, \"transcript_last_codon\"] = last_codon\n",
    "        df.at[idx, \"transcript_valid_stop\"] = is_valid_stop\n",
    "        df.at[idx, \"transcript_first_stop_codon\"] = first_stop\n",
    "        df.at[idx, \"transcript_first_stop_pos\"] = first_stop_pos\n",
    "        df.at[idx, \"transcript_num_stop_codons\"] = len(stop_codons)\n",
    "        df.at[idx, \"transcript_all_stop_codons\"] = stop_codons\n",
    "        df.at[idx, \"transcript_stop_codon_exons\"] = stop_exons  # for exon number\n",
    "\n",
    "    return df\n",
    "    \n",
    "analyze_transcript_df = analyze_transcript(loss_df)\n",
    "analyze_transcript_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e8fb75d2-0f5a-41db-9800-0a4857e3b29c",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = analyze_transcript_df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3573a588-f4d4-4ed5-a144-a7f3183638cb",
   "metadata": {},
   "source": [
    "# Add additional NMD features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "909491ba-7fb0-4c51-a223-907480896302",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "      <th>cds_in_transcript</th>\n",
       "      <th>ref_start_codon_pos</th>\n",
       "      <th>ref_start_codon_exon</th>\n",
       "      <th>ref_last_codon</th>\n",
       "      <th>ref_valid_stop</th>\n",
       "      <th>ref_first_stop_codon</th>\n",
       "      <th>ref_first_stop_pos</th>\n",
       "      <th>ref_num_stop_codons</th>\n",
       "      <th>ref_all_stop_codons</th>\n",
       "      <th>ref_stop_codon_exons</th>\n",
       "      <th>ref_is_premature</th>\n",
       "      <th>alt_start_codon_pos</th>\n",
       "      <th>alt_start_codon_exon</th>\n",
       "      <th>alt_last_codon</th>\n",
       "      <th>alt_valid_stop</th>\n",
       "      <th>alt_first_stop_codon</th>\n",
       "      <th>alt_first_stop_pos</th>\n",
       "      <th>alt_num_stop_codons</th>\n",
       "      <th>alt_all_stop_codons</th>\n",
       "      <th>alt_stop_codon_exons</th>\n",
       "      <th>alt_is_premature</th>\n",
       "      <th>start_loss</th>\n",
       "      <th>stop_loss</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>alt_transcript_seq</th>\n",
       "      <th>alt_transcript_length</th>\n",
       "      <th>transcript_exon_info</th>\n",
       "      <th>transcript_start_codon_pos</th>\n",
       "      <th>transcript_start_codon_exon</th>\n",
       "      <th>transcript_last_codon</th>\n",
       "      <th>transcript_valid_stop</th>\n",
       "      <th>transcript_first_stop_codon</th>\n",
       "      <th>transcript_first_stop_pos</th>\n",
       "      <th>transcript_num_stop_codons</th>\n",
       "      <th>transcript_all_stop_codons</th>\n",
       "      <th>transcript_stop_codon_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>351</td>\n",
       "      <td>2</td>\n",
       "      <td>[(351, TGA), (447, TGA)]</td>\n",
       "      <td>[4, 5]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1792.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>96</td>\n",
       "      <td>8</td>\n",
       "      <td>[(96, TGA), (102, TGA), (141, TAA), (144, TGA)...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 2, 2, 4]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1264.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3690.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GTG</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>42720608</td>\n",
       "      <td>42751741</td>\n",
       "      <td>AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...</td>\n",
       "      <td>580</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4042.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  cds_in_transcript  \\\n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]               True   \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]               True   \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]               True   \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]              False   \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "\n",
       "  ref_start_codon_pos ref_start_codon_exon ref_last_codon ref_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    1            GAT          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  ref_first_stop_codon ref_first_stop_pos ref_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TAA                600                   1   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                522                   1   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                774                   1   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                774                   1   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "  ref_all_stop_codons ref_stop_codon_exons ref_is_premature  \\\n",
       "0        [(600, TAA)]                  [5]            False   \n",
       "1        [(600, TAA)]                  [5]            False   \n",
       "2        [(600, TAA)]                  [5]            False   \n",
       "3        [(522, TGA)]                  [4]            False   \n",
       "4        [(522, TGA)]                  [4]            False   \n",
       "5        [(774, TGA)]                  [6]            False   \n",
       "6        [(774, TGA)]                  [6]            False   \n",
       "7                  []                   []            False   \n",
       "8        [(774, TGA)]                  [6]            False   \n",
       "9        [(774, TGA)]                  [6]            False   \n",
       "\n",
       "  alt_start_codon_pos alt_start_codon_exon alt_last_codon alt_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    2            GTG          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  alt_first_stop_codon alt_first_stop_pos alt_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TGA                351                   2   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                 96                   8   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                348                   8   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                348                   8   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "                                 alt_all_stop_codons  \\\n",
       "0                                       [(600, TAA)]   \n",
       "1                           [(351, TGA), (447, TGA)]   \n",
       "2                                       [(600, TAA)]   \n",
       "3  [(96, TGA), (102, TGA), (141, TAA), (144, TGA)...   \n",
       "4                                       [(522, TGA)]   \n",
       "5  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "6                                       [(774, TGA)]   \n",
       "7                                                 []   \n",
       "8  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "9                                       [(774, TGA)]   \n",
       "\n",
       "       alt_stop_codon_exons alt_is_premature  start_loss  stop_loss  \\\n",
       "0                       [5]            False       False      False   \n",
       "1                    [4, 5]             True       False      False   \n",
       "2                       [5]            False       False      False   \n",
       "3  [1, 1, 1, 1, 1, 2, 2, 4]             True       False      False   \n",
       "4                       [4]            False       False      False   \n",
       "5  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "6                       [6]            False       False      False   \n",
       "7                        []            False       False       True   \n",
       "8  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "9                       [6]            False       False      False   \n",
       "\n",
       "   transcript_start  transcript_end  \\\n",
       "0          42704779        42751728   \n",
       "1          42704779        42751728   \n",
       "2          42704779        42751728   \n",
       "3          42710849        42751849   \n",
       "4          42710849        42751849   \n",
       "5          42708443        42751748   \n",
       "6          42708443        42751748   \n",
       "7          42720608        42751741   \n",
       "8          42708439        42751866   \n",
       "9          42708439        42751866   \n",
       "\n",
       "                                      transcript_seq  transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "7  AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...                580   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "\n",
       "                                  alt_transcript_seq  alt_transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1792.0   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1264.0   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1338.0   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3690.0   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3764.0   \n",
       "7                                               None                    NaN   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4042.0   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4116.0   \n",
       "\n",
       "                                transcript_exon_info  \\\n",
       "0  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "1  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "2  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "3  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "4  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "5  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "6  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "7  [(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...   \n",
       "8  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "9  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "\n",
       "  transcript_start_codon_pos transcript_start_codon_exon  \\\n",
       "0                       None                        None   \n",
       "1                       None                        None   \n",
       "2                       None                        None   \n",
       "3                       None                        None   \n",
       "4                       None                        None   \n",
       "5                       None                        None   \n",
       "6                       None                        None   \n",
       "7                       None                        None   \n",
       "8                       None                        None   \n",
       "9                       None                        None   \n",
       "\n",
       "  transcript_last_codon transcript_valid_stop transcript_first_stop_codon  \\\n",
       "0                  None                  None                        None   \n",
       "1                  None                  None                        None   \n",
       "2                  None                  None                        None   \n",
       "3                  None                  None                        None   \n",
       "4                  None                  None                        None   \n",
       "5                  None                  None                        None   \n",
       "6                  None                  None                        None   \n",
       "7                  None                  None                        None   \n",
       "8                  None                  None                        None   \n",
       "9                  None                  None                        None   \n",
       "\n",
       "  transcript_first_stop_pos transcript_num_stop_codons  \\\n",
       "0                      None                       None   \n",
       "1                      None                       None   \n",
       "2                      None                       None   \n",
       "3                      None                       None   \n",
       "4                      None                       None   \n",
       "5                      None                       None   \n",
       "6                      None                       None   \n",
       "7                      None                       None   \n",
       "8                      None                       None   \n",
       "9                      None                       None   \n",
       "\n",
       "  transcript_all_stop_codons transcript_stop_codon_exons  \n",
       "0                       None                        None  \n",
       "1                       None                        None  \n",
       "2                       None                        None  \n",
       "3                       None                        None  \n",
       "4                       None                        None  \n",
       "5                       None                        None  \n",
       "6                       None                        None  \n",
       "7                       None                        None  \n",
       "8                       None                        None  \n",
       "9                       None                        None  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results # --> get row for calculation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d24634d1-06c6-4659-a24c-49de3d23e0d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "transcript_id                                                ENST00000319104.7_2\n",
       "variant_id                                                                     .\n",
       "ref_cds_start                                                           42705933\n",
       "ref_cds_stop                                                            42743007\n",
       "ref_cds_seq                    ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...\n",
       "ref_cds_len                                                                  603\n",
       "alt_cds_start                                                           42705933\n",
       "alt_cds_stop                                                            42743007\n",
       "alt_cds_seq                    ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...\n",
       "alt_cds_len                                                                  603\n",
       "chromosome                                                                  chr8\n",
       "gene_id                                                     ENSG00000120925.16_6\n",
       "strand                                                                         -\n",
       "ref                                                                            C\n",
       "alt                                                                            C\n",
       "start_variant                                                           42706139\n",
       "end_variant                                                             42706140\n",
       "ref_cds_info                    [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]\n",
       "alt_cds_info                    [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]\n",
       "cds_in_transcript                                                           True\n",
       "ref_start_codon_pos                                                            0\n",
       "ref_start_codon_exon                                                           1\n",
       "ref_last_codon                                                               TAA\n",
       "ref_valid_stop                                                              True\n",
       "ref_first_stop_codon                                                         TAA\n",
       "ref_first_stop_pos                                                           600\n",
       "ref_num_stop_codons                                                            1\n",
       "ref_all_stop_codons                                                 [(600, TAA)]\n",
       "ref_stop_codon_exons                                                         [5]\n",
       "ref_is_premature                                                           False\n",
       "alt_start_codon_pos                                                            0\n",
       "alt_start_codon_exon                                                           1\n",
       "alt_last_codon                                                               TAA\n",
       "alt_valid_stop                                                              True\n",
       "alt_first_stop_codon                                                         TAA\n",
       "alt_first_stop_pos                                                           600\n",
       "alt_num_stop_codons                                                            1\n",
       "alt_all_stop_codons                                                 [(600, TAA)]\n",
       "alt_stop_codon_exons                                                         [5]\n",
       "alt_is_premature                                                           False\n",
       "start_loss                                                                 False\n",
       "stop_loss                                                                  False\n",
       "transcript_start                                                        42704779\n",
       "transcript_end                                                          42751728\n",
       "transcript_seq                 GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...\n",
       "transcript_length                                                           1866\n",
       "alt_transcript_seq             GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...\n",
       "alt_transcript_length                                                     1866.0\n",
       "transcript_exon_info           [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...\n",
       "transcript_start_codon_pos                                                  None\n",
       "transcript_start_codon_exon                                                 None\n",
       "transcript_last_codon                                                       None\n",
       "transcript_valid_stop                                                       None\n",
       "transcript_first_stop_codon                                                 None\n",
       "transcript_first_stop_pos                                                   None\n",
       "transcript_num_stop_codons                                                  None\n",
       "transcript_all_stop_codons                                                  None\n",
       "transcript_stop_codon_exons                                                 None\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row = results.iloc[0]\n",
    "row"
   ]
  },
  {
   "cell_type": "raw",
   "id": "1db4acd1-ed75-4907-b944-7a824635ecf8",
   "metadata": {},
   "source": [
    "\"\"\"\n",
    "Compute additional features which might be relevant for analyzing nonsense-mediated decay (NMD) behavior,\n",
    "inspired by benchmark datasets from nmd_eff. These features include UTR lengths, exon structure and positional information\n",
    "of the premature termination codon (PTC).\n",
    "\n",
    ":param row: A DataFrame row with annotated transcript information\n",
    ":return: A dictionary with additional NMD related features.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "365931b3-23a8-482f-9e10-a3da716deb50",
   "metadata": {},
   "source": [
    "## needed functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c5d86f80-f88b-4122-a332-ff728c97df10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_nmd_features(row):\n",
    "\n",
    "    # 5' and 3' UTR lengths\n",
    "    utr_lengths = calculate_utr_lengths(row)\n",
    "    utr3_length = utr_lengths[\"utr3_length\"]\n",
    "    utr5_length = utr_lengths[\"utr5_length\"]\n",
    "\n",
    "    # Total, Upstream and Downstream exon count\n",
    "    exon_features = calculate_exon_features(row)\n",
    "    total_exon_count = exon_features[\"total_exon_count\"]\n",
    "    upstream_exon_count = exon_features[\"upstream_exon_count\"]\n",
    "    downstream_exon_count = exon_features[\"downstream_exon_count\"]\n",
    "\n",
    "    # Distance between PTC to start codon\n",
    "    ptc_to_start_codon = calculate_ptc_to_start_distance(row)\n",
    "    # PTC location < 150nt to start codon\n",
    "    ptc_less_than_150nt_to_start = (\n",
    "            ptc_to_start_codon is not None and ptc_to_start_codon < 150\n",
    "    )\n",
    "\n",
    "    # PTC exon length\n",
    "    ptc_exon_length = calculate_ptc_exon_length(row)\n",
    "\n",
    "    # Distance PTC to normal stop codon\n",
    "    stop_codon_distance = calculate_stop_codon_dist(row)\n",
    "\n",
    "    # Distance PTC to downstream exon junction\n",
    "    ptc_to_intron = calculate_ptc_to_downstream_ej(row)\n",
    "\n",
    "    # Add likely_misannotated flag\n",
    "    likely_misannotated = add_likely_misannotated_flag(row)\n",
    "\n",
    "\n",
    "    return {\n",
    "        \"utr3_length\": utr3_length,\n",
    "        \"utr5_length\": utr5_length,\n",
    "        \"total_exon_count\": total_exon_count,\n",
    "        \"upstream_exon_count\": upstream_exon_count,\n",
    "        \"downstream_exon_count\": downstream_exon_count,\n",
    "        # \"ptc_pos_codon\": ptc_pos_codon,\n",
    "        \"ptc_to_start_codon\": ptc_to_start_codon,\n",
    "        \"ptc_less_than_150nt_to_start\": ptc_less_than_150nt_to_start,\n",
    "        \"ptc_exon_length\": ptc_exon_length,\n",
    "        \"stop_codon_distance\": stop_codon_distance,\n",
    "        \"ptc_to_intron\": ptc_to_intron,\n",
    "        \"likely_misannotated\": likely_misannotated\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1e953965-5677-43ec-b173-f8f73712aa3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_utr_lengths(row):\n",
    "\n",
    "    strand = row.get(\"strand\")\n",
    "    ref_cds_info = row.get(\"ref_cds_info\") or []\n",
    "    transcript_exon_info = row.get(\"transcript_exon_info\") or []\n",
    "\n",
    "    if not ref_cds_info or not transcript_exon_info:\n",
    "        return {\"utr5_length\": None, \"utr3_length\": None}\n",
    "\n",
    "    # Convert to dicts for easier lookup\n",
    "    transcript_exon_dict = {int(k): int(v) for k, v in transcript_exon_info}\n",
    "    cds_exons_dict = {int(exon): int(length) for exon, length in ref_cds_info}\n",
    "\n",
    "    # Handle single exon\n",
    "    if len(transcript_exon_dict) == 1:\n",
    "        if strand == \"+\":\n",
    "            utr5 = row[\"ref_cds_start\"] - row[\"transcript_start\"]\n",
    "            utr3 = row[\"transcript_end\"] - row[\"ref_cds_stop\"]\n",
    "        else:\n",
    "            utr5 = row[\"transcript_end\"] - row[\"ref_cds_stop\"]\n",
    "            utr3 = row[\"ref_cds_start\"] - row[\"transcript_start\"]\n",
    "\n",
    "        utr5 = utr5 if utr5 >= 0 else None\n",
    "        utr3 = utr3 if utr3 >= 0 else None\n",
    "\n",
    "        return {\"utr5_length\": utr5, \"utr3_length\": utr3}\n",
    "\n",
    "    cds_exon_nums = sorted(cds_exons_dict.keys())\n",
    "    tx_exon_nums = sorted(transcript_exon_dict.keys())\n",
    "\n",
    "    utr5 = 0\n",
    "    utr3 = 0\n",
    "\n",
    "    for exon in tx_exon_nums:\n",
    "        exon_len = transcript_exon_dict[exon]\n",
    "        cds_len = cds_exons_dict.get(exon, 0)\n",
    "        utr_len = exon_len - cds_len\n",
    "\n",
    "        if utr_len <= 0:\n",
    "            continue\n",
    "\n",
    "        if exon in cds_exons_dict:\n",
    "            # Exon overlaps CDS, partial UTR\n",
    "            if strand == \"+\":\n",
    "                if exon == cds_exon_nums[0]:\n",
    "                    utr5 += utr_len\n",
    "                elif exon == cds_exon_nums[-1]:\n",
    "                    utr3 += utr_len\n",
    "            else:\n",
    "                if exon == cds_exon_nums[0]:\n",
    "                    utr3 += utr_len\n",
    "                elif exon == cds_exon_nums[-1]:\n",
    "                    utr5 += utr_len\n",
    "        else:\n",
    "            # Exon is outside CDS\n",
    "            if strand == \"+\":\n",
    "                if exon < cds_exon_nums[0]:\n",
    "                    utr5 += exon_len\n",
    "                elif exon > cds_exon_nums[-1]:\n",
    "                    utr3 += exon_len\n",
    "            else:\n",
    "                if exon > cds_exon_nums[-1]:\n",
    "                    utr5 += exon_len\n",
    "                elif exon < cds_exon_nums[0]:\n",
    "                    utr3 += exon_len\n",
    "\n",
    "    utr5 = utr5 if utr5 >= 0 else None\n",
    "    utr3 = utr3 if utr3 >= 0 else None\n",
    "    return {\"utr5_length\": utr5, \"utr3_length\": utr3}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "26992bd9-3b79-4670-b5ec-9dcf7ada57d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_exon_features(row):\n",
    "\n",
    "    \"\"\"\n",
    "    Calculate exon-related features:\n",
    "    - total_exon_count: always computed if transcript_exon_info is available\n",
    "    - upstream_exon_count / downstream_exon_count: only computed if a PTC exists\n",
    "    \"\"\"\n",
    "\n",
    "    exon_info = row.get(\"transcript_exon_info\") or []\n",
    "    stop_exons = row.get(\"alt_stop_codon_exons\") or []\n",
    "\n",
    "    total_exons = len(exon_info)\n",
    "\n",
    "    if not row.get(\"alt_is_premature\") or not stop_exons or not exon_info:\n",
    "        return {\n",
    "            \"total_exon_count\": total_exons if total_exons > 0 else None,\n",
    "            \"upstream_exon_count\": None,\n",
    "            \"downstream_exon_count\": None\n",
    "        }\n",
    "\n",
    "    # Take the PTC exon closest to CDS start\n",
    "    ptc_exon = min(int(e) for e in stop_exons)\n",
    "\n",
    "    # get exon numbers from transcript_exon_info\n",
    "    exon_numbers = [int(e[0]) for e in exon_info]\n",
    "\n",
    "    # If the PTC exon is not in transcript  cannot compute\n",
    "    if ptc_exon not in exon_numbers:\n",
    "        return {\n",
    "            \"total_exon_count\": total_exons,\n",
    "            \"upstream_exon_count\": None,\n",
    "            \"downstream_exon_count\": None\n",
    "        }\n",
    "\n",
    "    upstream = sum(1 for e in exon_numbers if e < ptc_exon)\n",
    "    downstream = sum(1 for e in exon_numbers if e > ptc_exon)\n",
    "\n",
    "    return {\n",
    "        \"total_exon_count\": int(total_exons),\n",
    "        \"upstream_exon_count\": int(upstream),\n",
    "        \"downstream_exon_count\": int(downstream)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "df887f46-d04b-4940-aff4-8a2b166f024b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_ptc_to_start_distance(row):\n",
    "\n",
    "    if not row.get(\"alt_is_premature\"):\n",
    "        return None\n",
    "\n",
    "    start = row.get(\"alt_start_codon_pos\")\n",
    "    stop = row.get(\"alt_first_stop_pos\")\n",
    "\n",
    "    if start is None or stop is None:\n",
    "        return None\n",
    "\n",
    "    # PTC codon position: is PTC_to_start_codon / 3 --> leave it out\n",
    "    #offset = stop - start\n",
    "    #return offset // 3 if offset >= 0 else None\n",
    "\n",
    "    if stop <= start:\n",
    "        return None\n",
    "\n",
    "    return stop-start # distance between the PTC to start codon in nt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "7457f469-21d5-41a2-a36a-afd9b3cc156c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_ptc_exon_length(row):\n",
    "    \"\"\"\n",
    "    Return the length of the exon containing the first premature stop codon (PTC).\n",
    "    \"\"\"\n",
    "\n",
    "    if not row.get(\"alt_is_premature\"):\n",
    "        return None\n",
    "\n",
    "    stop_exons = row.get(\"alt_stop_codon_exons\") or []\n",
    "    exon_info = row.get(\"transcript_exon_info\") or []\n",
    "\n",
    "    if not stop_exons or not exon_info:\n",
    "        return None\n",
    "\n",
    "    # First PTC exon = smallest exon number (transcript-order, strand-corrected)\n",
    "    ptc_exon = min(int(e) for e in stop_exons)\n",
    "\n",
    "    exon_dict = {int(e): int(length) for e, length in exon_info}\n",
    "    return exon_dict.get(ptc_exon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "1034a6fb-2e7e-4440-9629-5e1d661a2345",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_stop_codon_dist(row):\n",
    "\n",
    "    \"\"\"\n",
    "    Calculate the distance between the reference stop codon and the alternative stop codon.\n",
    "    Positive means the PTC is upstream of the reference stop codon.\n",
    "    \"\"\"\n",
    "\n",
    "    ref_stop = row.get(\"ref_first_stop_pos\")\n",
    "    alt_stop = row.get(\"alt_first_stop_pos\")\n",
    "\n",
    "    if ref_stop is None or alt_stop is None:\n",
    "        return None\n",
    "\n",
    "    return ref_stop - alt_stop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c35dbb26-3317-4ab8-bc27-0043a4c48957",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_ptc_to_downstream_ej(row):\n",
    "\n",
    "    \"\"\"\n",
    "    Calculate distance from PTC to the downstream exon junction (next exon start/end depending on strand).\n",
    "    Returns None if not applicable.\n",
    "    \"\"\"\n",
    "\n",
    "    # only calculate if we have PTC\n",
    "    if not row.get(\"alt_is_premature\"):\n",
    "        return None\n",
    "\n",
    "    stop_exons = row.get(\"alt_stop_codon_exons\") or []\n",
    "    \n",
    "    # exon_info = row.get(\"transcript_exon_info\") or []\n",
    "    exon_info = row.get(\"alt_cds_info\") or [] # Assuming that the PTC cannot be outside the CDS, since it needs to come before the original stop codon\n",
    "    \n",
    "    ptc_pos = row.get(\"alt_first_stop_pos\")\n",
    "\n",
    "    if not stop_exons or not exon_info or ptc_pos is None:\n",
    "        return None\n",
    "\n",
    "    # Choose the PTC exon (smallest number, closer to start)\n",
    "    ptc_exon = min(stop_exons)\n",
    "\n",
    "    # Sum lengths of exons up to and including ptc_exon\n",
    "    cumulative_length = 0\n",
    "    for exon_num, length in exon_info:\n",
    "        cumulative_length += length\n",
    "        if exon_num == ptc_exon:\n",
    "            break\n",
    "\n",
    "    # Distance from PTC to downstream exon junction\n",
    "    distance = cumulative_length - ptc_pos\n",
    "    return distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "e7c7d568-d139-447f-bf6c-8b1cbb100aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_likely_misannotated_flag(row):\n",
    "\n",
    "    \"\"\"\n",
    "    Flag rows that look inconsistent between CDS and transcript annotations and might be likely misannotated.\n",
    "    A row is flagged as likely misannotated if any of these conditions apply:\n",
    "        cds_in_transcript = False (the assembled CDS is not found in the transcript sequence)\n",
    "        ref_start_codon_pos is defined and not 0 (reference CDS has a start codon not at the very start)\n",
    "        ref_valid_stop is False (the last reference codon is not a valid stop codon)\n",
    "\n",
    "    :return: A boolean flag. True if any condition above is met and thus the row is likely misannotated, False otherwise.\n",
    "    \"\"\"\n",
    "\n",
    "    # Add likely_misannotated flag: when\n",
    "    # \"cds_in_transcript\" is FALSE\n",
    "    # \"ref_start_codon_pos\" is not 0\n",
    "    # \"ref_valid_stop\" is FALSE\n",
    "\n",
    "    cds_in_transcript = row.get(\"cds_in_transcript\")\n",
    "    ref_start_codon_pos = row.get(\"ref_start_codon_pos\")\n",
    "    ref_valid_stop = row.get(\"ref_valid_stop\")\n",
    "\n",
    "    # if any of these are missing entirely, flag as likely misannotated\n",
    "    if cds_in_transcript is None or ref_start_codon_pos is None or ref_valid_stop is None:\n",
    "        return True\n",
    "\n",
    "    flag = (\n",
    "        (cds_in_transcript is False) or\n",
    "        ((ref_start_codon_pos is not None) and (ref_start_codon_pos != 0)) or\n",
    "        (ref_valid_stop is False)\n",
    "    )\n",
    "\n",
    "    return(flag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6cd28ed3-3e65-4433-b7f3-50ca7bf965f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "extra_features = results.apply(add_nmd_features, axis=1, result_type='expand')\n",
    "results = pd.concat([results, extra_features], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "39109d06-9b37-44ab-b9a0-54f550b1e5b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "      <th>cds_in_transcript</th>\n",
       "      <th>ref_start_codon_pos</th>\n",
       "      <th>ref_start_codon_exon</th>\n",
       "      <th>ref_last_codon</th>\n",
       "      <th>ref_valid_stop</th>\n",
       "      <th>ref_first_stop_codon</th>\n",
       "      <th>ref_first_stop_pos</th>\n",
       "      <th>ref_num_stop_codons</th>\n",
       "      <th>ref_all_stop_codons</th>\n",
       "      <th>ref_stop_codon_exons</th>\n",
       "      <th>ref_is_premature</th>\n",
       "      <th>alt_start_codon_pos</th>\n",
       "      <th>alt_start_codon_exon</th>\n",
       "      <th>alt_last_codon</th>\n",
       "      <th>alt_valid_stop</th>\n",
       "      <th>alt_first_stop_codon</th>\n",
       "      <th>alt_first_stop_pos</th>\n",
       "      <th>alt_num_stop_codons</th>\n",
       "      <th>alt_all_stop_codons</th>\n",
       "      <th>alt_stop_codon_exons</th>\n",
       "      <th>alt_is_premature</th>\n",
       "      <th>start_loss</th>\n",
       "      <th>stop_loss</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>alt_transcript_seq</th>\n",
       "      <th>alt_transcript_length</th>\n",
       "      <th>transcript_exon_info</th>\n",
       "      <th>transcript_start_codon_pos</th>\n",
       "      <th>transcript_start_codon_exon</th>\n",
       "      <th>transcript_last_codon</th>\n",
       "      <th>transcript_valid_stop</th>\n",
       "      <th>transcript_first_stop_codon</th>\n",
       "      <th>transcript_first_stop_pos</th>\n",
       "      <th>transcript_num_stop_codons</th>\n",
       "      <th>transcript_all_stop_codons</th>\n",
       "      <th>transcript_stop_codon_exons</th>\n",
       "      <th>utr3_length</th>\n",
       "      <th>utr5_length</th>\n",
       "      <th>total_exon_count</th>\n",
       "      <th>upstream_exon_count</th>\n",
       "      <th>downstream_exon_count</th>\n",
       "      <th>ptc_to_start_codon</th>\n",
       "      <th>ptc_less_than_150nt_to_start</th>\n",
       "      <th>ptc_exon_length</th>\n",
       "      <th>stop_codon_distance</th>\n",
       "      <th>ptc_to_intron</th>\n",
       "      <th>likely_misannotated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1154</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>351</td>\n",
       "      <td>2</td>\n",
       "      <td>[(351, TGA), (447, TGA)]</td>\n",
       "      <td>[4, 5]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1792.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1154</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>351.0</td>\n",
       "      <td>False</td>\n",
       "      <td>76.0</td>\n",
       "      <td>249.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1154</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>96</td>\n",
       "      <td>8</td>\n",
       "      <td>[(96, TGA), (102, TGA), (141, TAA), (144, TGA)...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 2, 2, 4]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1264.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>452</td>\n",
       "      <td>361</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>True</td>\n",
       "      <td>722.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>452</td>\n",
       "      <td>361</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3690.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2858</td>\n",
       "      <td>129</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>False</td>\n",
       "      <td>111.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2858</td>\n",
       "      <td>129</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GTG</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>42720608</td>\n",
       "      <td>42751741</td>\n",
       "      <td>AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...</td>\n",
       "      <td>580</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>234</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4042.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2862</td>\n",
       "      <td>477</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>False</td>\n",
       "      <td>111.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2862</td>\n",
       "      <td>477</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  cds_in_transcript  \\\n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]               True   \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]               True   \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]               True   \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]              False   \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "\n",
       "  ref_start_codon_pos ref_start_codon_exon ref_last_codon ref_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    1            GAT          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  ref_first_stop_codon ref_first_stop_pos ref_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TAA                600                   1   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                522                   1   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                774                   1   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                774                   1   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "  ref_all_stop_codons ref_stop_codon_exons ref_is_premature  \\\n",
       "0        [(600, TAA)]                  [5]            False   \n",
       "1        [(600, TAA)]                  [5]            False   \n",
       "2        [(600, TAA)]                  [5]            False   \n",
       "3        [(522, TGA)]                  [4]            False   \n",
       "4        [(522, TGA)]                  [4]            False   \n",
       "5        [(774, TGA)]                  [6]            False   \n",
       "6        [(774, TGA)]                  [6]            False   \n",
       "7                  []                   []            False   \n",
       "8        [(774, TGA)]                  [6]            False   \n",
       "9        [(774, TGA)]                  [6]            False   \n",
       "\n",
       "  alt_start_codon_pos alt_start_codon_exon alt_last_codon alt_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    2            GTG          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  alt_first_stop_codon alt_first_stop_pos alt_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TGA                351                   2   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                 96                   8   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                348                   8   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                348                   8   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "                                 alt_all_stop_codons  \\\n",
       "0                                       [(600, TAA)]   \n",
       "1                           [(351, TGA), (447, TGA)]   \n",
       "2                                       [(600, TAA)]   \n",
       "3  [(96, TGA), (102, TGA), (141, TAA), (144, TGA)...   \n",
       "4                                       [(522, TGA)]   \n",
       "5  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "6                                       [(774, TGA)]   \n",
       "7                                                 []   \n",
       "8  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "9                                       [(774, TGA)]   \n",
       "\n",
       "       alt_stop_codon_exons alt_is_premature  start_loss  stop_loss  \\\n",
       "0                       [5]            False       False      False   \n",
       "1                    [4, 5]             True       False      False   \n",
       "2                       [5]            False       False      False   \n",
       "3  [1, 1, 1, 1, 1, 2, 2, 4]             True       False      False   \n",
       "4                       [4]            False       False      False   \n",
       "5  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "6                       [6]            False       False      False   \n",
       "7                        []            False       False       True   \n",
       "8  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "9                       [6]            False       False      False   \n",
       "\n",
       "   transcript_start  transcript_end  \\\n",
       "0          42704779        42751728   \n",
       "1          42704779        42751728   \n",
       "2          42704779        42751728   \n",
       "3          42710849        42751849   \n",
       "4          42710849        42751849   \n",
       "5          42708443        42751748   \n",
       "6          42708443        42751748   \n",
       "7          42720608        42751741   \n",
       "8          42708439        42751866   \n",
       "9          42708439        42751866   \n",
       "\n",
       "                                      transcript_seq  transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "7  AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...                580   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "\n",
       "                                  alt_transcript_seq  alt_transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1792.0   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1264.0   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1338.0   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3690.0   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3764.0   \n",
       "7                                               None                    NaN   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4042.0   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4116.0   \n",
       "\n",
       "                                transcript_exon_info  \\\n",
       "0  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "1  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "2  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "3  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "4  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "5  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "6  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "7  [(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...   \n",
       "8  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "9  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "\n",
       "  transcript_start_codon_pos transcript_start_codon_exon  \\\n",
       "0                       None                        None   \n",
       "1                       None                        None   \n",
       "2                       None                        None   \n",
       "3                       None                        None   \n",
       "4                       None                        None   \n",
       "5                       None                        None   \n",
       "6                       None                        None   \n",
       "7                       None                        None   \n",
       "8                       None                        None   \n",
       "9                       None                        None   \n",
       "\n",
       "  transcript_last_codon transcript_valid_stop transcript_first_stop_codon  \\\n",
       "0                  None                  None                        None   \n",
       "1                  None                  None                        None   \n",
       "2                  None                  None                        None   \n",
       "3                  None                  None                        None   \n",
       "4                  None                  None                        None   \n",
       "5                  None                  None                        None   \n",
       "6                  None                  None                        None   \n",
       "7                  None                  None                        None   \n",
       "8                  None                  None                        None   \n",
       "9                  None                  None                        None   \n",
       "\n",
       "  transcript_first_stop_pos transcript_num_stop_codons  \\\n",
       "0                      None                       None   \n",
       "1                      None                       None   \n",
       "2                      None                       None   \n",
       "3                      None                       None   \n",
       "4                      None                       None   \n",
       "5                      None                       None   \n",
       "6                      None                       None   \n",
       "7                      None                       None   \n",
       "8                      None                       None   \n",
       "9                      None                       None   \n",
       "\n",
       "  transcript_all_stop_codons transcript_stop_codon_exons  utr3_length  \\\n",
       "0                       None                        None         1154   \n",
       "1                       None                        None         1154   \n",
       "2                       None                        None         1154   \n",
       "3                       None                        None          452   \n",
       "4                       None                        None          452   \n",
       "5                       None                        None         2858   \n",
       "6                       None                        None         2858   \n",
       "7                       None                        None            0   \n",
       "8                       None                        None         2862   \n",
       "9                       None                        None         2862   \n",
       "\n",
       "   utr5_length  total_exon_count  upstream_exon_count  downstream_exon_count  \\\n",
       "0          109                 6                  NaN                    NaN   \n",
       "1          109                 6                  3.0                    2.0   \n",
       "2          109                 6                  NaN                    NaN   \n",
       "3          361                 7                  0.0                    6.0   \n",
       "4          361                 7                  NaN                    NaN   \n",
       "5          129                 7                  1.0                    5.0   \n",
       "6          129                 7                  NaN                    NaN   \n",
       "7          234                 6                  NaN                    NaN   \n",
       "8          477                 7                  1.0                    5.0   \n",
       "9          477                 7                  NaN                    NaN   \n",
       "\n",
       "   ptc_to_start_codon  ptc_less_than_150nt_to_start  ptc_exon_length  \\\n",
       "0                 NaN                         False              NaN   \n",
       "1               351.0                         False             76.0   \n",
       "2                 NaN                         False              NaN   \n",
       "3                96.0                          True            722.0   \n",
       "4                 NaN                         False              NaN   \n",
       "5               348.0                         False            111.0   \n",
       "6                 NaN                         False              NaN   \n",
       "7                 NaN                         False              NaN   \n",
       "8               348.0                         False            111.0   \n",
       "9                 NaN                         False              NaN   \n",
       "\n",
       "   stop_codon_distance  ptc_to_intron  likely_misannotated  \n",
       "0                  0.0            NaN                False  \n",
       "1                249.0           41.0                False  \n",
       "2                  0.0            NaN                False  \n",
       "3                426.0          174.0                False  \n",
       "4                  0.0            NaN                False  \n",
       "5                426.0           33.0                False  \n",
       "6                  0.0            NaN                False  \n",
       "7                  NaN            NaN                 True  \n",
       "8                426.0           33.0                False  \n",
       "9                  0.0            NaN                False  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7a75836-f938-4121-bec9-e82928097e9a",
   "metadata": {},
   "source": [
    "# Compute NMD-rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "e5b4e09d-0427-49d4-8d1c-93b4312308d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_nmd_escape_rules(row):\n",
    "\n",
    "    \"\"\"\n",
    "    Evaluate whether a premature stop codon in a transcript is likely to escape nonsense-mediated decay (NMD) based on\n",
    "    established biological rules. This function applies five NMD escape rules to determine if a premature termination\n",
    "    codon (PTC) is likely to escape degradation:\n",
    "    1. Last exon rule: The PTC is in the last exon\n",
    "    2. 50nt penultimate rule: The PTC is within 50 nucleotides upstream of the last exon junction\n",
    "    3. Long exon rule: The PTC is in an exon with >407 nucleotides\n",
    "    4. Start proximal rule: The PTC is within 150 nucleotides of the start codon\n",
    "    5. Single exon rule: The transcript where the PTC lays consists only of a single exon\n",
    "    A PTC is considered to escape NMD if it satisfies any of the above rules.\n",
    "\n",
    "    :param row: A row of the DataFrame including alt_is_premature (bool), alt_first_stop_pos (int),\n",
    "                alt_stop_codon_exons (list[int]), transcript_exon_info (list[tuple[exon_number (int), exon_length (int)]]),\n",
    "                alt_start_codon_pos (int)\n",
    "    :return: A dictionary with boolean flags for each rule and overall NMD escape\n",
    "    \"\"\"\n",
    "\n",
    "    # Only relevant for premature stop codons\n",
    "    if not row.get(\"alt_is_premature\"):\n",
    "        return {\n",
    "            \"nmd_last_exon_rule\": False,\n",
    "            \"nmd_50nt_penultimate_rule\": False,\n",
    "            \"nmd_long_exon_rule\": False,\n",
    "            \"nmd_start_proximal_rule\": False,\n",
    "            \"nmd_single_exon_rule\": False,\n",
    "            \"nmd_escape\": False\n",
    "        }\n",
    "\n",
    "    # Extract relevant data\n",
    "    stop_pos = row.get(\"alt_first_stop_pos\")\n",
    "    stop_exons = row.get(\"alt_stop_codon_exons\") or []\n",
    "    exon_info = row.get(\"transcript_exon_info\") or []\n",
    "    start_pos = row.get(\"alt_start_codon_pos\")\n",
    "\n",
    "    total_exons = row.get(\"total_exon_count\")\n",
    "    downstream_exons = row.get(\"downstream_exon_count\")\n",
    "    ptc_exon_length = row.get(\"ptc_exon_length\")\n",
    "\n",
    "    # Preprocess exon info\n",
    "    sorted_exons = sorted(exon_info, key=lambda x: x[0])\n",
    "    exon_length_map = {exon_num: length for exon_num, length in sorted_exons}\n",
    "    exon_offsets = {}\n",
    "    offset = 0\n",
    "    for exon_num, length in sorted_exons:\n",
    "        exon_offsets[exon_num] = (offset, offset + length)\n",
    "        offset += length\n",
    "\n",
    "    # Single exon rule\n",
    "    # rule_single_exon = len(sorted_exons) == 1 # old code\n",
    "    rule_single_exon = total_exons == 1\n",
    "\n",
    "    # Last exon rule\n",
    "    #if not rule_single_exon:\n",
    "    #    last_exon = sorted_exons[-1][0] if sorted_exons else None\n",
    "    #    rule_last_exon = bool(stop_exons and max(stop_exons) == last_exon)\n",
    "    #else: rule_last_exon = False\n",
    "    rule_last_exon = downstream_exons == 0 if downstream_exons is not None else False\n",
    "\n",
    "    # 50nt from penultimate exon end\n",
    "    if len(sorted_exons) >= 2:\n",
    "        penultimate_exon_num, penultimate_len = sorted_exons[-2]\n",
    "        pen_start, pen_end = exon_offsets.get(penultimate_exon_num, (None, None))\n",
    "        rule_50nt_penultimate = pen_end is not None and (stop_pos >= pen_end - 50) and (stop_pos < pen_end)\n",
    "    else:\n",
    "        rule_50nt_penultimate = False\n",
    "\n",
    "    # Long exon rule (with exon longer than >407nt)\n",
    "    # rule_long_exon = any(exon_length_map.get(exon, 0) > 407 for exon in stop_exons) # old code\n",
    "    rule_long_exon = ptc_exon_length is not None and ptc_exon_length > 407\n",
    "\n",
    "    # Start-proximal rule (closer than 150nt from the start codon)\n",
    "    rule_start_proximal = start_pos is not None and stop_pos is not None and (stop_pos - start_pos) < 150 and (stop_pos - start_pos) >= 0\n",
    "\n",
    "    # NMD escape if any rule is true\n",
    "    escape = rule_last_exon or rule_50nt_penultimate or rule_long_exon or rule_start_proximal or rule_single_exon\n",
    "\n",
    "    return {\n",
    "        \"nmd_last_exon_rule\": rule_last_exon,\n",
    "        \"nmd_50nt_penultimate_rule\": rule_50nt_penultimate,\n",
    "        \"nmd_long_exon_rule\": rule_long_exon,\n",
    "        \"nmd_start_proximal_rule\": rule_start_proximal,\n",
    "        \"nmd_single_exon_rule\": rule_single_exon,\n",
    "        \"nmd_escape\": escape\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "25fcbe00-96a0-414d-9a8a-eddefc5e0c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_results = results.apply(evaluate_nmd_escape_rules, axis=1, result_type='expand')\n",
    "results = pd.concat([results, nmd_results], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "176b45d2-f000-4ec7-9642-e9a68b6c38c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "      <th>cds_in_transcript</th>\n",
       "      <th>ref_start_codon_pos</th>\n",
       "      <th>ref_start_codon_exon</th>\n",
       "      <th>ref_last_codon</th>\n",
       "      <th>ref_valid_stop</th>\n",
       "      <th>ref_first_stop_codon</th>\n",
       "      <th>ref_first_stop_pos</th>\n",
       "      <th>ref_num_stop_codons</th>\n",
       "      <th>ref_all_stop_codons</th>\n",
       "      <th>ref_stop_codon_exons</th>\n",
       "      <th>ref_is_premature</th>\n",
       "      <th>alt_start_codon_pos</th>\n",
       "      <th>alt_start_codon_exon</th>\n",
       "      <th>alt_last_codon</th>\n",
       "      <th>alt_valid_stop</th>\n",
       "      <th>alt_first_stop_codon</th>\n",
       "      <th>alt_first_stop_pos</th>\n",
       "      <th>alt_num_stop_codons</th>\n",
       "      <th>alt_all_stop_codons</th>\n",
       "      <th>alt_stop_codon_exons</th>\n",
       "      <th>alt_is_premature</th>\n",
       "      <th>start_loss</th>\n",
       "      <th>stop_loss</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>alt_transcript_seq</th>\n",
       "      <th>alt_transcript_length</th>\n",
       "      <th>transcript_exon_info</th>\n",
       "      <th>transcript_start_codon_pos</th>\n",
       "      <th>transcript_start_codon_exon</th>\n",
       "      <th>transcript_last_codon</th>\n",
       "      <th>transcript_valid_stop</th>\n",
       "      <th>transcript_first_stop_codon</th>\n",
       "      <th>transcript_first_stop_pos</th>\n",
       "      <th>transcript_num_stop_codons</th>\n",
       "      <th>transcript_all_stop_codons</th>\n",
       "      <th>transcript_stop_codon_exons</th>\n",
       "      <th>utr3_length</th>\n",
       "      <th>utr5_length</th>\n",
       "      <th>total_exon_count</th>\n",
       "      <th>upstream_exon_count</th>\n",
       "      <th>downstream_exon_count</th>\n",
       "      <th>ptc_to_start_codon</th>\n",
       "      <th>ptc_less_than_150nt_to_start</th>\n",
       "      <th>ptc_exon_length</th>\n",
       "      <th>stop_codon_distance</th>\n",
       "      <th>ptc_to_intron</th>\n",
       "      <th>likely_misannotated</th>\n",
       "      <th>nmd_last_exon_rule</th>\n",
       "      <th>nmd_50nt_penultimate_rule</th>\n",
       "      <th>nmd_long_exon_rule</th>\n",
       "      <th>nmd_start_proximal_rule</th>\n",
       "      <th>nmd_single_exon_rule</th>\n",
       "      <th>nmd_escape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1154</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>351</td>\n",
       "      <td>2</td>\n",
       "      <td>[(351, TGA), (447, TGA)]</td>\n",
       "      <td>[4, 5]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1792.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1154</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>351.0</td>\n",
       "      <td>False</td>\n",
       "      <td>76.0</td>\n",
       "      <td>249.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1154</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>96</td>\n",
       "      <td>8</td>\n",
       "      <td>[(96, TGA), (102, TGA), (141, TAA), (144, TGA)...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 2, 2, 4]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1264.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>452</td>\n",
       "      <td>361</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>True</td>\n",
       "      <td>722.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>452</td>\n",
       "      <td>361</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3690.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2858</td>\n",
       "      <td>129</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>False</td>\n",
       "      <td>111.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2858</td>\n",
       "      <td>129</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GTG</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>42720608</td>\n",
       "      <td>42751741</td>\n",
       "      <td>AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...</td>\n",
       "      <td>580</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>234</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4042.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2862</td>\n",
       "      <td>477</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>False</td>\n",
       "      <td>111.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2862</td>\n",
       "      <td>477</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  cds_in_transcript  \\\n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]               True   \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]               True   \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]               True   \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]              False   \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "\n",
       "  ref_start_codon_pos ref_start_codon_exon ref_last_codon ref_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    1            GAT          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  ref_first_stop_codon ref_first_stop_pos ref_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TAA                600                   1   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                522                   1   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                774                   1   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                774                   1   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "  ref_all_stop_codons ref_stop_codon_exons ref_is_premature  \\\n",
       "0        [(600, TAA)]                  [5]            False   \n",
       "1        [(600, TAA)]                  [5]            False   \n",
       "2        [(600, TAA)]                  [5]            False   \n",
       "3        [(522, TGA)]                  [4]            False   \n",
       "4        [(522, TGA)]                  [4]            False   \n",
       "5        [(774, TGA)]                  [6]            False   \n",
       "6        [(774, TGA)]                  [6]            False   \n",
       "7                  []                   []            False   \n",
       "8        [(774, TGA)]                  [6]            False   \n",
       "9        [(774, TGA)]                  [6]            False   \n",
       "\n",
       "  alt_start_codon_pos alt_start_codon_exon alt_last_codon alt_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    2            GTG          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  alt_first_stop_codon alt_first_stop_pos alt_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TGA                351                   2   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                 96                   8   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                348                   8   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                348                   8   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "                                 alt_all_stop_codons  \\\n",
       "0                                       [(600, TAA)]   \n",
       "1                           [(351, TGA), (447, TGA)]   \n",
       "2                                       [(600, TAA)]   \n",
       "3  [(96, TGA), (102, TGA), (141, TAA), (144, TGA)...   \n",
       "4                                       [(522, TGA)]   \n",
       "5  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "6                                       [(774, TGA)]   \n",
       "7                                                 []   \n",
       "8  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "9                                       [(774, TGA)]   \n",
       "\n",
       "       alt_stop_codon_exons alt_is_premature  start_loss  stop_loss  \\\n",
       "0                       [5]            False       False      False   \n",
       "1                    [4, 5]             True       False      False   \n",
       "2                       [5]            False       False      False   \n",
       "3  [1, 1, 1, 1, 1, 2, 2, 4]             True       False      False   \n",
       "4                       [4]            False       False      False   \n",
       "5  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "6                       [6]            False       False      False   \n",
       "7                        []            False       False       True   \n",
       "8  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "9                       [6]            False       False      False   \n",
       "\n",
       "   transcript_start  transcript_end  \\\n",
       "0          42704779        42751728   \n",
       "1          42704779        42751728   \n",
       "2          42704779        42751728   \n",
       "3          42710849        42751849   \n",
       "4          42710849        42751849   \n",
       "5          42708443        42751748   \n",
       "6          42708443        42751748   \n",
       "7          42720608        42751741   \n",
       "8          42708439        42751866   \n",
       "9          42708439        42751866   \n",
       "\n",
       "                                      transcript_seq  transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "7  AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...                580   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "\n",
       "                                  alt_transcript_seq  alt_transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1792.0   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1264.0   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1338.0   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3690.0   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3764.0   \n",
       "7                                               None                    NaN   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4042.0   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4116.0   \n",
       "\n",
       "                                transcript_exon_info  \\\n",
       "0  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "1  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "2  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "3  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "4  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "5  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "6  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "7  [(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...   \n",
       "8  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "9  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "\n",
       "  transcript_start_codon_pos transcript_start_codon_exon  \\\n",
       "0                       None                        None   \n",
       "1                       None                        None   \n",
       "2                       None                        None   \n",
       "3                       None                        None   \n",
       "4                       None                        None   \n",
       "5                       None                        None   \n",
       "6                       None                        None   \n",
       "7                       None                        None   \n",
       "8                       None                        None   \n",
       "9                       None                        None   \n",
       "\n",
       "  transcript_last_codon transcript_valid_stop transcript_first_stop_codon  \\\n",
       "0                  None                  None                        None   \n",
       "1                  None                  None                        None   \n",
       "2                  None                  None                        None   \n",
       "3                  None                  None                        None   \n",
       "4                  None                  None                        None   \n",
       "5                  None                  None                        None   \n",
       "6                  None                  None                        None   \n",
       "7                  None                  None                        None   \n",
       "8                  None                  None                        None   \n",
       "9                  None                  None                        None   \n",
       "\n",
       "  transcript_first_stop_pos transcript_num_stop_codons  \\\n",
       "0                      None                       None   \n",
       "1                      None                       None   \n",
       "2                      None                       None   \n",
       "3                      None                       None   \n",
       "4                      None                       None   \n",
       "5                      None                       None   \n",
       "6                      None                       None   \n",
       "7                      None                       None   \n",
       "8                      None                       None   \n",
       "9                      None                       None   \n",
       "\n",
       "  transcript_all_stop_codons transcript_stop_codon_exons  utr3_length  \\\n",
       "0                       None                        None         1154   \n",
       "1                       None                        None         1154   \n",
       "2                       None                        None         1154   \n",
       "3                       None                        None          452   \n",
       "4                       None                        None          452   \n",
       "5                       None                        None         2858   \n",
       "6                       None                        None         2858   \n",
       "7                       None                        None            0   \n",
       "8                       None                        None         2862   \n",
       "9                       None                        None         2862   \n",
       "\n",
       "   utr5_length  total_exon_count  upstream_exon_count  downstream_exon_count  \\\n",
       "0          109                 6                  NaN                    NaN   \n",
       "1          109                 6                  3.0                    2.0   \n",
       "2          109                 6                  NaN                    NaN   \n",
       "3          361                 7                  0.0                    6.0   \n",
       "4          361                 7                  NaN                    NaN   \n",
       "5          129                 7                  1.0                    5.0   \n",
       "6          129                 7                  NaN                    NaN   \n",
       "7          234                 6                  NaN                    NaN   \n",
       "8          477                 7                  1.0                    5.0   \n",
       "9          477                 7                  NaN                    NaN   \n",
       "\n",
       "   ptc_to_start_codon  ptc_less_than_150nt_to_start  ptc_exon_length  \\\n",
       "0                 NaN                         False              NaN   \n",
       "1               351.0                         False             76.0   \n",
       "2                 NaN                         False              NaN   \n",
       "3                96.0                          True            722.0   \n",
       "4                 NaN                         False              NaN   \n",
       "5               348.0                         False            111.0   \n",
       "6                 NaN                         False              NaN   \n",
       "7                 NaN                         False              NaN   \n",
       "8               348.0                         False            111.0   \n",
       "9                 NaN                         False              NaN   \n",
       "\n",
       "   stop_codon_distance  ptc_to_intron  likely_misannotated  \\\n",
       "0                  0.0            NaN                False   \n",
       "1                249.0           41.0                False   \n",
       "2                  0.0            NaN                False   \n",
       "3                426.0          174.0                False   \n",
       "4                  0.0            NaN                False   \n",
       "5                426.0           33.0                False   \n",
       "6                  0.0            NaN                False   \n",
       "7                  NaN            NaN                 True   \n",
       "8                426.0           33.0                False   \n",
       "9                  0.0            NaN                False   \n",
       "\n",
       "   nmd_last_exon_rule  nmd_50nt_penultimate_rule  nmd_long_exon_rule  \\\n",
       "0               False                      False               False   \n",
       "1               False                      False               False   \n",
       "2               False                      False               False   \n",
       "3               False                      False                True   \n",
       "4               False                      False               False   \n",
       "5               False                      False               False   \n",
       "6               False                      False               False   \n",
       "7               False                      False               False   \n",
       "8               False                      False               False   \n",
       "9               False                      False               False   \n",
       "\n",
       "   nmd_start_proximal_rule  nmd_single_exon_rule  nmd_escape  \n",
       "0                    False                 False       False  \n",
       "1                    False                 False       False  \n",
       "2                    False                 False       False  \n",
       "3                     True                 False        True  \n",
       "4                    False                 False       False  \n",
       "5                    False                 False       False  \n",
       "6                    False                 False       False  \n",
       "7                    False                 False       False  \n",
       "8                    False                 False       False  \n",
       "9                    False                 False       False  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40de9b04-c18e-4065-b2c3-7653757230b7",
   "metadata": {},
   "source": [
    "# Write output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2c34780d-615f-4dd7-88dc-a4d0570cbc90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 9632\n",
      "-rw-r--r--. 1 l_schroeder users   38865 Jan 11 14:06 create_test_VCF.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users 6252969 Dec 18 14:04 NMD.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users  825558 Jan 11 16:22 nmd-vep.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users    2465 Dec 18 14:34 README.md\n",
      "-rw-r--r--. 1 l_schroeder users 1307856 Dec 18 14:04 train.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users 1307856 Jan 11 13:57 train_new.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users   54222 Dec 18 14:04 validation_MMRF_TARGET.ipynb\n",
      "-rw-r--r--. 1 l_schroeder users   54222 Jan 11 14:05 validation_new.ipynb\n"
     ]
    }
   ],
   "source": [
    "ls -ll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96b6c7b1-dd4b-434f-8875-8aabfbb39979",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = \"~/NMD/nmd-variant-effect-prediction/test_variant/output.csv\"\n",
    "\n",
    "results.to_csv(output_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "39a1bb24-9ecc-4a0a-a6e0-4b9d8d49203e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       nmd_last_exon_rule  nmd_50nt_penultimate_rule  nmd_long_exon_rule  \\\n",
      "False                10.0                       10.0                   9   \n",
      "True                  NaN                        NaN                   1   \n",
      "\n",
      "       nmd_start_proximal_rule  nmd_single_exon_rule  nmd_escape  \n",
      "False                        9                  10.0           9  \n",
      "True                         1                   NaN           1  \n"
     ]
    }
   ],
   "source": [
    "rule_cols = [\n",
    "    \"nmd_last_exon_rule\",\n",
    "    \"nmd_50nt_penultimate_rule\",\n",
    "    \"nmd_long_exon_rule\",\n",
    "    \"nmd_start_proximal_rule\",\n",
    "    \"nmd_single_exon_rule\",\n",
    "    \"nmd_escape\",\n",
    "]\n",
    "\n",
    "summary = results[rule_cols].apply(lambda x: x.value_counts(dropna=False))\n",
    "print(summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9f4777c4-5e85-4788-897a-b923e8c88e65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref_cds_start</th>\n",
       "      <th>ref_cds_stop</th>\n",
       "      <th>ref_cds_seq</th>\n",
       "      <th>ref_cds_len</th>\n",
       "      <th>alt_cds_start</th>\n",
       "      <th>alt_cds_stop</th>\n",
       "      <th>alt_cds_seq</th>\n",
       "      <th>alt_cds_len</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start_variant</th>\n",
       "      <th>end_variant</th>\n",
       "      <th>ref_cds_info</th>\n",
       "      <th>alt_cds_info</th>\n",
       "      <th>cds_in_transcript</th>\n",
       "      <th>ref_start_codon_pos</th>\n",
       "      <th>ref_start_codon_exon</th>\n",
       "      <th>ref_last_codon</th>\n",
       "      <th>ref_valid_stop</th>\n",
       "      <th>ref_first_stop_codon</th>\n",
       "      <th>ref_first_stop_pos</th>\n",
       "      <th>ref_num_stop_codons</th>\n",
       "      <th>ref_all_stop_codons</th>\n",
       "      <th>ref_stop_codon_exons</th>\n",
       "      <th>ref_is_premature</th>\n",
       "      <th>alt_start_codon_pos</th>\n",
       "      <th>alt_start_codon_exon</th>\n",
       "      <th>alt_last_codon</th>\n",
       "      <th>alt_valid_stop</th>\n",
       "      <th>alt_first_stop_codon</th>\n",
       "      <th>alt_first_stop_pos</th>\n",
       "      <th>alt_num_stop_codons</th>\n",
       "      <th>alt_all_stop_codons</th>\n",
       "      <th>alt_stop_codon_exons</th>\n",
       "      <th>alt_is_premature</th>\n",
       "      <th>start_loss</th>\n",
       "      <th>stop_loss</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>alt_transcript_seq</th>\n",
       "      <th>alt_transcript_length</th>\n",
       "      <th>transcript_exon_info</th>\n",
       "      <th>transcript_start_codon_pos</th>\n",
       "      <th>transcript_start_codon_exon</th>\n",
       "      <th>transcript_last_codon</th>\n",
       "      <th>transcript_valid_stop</th>\n",
       "      <th>transcript_first_stop_codon</th>\n",
       "      <th>transcript_first_stop_pos</th>\n",
       "      <th>transcript_num_stop_codons</th>\n",
       "      <th>transcript_all_stop_codons</th>\n",
       "      <th>transcript_stop_codon_exons</th>\n",
       "      <th>utr3_length</th>\n",
       "      <th>utr5_length</th>\n",
       "      <th>total_exon_count</th>\n",
       "      <th>upstream_exon_count</th>\n",
       "      <th>downstream_exon_count</th>\n",
       "      <th>ptc_to_start_codon</th>\n",
       "      <th>ptc_less_than_150nt_to_start</th>\n",
       "      <th>ptc_exon_length</th>\n",
       "      <th>stop_codon_distance</th>\n",
       "      <th>ptc_to_intron</th>\n",
       "      <th>likely_misannotated</th>\n",
       "      <th>nmd_last_exon_rule</th>\n",
       "      <th>nmd_50nt_penultimate_rule</th>\n",
       "      <th>nmd_long_exon_rule</th>\n",
       "      <th>nmd_start_proximal_rule</th>\n",
       "      <th>nmd_single_exon_rule</th>\n",
       "      <th>nmd_escape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>42706139</td>\n",
       "      <td>42706140</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1154</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>529</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>351</td>\n",
       "      <td>2</td>\n",
       "      <td>[(351, TGA), (447, TGA)]</td>\n",
       "      <td>[4, 5]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1792.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1154</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>351.0</td>\n",
       "      <td>False</td>\n",
       "      <td>76.0</td>\n",
       "      <td>249.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000319104.7_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>42705933</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>603</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>[(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "      <td>600</td>\n",
       "      <td>1</td>\n",
       "      <td>[(600, TAA)]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42704779</td>\n",
       "      <td>42751728</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866</td>\n",
       "      <td>GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1154</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>451</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>96</td>\n",
       "      <td>8</td>\n",
       "      <td>[(96, TGA), (102, TGA), (141, TAA), (144, TGA)...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 2, 2, 4]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1264.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>452</td>\n",
       "      <td>361</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>True</td>\n",
       "      <td>722.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000526349.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42725216</td>\n",
       "      <td>ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...</td>\n",
       "      <td>525</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 70)]</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>522</td>\n",
       "      <td>1</td>\n",
       "      <td>[(522, TGA)]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42710849</td>\n",
       "      <td>42751849</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338</td>\n",
       "      <td>GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...</td>\n",
       "      <td>1338.0</td>\n",
       "      <td>[(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>452</td>\n",
       "      <td>361</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3690.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2858</td>\n",
       "      <td>129</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>False</td>\n",
       "      <td>111.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000527424.6_3</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708443</td>\n",
       "      <td>42751748</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764</td>\n",
       "      <td>GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...</td>\n",
       "      <td>3764.0</td>\n",
       "      <td>[(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2858</td>\n",
       "      <td>129</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000531440.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>349</td>\n",
       "      <td>42720605</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>322</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 27), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>[(1, 0), (2, 109), (3, 76), (4, 137)]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GTG</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>42720608</td>\n",
       "      <td>42751741</td>\n",
       "      <td>AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...</td>\n",
       "      <td>580</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>234</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>703</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>42720557</td>\n",
       "      <td>42720632</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>348</td>\n",
       "      <td>8</td>\n",
       "      <td>[(348, TGA), (354, TGA), (393, TAA), (396, TGA...</td>\n",
       "      <td>[2, 2, 4, 4, 4, 6, 6, 6]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4042.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2862</td>\n",
       "      <td>477</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>False</td>\n",
       "      <td>111.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000534961.5_2</td>\n",
       "      <td>.</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>42711301</td>\n",
       "      <td>42743007</td>\n",
       "      <td>ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...</td>\n",
       "      <td>777</td>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000120925.16_6</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42720558</td>\n",
       "      <td>42720559</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>[(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "      <td>774</td>\n",
       "      <td>1</td>\n",
       "      <td>[(774, TGA)]</td>\n",
       "      <td>[6]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42708439</td>\n",
       "      <td>42751866</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116</td>\n",
       "      <td>AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...</td>\n",
       "      <td>4116.0</td>\n",
       "      <td>[(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2862</td>\n",
       "      <td>477</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id variant_id  ref_cds_start  ref_cds_stop  \\\n",
       "0  ENST00000319104.7_2          .       42705933      42743007   \n",
       "1  ENST00000319104.7_2          .       42705933      42743007   \n",
       "2  ENST00000319104.7_2          .       42705933      42743007   \n",
       "3  ENST00000526349.5_2          .       42711301      42725216   \n",
       "4  ENST00000526349.5_2          .       42711301      42725216   \n",
       "5  ENST00000527424.6_3          .       42711301      42743007   \n",
       "6  ENST00000527424.6_3          .       42711301      42743007   \n",
       "7  ENST00000531440.5_2          .       42720605      42743007   \n",
       "8  ENST00000534961.5_2          .       42711301      42743007   \n",
       "9  ENST00000534961.5_2          .       42711301      42743007   \n",
       "\n",
       "                                         ref_cds_seq  ref_cds_len  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          349   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777   \n",
       "\n",
       "   alt_cds_start  alt_cds_stop  \\\n",
       "0       42705933      42743007   \n",
       "1       42705933      42743007   \n",
       "2       42705933      42743007   \n",
       "3       42711301      42725216   \n",
       "4       42711301      42725216   \n",
       "5       42711301      42743007   \n",
       "6       42711301      42743007   \n",
       "7       42720605      42743007   \n",
       "8       42711301      42743007   \n",
       "9       42711301      42743007   \n",
       "\n",
       "                                         alt_cds_seq  alt_cds_len chromosome  \\\n",
       "0  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "1  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          529       chr8   \n",
       "2  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          603       chr8   \n",
       "3  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          451       chr8   \n",
       "4  ATGTACTGTCCCATCTGCCTGCACCAAGCCTCCTTCCCGGTGGAGA...          525       chr8   \n",
       "5  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "6  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "7  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          322       chr8   \n",
       "8  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          703       chr8   \n",
       "9  ATGGCCAAATATCAAGGTGAAGTTCAAAGTTTGAAACTGGATGATG...          777       chr8   \n",
       "\n",
       "                gene_id strand  \\\n",
       "0  ENSG00000120925.16_6      -   \n",
       "1  ENSG00000120925.16_6      -   \n",
       "2  ENSG00000120925.16_6      -   \n",
       "3  ENSG00000120925.16_6      -   \n",
       "4  ENSG00000120925.16_6      -   \n",
       "5  ENSG00000120925.16_6      -   \n",
       "6  ENSG00000120925.16_6      -   \n",
       "7  ENSG00000120925.16_6      -   \n",
       "8  ENSG00000120925.16_6      -   \n",
       "9  ENSG00000120925.16_6      -   \n",
       "\n",
       "                                                 ref alt  start_variant  \\\n",
       "0                                                  C   C       42706139   \n",
       "1  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "2                                                  C   T       42720558   \n",
       "3  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "4                                                  C   T       42720558   \n",
       "5  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "6                                                  C   T       42720558   \n",
       "7  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "8  CCGTTTGTCTACAGATTGGACAACTGATTGCCCCAAGCCATGAACC...   C       42720557   \n",
       "9                                                  C   T       42720558   \n",
       "\n",
       "   end_variant                                       ref_cds_info  \\\n",
       "0     42706140   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "1     42720632   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "2     42720559   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]   \n",
       "3     42720632             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "4     42720559             [(1, 270), (2, 111), (3, 74), (4, 70)]   \n",
       "5     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "6     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "7     42720632             [(1, 27), (2, 109), (3, 76), (4, 137)]   \n",
       "8     42720632  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "9     42720559  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...   \n",
       "\n",
       "                                        alt_cds_info  cds_in_transcript  \\\n",
       "0   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "1    [(1, 207), (2, 0), (3, 109), (4, 76), (5, 137)]               True   \n",
       "2   [(1, 207), (2, 74), (3, 109), (4, 76), (5, 137)]               True   \n",
       "3              [(1, 270), (2, 111), (3, 0), (4, 70)]               True   \n",
       "4             [(1, 270), (2, 111), (3, 74), (4, 70)]               True   \n",
       "5  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "6  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "7              [(1, 0), (2, 109), (3, 76), (4, 137)]              False   \n",
       "8  [(1, 270), (2, 111), (3, 0), (4, 109), (5, 76)...               True   \n",
       "9  [(1, 270), (2, 111), (3, 74), (4, 109), (5, 76...               True   \n",
       "\n",
       "  ref_start_codon_pos ref_start_codon_exon ref_last_codon ref_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    1            GAT          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  ref_first_stop_codon ref_first_stop_pos ref_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TAA                600                   1   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                522                   1   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                774                   1   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                774                   1   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "  ref_all_stop_codons ref_stop_codon_exons ref_is_premature  \\\n",
       "0        [(600, TAA)]                  [5]            False   \n",
       "1        [(600, TAA)]                  [5]            False   \n",
       "2        [(600, TAA)]                  [5]            False   \n",
       "3        [(522, TGA)]                  [4]            False   \n",
       "4        [(522, TGA)]                  [4]            False   \n",
       "5        [(774, TGA)]                  [6]            False   \n",
       "6        [(774, TGA)]                  [6]            False   \n",
       "7                  []                   []            False   \n",
       "8        [(774, TGA)]                  [6]            False   \n",
       "9        [(774, TGA)]                  [6]            False   \n",
       "\n",
       "  alt_start_codon_pos alt_start_codon_exon alt_last_codon alt_valid_stop  \\\n",
       "0                   0                    1            TAA           True   \n",
       "1                   0                    1            TAA           True   \n",
       "2                   0                    1            TAA           True   \n",
       "3                   0                    1            TGA           True   \n",
       "4                   0                    1            TGA           True   \n",
       "5                   0                    1            TGA           True   \n",
       "6                   0                    1            TGA           True   \n",
       "7                   0                    2            GTG          False   \n",
       "8                   0                    1            TGA           True   \n",
       "9                   0                    1            TGA           True   \n",
       "\n",
       "  alt_first_stop_codon alt_first_stop_pos alt_num_stop_codons  \\\n",
       "0                  TAA                600                   1   \n",
       "1                  TGA                351                   2   \n",
       "2                  TAA                600                   1   \n",
       "3                  TGA                 96                   8   \n",
       "4                  TGA                522                   1   \n",
       "5                  TGA                348                   8   \n",
       "6                  TGA                774                   1   \n",
       "7                 None               None                   0   \n",
       "8                  TGA                348                   8   \n",
       "9                  TGA                774                   1   \n",
       "\n",
       "                                 alt_all_stop_codons  \\\n",
       "0                                       [(600, TAA)]   \n",
       "1                           [(351, TGA), (447, TGA)]   \n",
       "2                                       [(600, TAA)]   \n",
       "3  [(96, TGA), (102, TGA), (141, TAA), (144, TGA)...   \n",
       "4                                       [(522, TGA)]   \n",
       "5  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "6                                       [(774, TGA)]   \n",
       "7                                                 []   \n",
       "8  [(348, TGA), (354, TGA), (393, TAA), (396, TGA...   \n",
       "9                                       [(774, TGA)]   \n",
       "\n",
       "       alt_stop_codon_exons alt_is_premature  start_loss  stop_loss  \\\n",
       "0                       [5]            False       False      False   \n",
       "1                    [4, 5]             True       False      False   \n",
       "2                       [5]            False       False      False   \n",
       "3  [1, 1, 1, 1, 1, 2, 2, 4]             True       False      False   \n",
       "4                       [4]            False       False      False   \n",
       "5  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "6                       [6]            False       False      False   \n",
       "7                        []            False       False       True   \n",
       "8  [2, 2, 4, 4, 4, 6, 6, 6]             True       False      False   \n",
       "9                       [6]            False       False      False   \n",
       "\n",
       "   transcript_start  transcript_end  \\\n",
       "0          42704779        42751728   \n",
       "1          42704779        42751728   \n",
       "2          42704779        42751728   \n",
       "3          42710849        42751849   \n",
       "4          42710849        42751849   \n",
       "5          42708443        42751748   \n",
       "6          42708443        42751748   \n",
       "7          42720608        42751741   \n",
       "8          42708439        42751866   \n",
       "9          42708439        42751866   \n",
       "\n",
       "                                      transcript_seq  transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...               1866   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...               1338   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...               3764   \n",
       "7  AAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGG...                580   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...               4116   \n",
       "\n",
       "                                  alt_transcript_seq  alt_transcript_length  \\\n",
       "0  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "1  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1792.0   \n",
       "2  GACGTCTAGTGGGTTGCCCGGGAGGGGTGGCGGGAGCGGTCCTGGA...                 1866.0   \n",
       "3  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1264.0   \n",
       "4  GGCGAGCACGCGCTCCGCCCTGGAGGCTGCGGCGACGGGTCCTCCT...                 1338.0   \n",
       "5  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3690.0   \n",
       "6  GCACTCCAAATTAGAAAGGGGACGTCTAGTGGGTTGCCCGGGAGGG...                 3764.0   \n",
       "7                                               None                    NaN   \n",
       "8  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4042.0   \n",
       "9  AGTTCGCTGCGTGTCGAGGCGAGCACGCGCTCCGCCCTGGAGGCTG...                 4116.0   \n",
       "\n",
       "                                transcript_exon_info  \\\n",
       "0  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "1  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "2  [(6, 102), (5, 144), (4, 76), (3, 109), (2, 74...   \n",
       "3  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "4  [(7, 223), (6, 23), (5, 76), (4, 109), (3, 74)...   \n",
       "5  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "6  [(7, 122), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "7  [(6, 115), (5, 112), (4, 144), (3, 76), (2, 10...   \n",
       "8  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "9  [(7, 470), (6, 144), (5, 76), (4, 109), (3, 74...   \n",
       "\n",
       "  transcript_start_codon_pos transcript_start_codon_exon  \\\n",
       "0                       None                        None   \n",
       "1                       None                        None   \n",
       "2                       None                        None   \n",
       "3                       None                        None   \n",
       "4                       None                        None   \n",
       "5                       None                        None   \n",
       "6                       None                        None   \n",
       "7                       None                        None   \n",
       "8                       None                        None   \n",
       "9                       None                        None   \n",
       "\n",
       "  transcript_last_codon transcript_valid_stop transcript_first_stop_codon  \\\n",
       "0                  None                  None                        None   \n",
       "1                  None                  None                        None   \n",
       "2                  None                  None                        None   \n",
       "3                  None                  None                        None   \n",
       "4                  None                  None                        None   \n",
       "5                  None                  None                        None   \n",
       "6                  None                  None                        None   \n",
       "7                  None                  None                        None   \n",
       "8                  None                  None                        None   \n",
       "9                  None                  None                        None   \n",
       "\n",
       "  transcript_first_stop_pos transcript_num_stop_codons  \\\n",
       "0                      None                       None   \n",
       "1                      None                       None   \n",
       "2                      None                       None   \n",
       "3                      None                       None   \n",
       "4                      None                       None   \n",
       "5                      None                       None   \n",
       "6                      None                       None   \n",
       "7                      None                       None   \n",
       "8                      None                       None   \n",
       "9                      None                       None   \n",
       "\n",
       "  transcript_all_stop_codons transcript_stop_codon_exons  utr3_length  \\\n",
       "0                       None                        None         1154   \n",
       "1                       None                        None         1154   \n",
       "2                       None                        None         1154   \n",
       "3                       None                        None          452   \n",
       "4                       None                        None          452   \n",
       "5                       None                        None         2858   \n",
       "6                       None                        None         2858   \n",
       "7                       None                        None            0   \n",
       "8                       None                        None         2862   \n",
       "9                       None                        None         2862   \n",
       "\n",
       "   utr5_length  total_exon_count  upstream_exon_count  downstream_exon_count  \\\n",
       "0          109                 6                  NaN                    NaN   \n",
       "1          109                 6                  3.0                    2.0   \n",
       "2          109                 6                  NaN                    NaN   \n",
       "3          361                 7                  0.0                    6.0   \n",
       "4          361                 7                  NaN                    NaN   \n",
       "5          129                 7                  1.0                    5.0   \n",
       "6          129                 7                  NaN                    NaN   \n",
       "7          234                 6                  NaN                    NaN   \n",
       "8          477                 7                  1.0                    5.0   \n",
       "9          477                 7                  NaN                    NaN   \n",
       "\n",
       "   ptc_to_start_codon  ptc_less_than_150nt_to_start  ptc_exon_length  \\\n",
       "0                 NaN                         False              NaN   \n",
       "1               351.0                         False             76.0   \n",
       "2                 NaN                         False              NaN   \n",
       "3                96.0                          True            722.0   \n",
       "4                 NaN                         False              NaN   \n",
       "5               348.0                         False            111.0   \n",
       "6                 NaN                         False              NaN   \n",
       "7                 NaN                         False              NaN   \n",
       "8               348.0                         False            111.0   \n",
       "9                 NaN                         False              NaN   \n",
       "\n",
       "   stop_codon_distance  ptc_to_intron  likely_misannotated  \\\n",
       "0                  0.0            NaN                False   \n",
       "1                249.0           41.0                False   \n",
       "2                  0.0            NaN                False   \n",
       "3                426.0          174.0                False   \n",
       "4                  0.0            NaN                False   \n",
       "5                426.0           33.0                False   \n",
       "6                  0.0            NaN                False   \n",
       "7                  NaN            NaN                 True   \n",
       "8                426.0           33.0                False   \n",
       "9                  0.0            NaN                False   \n",
       "\n",
       "   nmd_last_exon_rule  nmd_50nt_penultimate_rule  nmd_long_exon_rule  \\\n",
       "0               False                      False               False   \n",
       "1               False                      False               False   \n",
       "2               False                      False               False   \n",
       "3               False                      False                True   \n",
       "4               False                      False               False   \n",
       "5               False                      False               False   \n",
       "6               False                      False               False   \n",
       "7               False                      False               False   \n",
       "8               False                      False               False   \n",
       "9               False                      False               False   \n",
       "\n",
       "   nmd_start_proximal_rule  nmd_single_exon_rule  nmd_escape  \n",
       "0                    False                 False       False  \n",
       "1                    False                 False       False  \n",
       "2                    False                 False       False  \n",
       "3                     True                 False        True  \n",
       "4                    False                 False       False  \n",
       "5                    False                 False       False  \n",
       "6                    False                 False       False  \n",
       "7                    False                 False       False  \n",
       "8                    False                 False       False  \n",
       "9                    False                 False       False  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.head(20)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "NMD_test",
   "language": "python",
   "name": "nmd_test"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
